// wix-style-react/dist/stories/Heading/index.story.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Heading = require('wix-style-react/Heading');

var _Heading2 = _interopRequireDefault(_Heading);

var _CodeExample = require('wix-storybook-utils/CodeExample');

var _CodeExample2 = _interopRequireDefault(_CodeExample);

var _ExampleEllipsis = require('./ExampleEllipsis');

var _ExampleEllipsis2 = _interopRequireDefault(_ExampleEllipsis);

var _ExampleEllipsis3 = require('!raw-loader!./ExampleEllipsis');

var _ExampleEllipsis4 = _interopRequireDefault(_ExampleEllipsis3);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  category: '1. Foundation',
  storyName: '1.3 Heading',
  component: _Heading2.default,
  componentPath: '../../src/Heading/Heading.js',

  componentProps: {
    children: 'Hey there, good looking',
    light: false,
    dataHook: 'storybook-heading',
    appearance: 'H1',
    ellipsis: false
  },

  examples: _react2.default.createElement(
    'div',
    null,
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Ellipsis Example', code: _ExampleEllipsis4.default },
      _react2.default.createElement(_ExampleEllipsis2.default, null)
    )
  )
};
