// wix-ui-core/dist/src/components/Input/Input.driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var test_utils_1 = require("stylable/test-utils");
var Input_st_css_1 = require("./Input.st.css");
exports.inputDriverFactory = function (_a) {
    var element = _a.element, eventTrigger = _a.eventTrigger;
    var input = element && element.querySelector('input');
    var styleUtil = new test_utils_1.StylableDOMUtil(Input_st_css_1.default);
    return {
        /** checks if element exists */
        exists: function () { return !!element; },
        hasStyleState: function (stateName) { return styleUtil.hasStyleState(element, stateName); },
        /** get input element */
        getInput: function () { return input; },
        /** get value */
        getValue: function () { return input.value; },
        /** get placeholder */
        getPlaceholder: function () { return input.placeholder; },
        /** get prefix */
        getPrefix: function () { return input.previousSibling; },
        /** get suffix */
        getSuffix: function () { return input.nextSibling; },
        /** is disabled */
        isDisabled: function () { return input.disabled; },
        /** is read only */
        isReadOnly: function () { return input.readOnly; },
        /** set value */
        setValue: function (value) {
            input.value = value;
            eventTrigger.change(input, { target: { value: value } });
        },
        /** trigger focus */
        focus: function () { return eventTrigger.focus(input); },
        /** trigger blur */
        blur: function () { return eventTrigger.blur(input); },
        /** trigger keyDown */
        keyDown: function (key) { return eventTrigger.keyDown(input, { key: key }); }
    };
};
