// playable/dist/src/modules/ui/loading-cover/loading-cover.view.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var view_1 = tslib_1.__importDefault(require("../core/view"));
var templates_1 = require("./templates");
var getElementByHook_1 = tslib_1.__importDefault(require("../core/getElementByHook"));
var htmlToElement_1 = tslib_1.__importDefault(require("../core/htmlToElement"));
var loading_cover_scss_1 = tslib_1.__importDefault(require("./loading-cover.scss"));
var LoadingCoverView = /** @class */ (function (_super) {
    tslib_1.__extends(LoadingCoverView, _super);
    function LoadingCoverView(config) {
        var _this = _super.call(this) || this;
        _this._initDOM();
        _this.setCover(config.url);
        return _this;
    }
    LoadingCoverView.prototype.getNode = function () {
        return this._$node;
    };
    LoadingCoverView.prototype._initDOM = function () {
        this._$node = htmlToElement_1.default(templates_1.loadingCoverTemplate({
            styles: this.styleNames,
        }));
        this._$image = getElementByHook_1.default(this._$node, 'loading-cover-image');
    };
    LoadingCoverView.prototype.hide = function () {
        this._$node.classList.add(this.styleNames.hidden);
    };
    LoadingCoverView.prototype.show = function () {
        this._$node.classList.remove(this.styleNames.hidden);
    };
    LoadingCoverView.prototype.setCover = function (url) {
        var _this = this;
        if (url && typeof url === 'string') {
            this._$image.classList.add(this.styleNames.hidden);
            var onImageLoad_1 = function () {
                _this._$image.classList.remove(_this.styleNames.hidden);
                _this._$image.removeEventListener('load', onImageLoad_1);
            };
            this._$image.addEventListener('load', onImageLoad_1);
            this._$image.setAttribute('src', url);
        }
    };
    LoadingCoverView.prototype.destroy = function () {
        if (this._$node.parentNode) {
            this._$node.parentNode.removeChild(this._$node);
        }
        this._$node = null;
    };
    return LoadingCoverView;
}(view_1.default));
LoadingCoverView.extendStyleNames(loading_cover_scss_1.default);
exports.default = LoadingCoverView;
