// wix-ui-core/dist/src/components/LabelWithOptions/LabelWithOptions.driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Dropdown_driver_1 = require("../../baseComponents/Dropdown/Dropdown.driver");
var Label_driver_1 = require("../Label/Label.driver");
var Checkbox_driver_1 = require("../Checkbox/Checkbox.driver");
var test_utils_1 = require("stylable/test-utils");
var LabelWithOptions_st_css_1 = require("./LabelWithOptions.st.css");
exports.labelWithOptionsDriverFactory = function (_a) {
    var element = _a.element, eventTrigger = _a.eventTrigger;
    var dropdownDriver = Dropdown_driver_1.dropdownDriverFactory({ element: element, eventTrigger: eventTrigger });
    var labelElement = element.querySelector('[data-hook="label"]');
    var suffixElement = element.querySelector('[data-hook="suffix"]');
    var domUtils = new test_utils_1.StylableDOMUtil(LabelWithOptions_st_css_1.default, element);
    var labelDriver = Label_driver_1.labelDriverFactory({
        element: labelElement,
        eventTrigger: eventTrigger,
    });
    return Object.assign({}, dropdownDriver, labelDriver, {
        getSuffix: function () { return suffixElement; },
        isRequired: function () { return domUtils.hasStyleState(element, 'required'); },
        isError: function () { return domUtils.hasStyleState(element, 'error'); },
        isDisabled: function () { return domUtils.hasStyleState(element, 'disabled'); },
        checkboxDriverAt: function (index) {
            var checkbox = domUtils.selectAll('.checkbox')[index];
            return Checkbox_driver_1.checkboxDriverFactory({ element: checkbox, eventTrigger: eventTrigger });
        },
    });
};
