// wix-style-react/dist/src/Tabs/core/withTooltip/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Tooltip = require('../../../Tooltip');

var _Tooltip2 = _interopRequireDefault(_Tooltip);

var _tabPropTypes = require('../constants/tab-prop-types');

var TabPropTypes = _interopRequireWildcard(_tabPropTypes);

var _utils = require('../utils');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var withTooltip = function withTooltip(WrappedComponent) {

  var VERTICAL_PADDING = 18;

  var Wrapper = function (_React$Component) {
    _inherits(Wrapper, _React$Component);

    function Wrapper() {
      var _ref;

      var _temp, _this, _ret;

      _classCallCheck(this, Wrapper);

      for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Wrapper.__proto__ || Object.getPrototypeOf(Wrapper)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
        hasTooltip: false
      }, _temp), _possibleConstructorReturn(_this, _ret);
    }

    _createClass(Wrapper, [{
      key: 'initHasTooltip',
      value: function initHasTooltip(content) {
        var type = this.props.type;


        if (!content || type !== 'compactSide') {
          return;
        }

        var hasTooltip = (0, _utils.getWidth)(content) > (0, _utils.getWidth)(content.parentElement);
        if (this.state.hasTooltip !== hasTooltip) {
          this.setState({ hasTooltip: hasTooltip });
        }
      }
    }, {
      key: 'withTooltipProperties',
      value: function withTooltipProperties(component, key) {
        var tooltipProperties = ['onMouseEnter', 'onMouseLeave', 'onFocus', 'onBlur'];
        return _react2.default.cloneElement(component, _defineProperty({}, key, tooltipProperties));
      }
    }, {
      key: 'withTooltip',
      value: function withTooltip(component) {
        var _props$item = this.props.item,
            id = _props$item.id,
            title = _props$item.title;

        return _react2.default.createElement(
          _Tooltip2.default,
          { key: id, content: title, moveBy: { y: VERTICAL_PADDING } },
          this.withTooltipProperties(component, 'dynamicProperties')
        );
      }
    }, {
      key: 'renderWrappedComponent',
      value: function renderWrappedComponent() {
        var _this2 = this;

        return _react2.default.createElement(WrappedComponent, _extends({}, this.props, { initHasTooltip: function initHasTooltip(el) {
            return _this2.initHasTooltip(el);
          } }));
      }
    }, {
      key: 'render',
      value: function render() {
        var component = this.renderWrappedComponent();
        return this.state.hasTooltip ? this.withTooltip(component) : component;
      }
    }]);

    return Wrapper;
  }(_react2.default.Component);

  Wrapper.propTypes = {
    type: TabPropTypes.type,
    item: TabPropTypes.item.isRequired
  };

  return Wrapper;
};

exports.default = withTooltip;
