// playable/dist/src/modules/ui/interaction-indicator/interaction-indicator.view.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var view_1 = tslib_1.__importDefault(require("../core/view"));
var templates_1 = require("./templates");
var htmlToElement_1 = tslib_1.__importDefault(require("../core/htmlToElement"));
var interaction_indicator_scss_1 = tslib_1.__importDefault(require("./interaction-indicator.scss"));
var SECONDS_COUNT = 5;
var InteractionIndicatorView = /** @class */ (function (_super) {
    tslib_1.__extends(InteractionIndicatorView, _super);
    function InteractionIndicatorView() {
        var _this = _super.call(this) || this;
        _this._$node = htmlToElement_1.default(templates_1.containerTemplate({
            styles: _this.styleNames,
        }));
        _this._playIcon = templates_1.playIconTemplate({
            styles: _this.styleNames,
        });
        _this._pauseIcon = templates_1.pauseIconTemplate({
            styles: _this.styleNames,
        });
        _this._forwardIcon = templates_1.forwardIconTemplate({
            texts: {
                SECONDS_COUNT: SECONDS_COUNT,
            },
            styles: _this.styleNames,
        });
        _this._rewindIcon = templates_1.rewindIconTemplate({
            texts: {
                SECONDS_COUNT: SECONDS_COUNT,
            },
            styles: _this.styleNames,
        });
        _this._increaseVolumeIcon = templates_1.increaseVolumeIconTemplate({
            styles: _this.styleNames,
        });
        _this._decreaseVolumeIcon = templates_1.decreaseVolumeIconTemplate({
            styles: _this.styleNames,
        });
        _this._muteIcon = templates_1.muteIconTemplate({
            styles: _this.styleNames,
        });
        return _this;
    }
    InteractionIndicatorView.prototype.activatePlayIcon = function () {
        this._$node.innerHTML = this._playIcon;
    };
    InteractionIndicatorView.prototype.activatePauseIcon = function () {
        this._$node.innerHTML = this._pauseIcon;
    };
    InteractionIndicatorView.prototype.activateForwardIcon = function () {
        this._$node.innerHTML = this._forwardIcon;
    };
    InteractionIndicatorView.prototype.activateRewindIcon = function () {
        this._$node.innerHTML = this._rewindIcon;
    };
    InteractionIndicatorView.prototype.activateIncreaseVolumeIcon = function () {
        this._$node.innerHTML = this._increaseVolumeIcon;
    };
    InteractionIndicatorView.prototype.activateDecreaseVolumeIcon = function () {
        this._$node.innerHTML = this._decreaseVolumeIcon;
    };
    InteractionIndicatorView.prototype.activateMuteVolumeIcon = function () {
        this._$node.innerHTML = this._muteIcon;
    };
    InteractionIndicatorView.prototype.deactivateIcon = function () {
        this._$node.innerHTML = '';
    };
    InteractionIndicatorView.prototype.hide = function () {
        this._$node.classList.add(this.styleNames.hidden);
    };
    InteractionIndicatorView.prototype.show = function () {
        this._$node.classList.remove(this.styleNames.hidden);
    };
    InteractionIndicatorView.prototype.getNode = function () {
        return this._$node;
    };
    InteractionIndicatorView.prototype.destroy = function () {
        if (this._$node.parentNode) {
            this._$node.parentNode.removeChild(this._$node);
        }
        this._$node = null;
    };
    return InteractionIndicatorView;
}(view_1.default));
InteractionIndicatorView.extendStyleNames(interaction_indicator_scss_1.default);
exports.default = InteractionIndicatorView;
