// wix-style-react/dist/stories/Table/Table.story.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _TableStory = require('./Table.story.scss');

var _TableStory2 = _interopRequireDefault(_TableStory);

var _storySettings = require('./storySettings');

var _CodeExample = require('wix-storybook-utils/CodeExample');

var _CodeExample2 = _interopRequireDefault(_CodeExample);

var _Table = require('../../src/Table/Table');

var _TableExample = require('./TableExample');

var _TableExample2 = require('!raw-loader!./TableExample');

var _TableExample3 = _interopRequireDefault(_TableExample2);

var _TablePageExample = require('./TablePageExample');

var _TablePageExample2 = require('!raw-loader!./TablePageExample');

var _TablePageExample3 = _interopRequireDefault(_TablePageExample2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var data = [{ firstName: 'Meghan', lastName: 'Bishop' }, { firstName: 'Sara', lastName: 'Porter' }, { firstName: 'Deborah', lastName: 'Rhodes' }, { firstName: 'Walter', lastName: 'Jenning' }];

var dataLong = [1, 2, 3, 4, 5].reduce(function (accum) {
  return accum.concat(data);
}, []);

var columnsOption1 = [{ title: 'First', width: '30%', render: function render(row) {
    return row.firstName;
  } }, { title: 'Last', width: '30%', render: function render(row) {
    return row.lastName;
  } }];

var columnsOption2 = [{ title: 'Row Num', render: function render(row, rowNum) {
    return rowNum;
  } }, { title: 'First', render: function render(row) {
    return row.firstName;
  } }, { title: 'Last', render: function render(row) {
    return row.lastName;
  } }, { title: 'Full', render: function render(row) {
    return row.firstName + row.lastName;
  } }];

exports.default = {
  category: _storySettings.storySettings.kind,
  storyName: _storySettings.storySettings.storyName,

  component: _Table.Table,
  componentPath: '../../src/Table/Table.js',

  componentProps: {
    dataHook: _storySettings.storySettings.dataHook,
    id: 'id',
    data: data,
    columns: columnsOption1,
    showSelection: true
  },
  exampleProps: {
    columns: [{ label: '2 columns example', value: columnsOption1 }, { label: '4 columns example', value: columnsOption2 }],
    data: [{ label: '4 rows', value: data }, { label: '40 rows', value: dataLong }]
  },
  codeExample: false,
  examples: _react2.default.createElement(
    'div',
    null,
    _react2.default.createElement(
      'div',
      { className: _TableStory2.default.examples },
      _react2.default.createElement(
        'div',
        { className: _TableStory2.default.example },
        _react2.default.createElement(
          _CodeExample2.default,
          { title: 'Typical (With Toolbar)', code: _TableExample3.default },
          _react2.default.createElement(_TableExample.TableExample, null)
        )
      ),
      _react2.default.createElement(
        'div',
        { className: _TableStory2.default.example },
        _react2.default.createElement(
          _CodeExample2.default,
          { title: 'Table in a Page (Fixed Header)', code: _TablePageExample3.default },
          _react2.default.createElement(_TablePageExample.TablePageExample, null)
        )
      )
    )
  )
};
