// wix-ui-core/dist/src/components/Badge/styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var theme_1 = require("./theme");
var defaultsDeep = require('lodash/defaultsDeep');
var contentSelector = '& [data-class="badge-content"]';
exports.styles = function (theme) {
    theme = defaultsDeep(theme, theme_1.core);
    return {
        badge: (_a = {
                width: theme.width,
                minWidth: theme.minWidth,
                height: theme.height,
                lineHeight: theme.lineHeight,
                padding: theme.padding,
                border: theme.border,
                borderRadius: theme.borderRadius,
                color: theme.color,
                background: theme.backgroundColor,
                borderColor: theme.borderColor,
                cursor: theme.cursor,
                display: theme.display,
                textAlign: theme.textAlign,
                verticalAlign: theme.verticalAlign,
                boxSizing: 'border-box'
            },
            _a[contentSelector] = {
                color: theme.color
            },
            _a)
    };
    var _a;
};
