// wix-style-react/dist/src/Tabs/Tabs.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _omit = require('omit');

var _omit2 = _interopRequireDefault(_omit);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _WixComponent2 = require('../BaseComponents/WixComponent');

var _WixComponent3 = _interopRequireDefault(_WixComponent2);

var _SideContent = require('./core/SideContent');

var _SideContent2 = _interopRequireDefault(_SideContent);

var _TabItems = require('./core/TabItems');

var _TabItems2 = _interopRequireDefault(_TabItems);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _tabPropTypes = require('./core/constants/tab-prop-types');

var TabPropTypes = _interopRequireWildcard(_tabPropTypes);

var _Tabs = require('./Tabs.scss');

var _Tabs2 = _interopRequireDefault(_Tabs);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Tabs = function (_WixComponent) {
  _inherits(Tabs, _WixComponent);

  function Tabs() {
    _classCallCheck(this, Tabs);

    return _possibleConstructorReturn(this, (Tabs.__proto__ || Object.getPrototypeOf(Tabs)).apply(this, arguments));
  }

  _createClass(Tabs, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          sideContent = _props.sideContent,
          hasDivider = _props.hasDivider;

      var tabItemsProps = (0, _omit2.default)(['sideContent'], this.props);
      var className = (0, _classnames2.default)(_Tabs2.default.container, _defineProperty({}, _Tabs2.default.hasDivider, hasDivider));

      return _react2.default.createElement(
        'div',
        { className: className },
        _react2.default.createElement(_TabItems2.default, tabItemsProps),
        _react2.default.createElement(_SideContent2.default, { content: sideContent })
      );
    }
  }]);

  return Tabs;
}(_WixComponent3.default);

Tabs.defaultProps = {
  hasDivider: true
};


Tabs.propTypes = {
  activeId: TabPropTypes.activeId,
  dataHook: _propTypes2.default.string,
  hasDivider: _propTypes2.default.bool,
  items: TabPropTypes.items.isRequired,
  minWidth: TabPropTypes.width,
  type: TabPropTypes.type,
  sideContent: TabPropTypes.sideContent,
  width: TabPropTypes.width,
  onClick: TabPropTypes.onClick
};

exports.default = Tabs;
