// wix-ui-core/dist/src/clients/GoogleMaps/IframesManager/IframesManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var GoogleRequestHandler_1 = require("../GoogleRequestHandler/GoogleRequestHandler");
var handlerNames = require("../handlersName");
var IframesManager = /** @class */ (function () {
    function IframesManager() {
        this._iframeMap = new Map();
    }
    IframesManager.getKey = function (apiKey, lang) {
        return apiKey + "-" + lang;
    };
    IframesManager.prototype.addIframe = function (apiKey, lang) {
        if (this.hasIframe(apiKey, lang)) {
            return this.getIframe(apiKey, lang);
        }
        var iframeKey = IframesManager.getKey(apiKey, lang);
        var iframe = document.createElement('iframe');
        iframe.setAttribute('data-lang', lang);
        iframe.setAttribute('data-api-key', apiKey);
        iframe.style.display = 'none';
        document.body.appendChild(iframe);
        this.populateIframe(iframe, apiKey, lang);
        this._iframeMap.set(iframeKey, iframe);
        return iframe.contentWindow;
    };
    IframesManager.prototype.getIframe = function (apiKey, lang) {
        var iframeKey = IframesManager.getKey(apiKey, lang);
        return this._iframeMap.get(iframeKey).contentWindow;
    };
    IframesManager.prototype.hasIframe = function (apiKey, lang) {
        var iframeKey = IframesManager.getKey(apiKey, lang);
        return this._iframeMap.has(iframeKey);
    };
    IframesManager.prototype.removeAllIframes = function () {
        this._iframeMap.forEach(function (iframe) { return iframe.remove(); });
        this._iframeMap.clear();
    };
    IframesManager.prototype.populateIframe = function (iframe, apiKey, lang) {
        var iframeBody = iframe.contentWindow.document.body;
        iframeBody.appendChild(this.createInitializationScript());
        iframeBody.appendChild(this.createGoogleMapsScript(apiKey, lang));
    };
    IframesManager.prototype.createInitializationScript = function () {
        var script = document.createElement('script');
        script.innerText = 'window.initAutoComplete = (' + GoogleRequestHandler_1.googleRequestHandler.toString() + ')(window, ' + JSON.stringify(handlerNames) + '); window.googleReady = () => window.initAutoComplete(window.google);';
        return script;
    };
    IframesManager.prototype.createGoogleMapsScript = function (apiKey, lang) {
        var script = document.createElement('script');
        script.src = "https://maps.googleapis.com/maps/api/js?libraries=places&key=" + apiKey + "&language=" + lang + "&callback=googleReady";
        return script;
    };
    return IframesManager;
}());
exports.IframesManager = IframesManager;
