// wix-style-react/dist/stories/Badge/Example.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Email = require('wix-style-react/new-icons/Email');

var _Email2 = _interopRequireDefault(_Email);

var _DropdownComposite = require('../../src/DropdownComposite');

var _DropdownComposite2 = _interopRequireDefault(_DropdownComposite);

var _Dropdown = require('../../src/Dropdown');

var _Dropdown2 = _interopRequireDefault(_Dropdown);

var _Label = require('../../src/Label');

var _Label2 = _interopRequireDefault(_Label);

var _Text = require('../../src/Text');

var _Text2 = _interopRequireDefault(_Text);

var _TextField = require('../../src/TextField');

var _TextField2 = _interopRequireDefault(_TextField);

var _Input = require('../../src/Input');

var _Input2 = _interopRequireDefault(_Input);

var _constants = require('wix-ui-backoffice/dist/src/components/Badge/constants');

var _Badge = require('./Badge.scss');

var _Badge2 = _interopRequireDefault(_Badge);

var _Badge3 = require('wix-style-react/Badge');

var _Badge4 = _interopRequireDefault(_Badge3);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var skinOptions = Object.keys(_constants.SKIN).map(function (value) {
  return { id: value, value: value };
});
var typeOptions = Object.keys(_constants.TYPE).map(function (value) {
  return { id: value, value: value };
});

var iconsOptions = [{ id: 'none', value: 'none' }, { id: 'Email', value: 'Email' }];

var ControlledBadgeExample = function (_React$Component) {
  _inherits(ControlledBadgeExample, _React$Component);

  function ControlledBadgeExample(props) {
    _classCallCheck(this, ControlledBadgeExample);

    var _this = _possibleConstructorReturn(this, (ControlledBadgeExample.__proto__ || Object.getPrototypeOf(ControlledBadgeExample)).call(this, props));

    _this.state = {
      skin: _constants.SKIN.general,
      type: _constants.TYPE.solid,
      children: 'I\'M A BADGE!',
      prefixIcon: 'none',
      suffixIcon: 'none'
    };
    return _this;
  }

  _createClass(ControlledBadgeExample, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      return _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
          'div',
          { style: { background: 'azure', width: '310px' } },
          _react2.default.createElement(
            _Text2.default,
            null,
            'import Badge from \'wix-style-react/Badge\';'
          )
        ),
        _react2.default.createElement(
          'div',
          { className: _Badge2.default.container },
          _react2.default.createElement(
            'div',
            { style: { width: '200px' } },
            _react2.default.createElement(
              'h1',
              null,
              ' Props '
            ),
            _react2.default.createElement(
              _TextField2.default,
              null,
              _react2.default.createElement(
                _Label2.default,
                null,
                'children'
              ),
              _react2.default.createElement(_Input2.default, { value: this.state.children, onChange: function onChange(e) {
                  return _this2.setState({ children: e.target.value });
                } })
            ),
            _react2.default.createElement(
              _DropdownComposite2.default,
              null,
              _react2.default.createElement(
                _Label2.default,
                null,
                'skin'
              ),
              _react2.default.createElement(_Dropdown2.default, { selectedId: this.state.skin, options: skinOptions, onSelect: function onSelect(option) {
                  return _this2.setState({ skin: option.value });
                } })
            ),
            _react2.default.createElement(
              _DropdownComposite2.default,
              null,
              _react2.default.createElement(
                _Label2.default,
                null,
                'type'
              ),
              _react2.default.createElement(_Dropdown2.default, { selectedId: this.state.type, options: typeOptions, onSelect: function onSelect(option) {
                  return _this2.setState({ type: option.value });
                } })
            ),
            _react2.default.createElement(
              _DropdownComposite2.default,
              null,
              _react2.default.createElement(
                _Label2.default,
                null,
                'prefixIcon'
              ),
              _react2.default.createElement(_Dropdown2.default, { selectedId: this.state.prefixIcon, options: iconsOptions, onSelect: function onSelect(option) {
                  return _this2.setState({ prefixIcon: option.value });
                } })
            ),
            _react2.default.createElement(
              _DropdownComposite2.default,
              null,
              _react2.default.createElement(
                _Label2.default,
                null,
                'suffixIcon'
              ),
              _react2.default.createElement(_Dropdown2.default, { selectedId: this.state.suffixIcon, options: iconsOptions, onSelect: function onSelect(option) {
                  return _this2.setState({ suffixIcon: option.value });
                } })
            )
          ),
          _react2.default.createElement(
            'div',
            null,
            _react2.default.createElement(
              'h1',
              null,
              ' Preview '
            ),
            _react2.default.createElement('br', null),
            _react2.default.createElement('br', null),
            _react2.default.createElement('br', null),
            _react2.default.createElement('br', null),
            _react2.default.createElement('br', null),
            _react2.default.createElement('br', null),
            _react2.default.createElement(
              _Badge4.default,
              {
                skin: this.state.skin,
                type: this.state.type,
                prefixIcon: this.state.prefixIcon === 'Email' ? _react2.default.createElement(_Email2.default, null) : null,
                suffixIcon: this.state.suffixIcon === 'Email' ? _react2.default.createElement(_Email2.default, null) : null,
                'data-hook': 'storybook-badge'
              },
              this.state.children
            )
          )
        )
      );
    }
  }]);

  return ControlledBadgeExample;
}(_react2.default.Component);

exports.default = function () {
  return _react2.default.createElement(ControlledBadgeExample, null);
};
