// wix-ui-core/dist/src/components/Label/Label.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var prop_types_1 = require("prop-types");
var Label_st_css_1 = require("./Label.st.css");
var defaultProps = {
    ellipsis: false
};
/**
 * Label
 */
exports.Label = function (props) {
    var id = props.id, children = props.children, ellipsis = props.ellipsis, disabled = props.disabled;
    return React.createElement("label", __assign({}, Label_st_css_1.default('root', { ellipsis: ellipsis, disabled: disabled }, props), { htmlFor: props.for, id: id }), children);
};
exports.Label.displayName = 'Label';
exports.Label.propTypes = {
    /** class name */
    className: prop_types_1.string,
    /** children */
    children: prop_types_1.node,
    /** For property */
    for: prop_types_1.string,
    /** ID of element */
    id: prop_types_1.string,
    /** should the text be ellipsed or not */
    ellipsis: prop_types_1.bool,
    /** Is the Label disabled */
    disabled: prop_types_1.bool
};
exports.Label.defaultProps = defaultProps;
