// wix-ui-core/dist/src/components/Label/Label.driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var test_utils_1 = require("stylable/test-utils");
var Label_st_css_1 = require("./Label.st.css");
exports.labelDriverFactory = function (_a) {
    var element = _a.element, eventTrigger = _a.eventTrigger;
    var stylableDOMUtil = new test_utils_1.StylableDOMUtil(Label_st_css_1.default, element);
    return {
        /** check if element exists */
        exists: function () { return !!element; },
        /** get the label's text */
        getLabelText: function () { return element.textContent; },
        /** get the id of the component */
        getId: function () { return element.getAttribute('id'); },
        /** get the "for" attribute of the component */
        getForAttribute: function () { return element.getAttribute('for'); },
        /** true if disabled */
        isDisabled: function () { return stylableDOMUtil.hasStyleState(element, 'disabled'); },
        /** click the label */
        click: function () { return eventTrigger.click(element); },
        /** send key down on the label */
        keyDown: function (key) { return eventTrigger.keyDown(element, { key: key }); },
        /** returns true if the label is in ellipsis state */
        hasEllipsis: function () { return stylableDOMUtil.hasStyleState(element, 'ellipsis'); }
    };
};
