// wix-ui-backoffice/dist/src/components/Thumbnail/Thumbnail.driver.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var Thumbnail_driver_1 = require("wix-ui-core/dist/src/components/Thumbnail/Thumbnail.driver");
var test_utils_1 = require("stylable/test-utils");
var Thumbnail_st_css_1 = require("./Thumbnail.st.css");
var Text_driver_1 = require("../Text/Text.driver");
exports.thumbnailDriverFactory = function (_a) {
    var element = _a.element, eventTrigger = _a.eventTrigger, wrapper = _a.wrapper;
    var coreThumbnailDriver = Thumbnail_driver_1.thumbnailDriverFactory({ element: element, eventTrigger: eventTrigger, wrapper: wrapper });
    var stylableDOMUtil = new test_utils_1.StylableDOMUtil(Thumbnail_st_css_1.default, element);
    var titleDriver = Text_driver_1.textDriverFactory({ element: stylableDOMUtil.select('.title'), wrapper: wrapper, eventTrigger: eventTrigger });
    var descriptionDriver = Text_driver_1.textDriverFactory({ element: stylableDOMUtil.select('.description'), wrapper: wrapper, eventTrigger: eventTrigger });
    var image = element.querySelector('[data-hook="image"]');
    return __assign({}, coreThumbnailDriver, { titleDriver: function () { return titleDriver; }, 
        /** returns the title of the thumbnail */
        getTitle: function () { return titleDriver.getText(); }, 
        /** returns true if the thumbnail has description */
        hasDescription: function () { return descriptionDriver.exists(); }, 
        /** returns the description of the thumbnail */
        getDescription: function () { return descriptionDriver.getText(); }, 
        /** returns the image of the thumbnail */
        getImage: function () { return image && image.childNodes[0]; }, 
        /** returns true if the thumbnail has an image */
        hasImage: function () { return !!image; } });
};
