// wix-style-react/dist/src/ColorPicker/color-picker-hsb.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _color = require('color');

var _color2 = _interopRequireDefault(_color);

var _clamp = require('lodash/clamp');

var _clamp2 = _interopRequireDefault(_clamp);

var _WixComponent2 = require('../BaseComponents/WixComponent');

var _WixComponent3 = _interopRequireDefault(_WixComponent2);

var _utils = require('./utils');

var _colorPickerHsb = require('./color-picker-hsb.scss');

var _colorPickerHsb2 = _interopRequireDefault(_colorPickerHsb);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ColorPickerHsb = function (_WixComponent) {
  _inherits(ColorPickerHsb, _WixComponent);

  function ColorPickerHsb() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ColorPickerHsb);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ColorPickerHsb.__proto__ || Object.getPrototypeOf(ColorPickerHsb)).call.apply(_ref, [this].concat(args))), _this), _this.onMarkerDragStart = function (e) {
      e.preventDefault();
      window.addEventListener('mousemove', _this.onMarkerDrag);
      window.addEventListener('mouseup', _this.onMarkerDragEnd);
      _this.gradientRect = (0, _utils.getBoundingRect)(_this.gradient);
      _this.setNewColorByMouseEvent(e);
    }, _this.onMarkerDrag = function (e) {
      _this.setNewColorByMouseEvent(e);
    }, _this.onMarkerDragEnd = function () {
      window.removeEventListener('mousemove', _this.onMarkerDrag);
      window.removeEventListener('mouseup', _this.onMarkerDragEnd);
    }, _this.getSVByMouseEvent = function (e) {
      var x = e.clientX - _this.gradientRect.left;
      var y = e.clientY - _this.gradientRect.top;

      var s = (0, _clamp2.default)(100 * x / _this.gradientRect.width, 0, 100);
      var v = (0, _clamp2.default)(100 - 100 * y / _this.gradientRect.height, 0, 100);

      return { s: s, v: v };
    }, _this.setNewColorByMouseEvent = function (e) {
      var _this$props = _this.props,
          onChange = _this$props.onChange,
          current = _this$props.current;

      var _this$getSVByMouseEve = _this.getSVByMouseEvent(e),
          s = _this$getSVByMouseEve.s,
          v = _this$getSVByMouseEve.v;

      onChange((0, _color2.default)({ h: current.hue(), s: s, v: v }));
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ColorPickerHsb, [{
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.onMarkerDragEnd();
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var current = this.props.current;

      var hue = current.saturationv(100).lightness(50);
      var style = {
        left: current.saturationv() + '%',
        top: 100 - current.value() + '%'
      };
      return _react2.default.createElement(
        'div',
        {
          className: _colorPickerHsb2.default.root,
          'data-hook': 'color-picker-hsb',
          ref: function ref(e) {
            return _this2.gradient = e;
          },
          onMouseDown: this.onMarkerDragStart
        },
        _react2.default.createElement('div', { className: _colorPickerHsb2.default.hue, style: { background: hue.hex() } }),
        _react2.default.createElement('div', { className: _colorPickerHsb2.default.saturation }),
        _react2.default.createElement('div', { className: _colorPickerHsb2.default.brightness }),
        _react2.default.createElement('div', { className: _colorPickerHsb2.default.handle, style: style })
      );
    }
  }]);

  return ColorPickerHsb;
}(_WixComponent3.default);

ColorPickerHsb.propTypes = {
  current: _propTypes.object.isRequired,
  onChange: _propTypes.func.isRequired
};
exports.default = ColorPickerHsb;
