// wix-style-react/dist/src/ColorPicker/color-picker-hue.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _color = require('color');

var _color2 = _interopRequireDefault(_color);

var _clamp = require('lodash/clamp');

var _clamp2 = _interopRequireDefault(_clamp);

var _WixComponent2 = require('../BaseComponents/WixComponent');

var _WixComponent3 = _interopRequireDefault(_WixComponent2);

var _utils = require('./utils');

var _colorPickerHue = require('./color-picker-hue.scss');

var _colorPickerHue2 = _interopRequireDefault(_colorPickerHue);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ColorPickerHue = function (_WixComponent) {
  _inherits(ColorPickerHue, _WixComponent);

  function ColorPickerHue() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ColorPickerHue);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ColorPickerHue.__proto__ || Object.getPrototypeOf(ColorPickerHue)).call.apply(_ref, [this].concat(args))), _this), _this.onMarkerDragStart = function (e) {
      e.preventDefault();
      window.addEventListener('mousemove', _this.onMarkerDrag);
      window.addEventListener('mouseup', _this.onMarkerDragEnd);
      _this.sliderRect = (0, _utils.getBoundingRect)(_this.slider);
      _this.setNewColorByMouseEvent(e);
    }, _this.onMarkerDrag = function (e) {
      _this.setNewColorByMouseEvent(e);
    }, _this.onMarkerDragEnd = function () {
      window.removeEventListener('mousemove', _this.onMarkerDrag);
      window.removeEventListener('mouseup', _this.onMarkerDragEnd);
    }, _this.getHueByMouseEvent = function (e) {
      var x = e.clientX - _this.sliderRect.left;
      return (0, _clamp2.default)(360 * x / _this.sliderRect.width, 0, 359);
    }, _this.setNewColorByMouseEvent = function (e) {
      var _this$props = _this.props,
          onChange = _this$props.onChange,
          current = _this$props.current;

      var h = _this.getHueByMouseEvent(e);
      onChange((0, _color2.default)({ h: h, s: current.saturationv(), v: current.value() }));
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ColorPickerHue, [{
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.onMarkerDragEnd();
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      // HUE is an intenger value from 0 to 360.
      var percentage = this.props.current.hue() / 360 * 100;
      return _react2.default.createElement(
        'div',
        { className: _colorPickerHue2.default.root, ref: function ref(e) {
            return _this2.slider = e;
          }, onMouseDown: this.onMarkerDragStart },
        _react2.default.createElement('div', { className: _colorPickerHue2.default.handle, style: { left: percentage + '%' } })
      );
    }
  }]);

  return ColorPickerHue;
}(_WixComponent3.default);

ColorPickerHue.propTypes = {
  current: _propTypes.object.isRequired,
  onChange: _propTypes.func.isRequired
};
exports.default = ColorPickerHue;
