// wix-style-react/dist/src/ColorPicker/color-picker-converter.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _WixComponent2 = require('../BaseComponents/WixComponent');

var _WixComponent3 = _interopRequireDefault(_WixComponent2);

var _Tabs = require('../Tabs');

var _Tabs2 = _interopRequireDefault(_Tabs);

var _colorPickerConverterHex = require('./color-picker-converter-hex');

var _colorPickerConverterHex2 = _interopRequireDefault(_colorPickerConverterHex);

var _colorPickerConverterRgb = require('./color-picker-converter-rgb');

var _colorPickerConverterRgb2 = _interopRequireDefault(_colorPickerConverterRgb);

var _colorPickerConverterHsb = require('./color-picker-converter-hsb');

var _colorPickerConverterHsb2 = _interopRequireDefault(_colorPickerConverterHsb);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var HEX = 'HEX';
var RGB = 'RGB';
var HSB = 'HSB';

var tabs = [{ id: HEX, title: HEX }, { id: RGB, title: RGB }, { id: HSB, title: HSB }];

var ColorPickerConverter = function (_WixComponent) {
  _inherits(ColorPickerConverter, _WixComponent);

  function ColorPickerConverter(props) {
    _classCallCheck(this, ColorPickerConverter);

    var _this = _possibleConstructorReturn(this, (ColorPickerConverter.__proto__ || Object.getPrototypeOf(ColorPickerConverter)).call(this, props));

    _this.state = {
      activeTab: HEX
    };

    _this.changeTab = _this.changeTab.bind(_this);
    return _this;
  }

  _createClass(ColorPickerConverter, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          current = _props.current,
          showConverter = _props.showConverter,
          showInput = _props.showInput;


      if (!showConverter && !showInput) {
        return null;
      }

      if (!showConverter) {
        return _react2.default.createElement(_colorPickerConverterHex2.default, { current: current, onChange: this.props.onChange });
      }

      var activeTab = this.state.activeTab;


      return _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(_Tabs2.default, {
          minWidth: 0,
          items: tabs,
          activeId: activeTab,
          type: 'uniformFull',
          onClick: this.changeTab
        }),
        activeTab === HEX && _react2.default.createElement(_colorPickerConverterHex2.default, { current: current, onChange: this.props.onChange }),
        activeTab === RGB && _react2.default.createElement(_colorPickerConverterRgb2.default, { current: current, onChange: this.props.onChange }),
        activeTab === HSB && _react2.default.createElement(_colorPickerConverterHsb2.default, { current: current, onChange: this.props.onChange })
      );
    }
  }, {
    key: 'changeTab',
    value: function changeTab(_ref) {
      var id = _ref.id;

      this.setState({ activeTab: id });
    }
  }]);

  return ColorPickerConverter;
}(_WixComponent3.default);

ColorPickerConverter.propTypes = {
  current: _propTypes.object.isRequired,
  showConverter: _propTypes.bool.isRequired,
  showInput: _propTypes.bool.isRequired,
  onChange: _propTypes.func.isRequired
};
exports.default = ColorPickerConverter;
