// wix-style-react/dist/src/ColorPicker/color-picker-converter-rgb.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _color = require('color');

var _color2 = _interopRequireDefault(_color);

var _propTypes = require('prop-types');

var _mapValues = require('lodash/mapValues');

var _mapValues2 = _interopRequireDefault(_mapValues);

var _WixComponent2 = require('../BaseComponents/WixComponent');

var _WixComponent3 = _interopRequireDefault(_WixComponent2);

var _Input = require('../Input');

var _Input2 = _interopRequireDefault(_Input);

var _colorPickerConverter = require('./color-picker-converter.scss');

var _colorPickerConverter2 = _interopRequireDefault(_colorPickerConverter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ColorPickerConverterRgb = function (_WixComponent) {
  _inherits(ColorPickerConverterRgb, _WixComponent);

  function ColorPickerConverterRgb(props) {
    _classCallCheck(this, ColorPickerConverterRgb);

    var _this = _possibleConstructorReturn(this, (ColorPickerConverterRgb.__proto__ || Object.getPrototypeOf(ColorPickerConverterRgb)).call(this, props));

    _this.state = (0, _mapValues2.default)(props.current.rgb().object(), Math.round);
    return _this;
  }

  _createClass(ColorPickerConverterRgb, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      return _react2.default.createElement(
        'div',
        { className: _colorPickerConverter2.default.root },
        _react2.default.createElement(
          'div',
          { className: _colorPickerConverter2.default.distribute },
          _react2.default.createElement(_Input2.default, { size: 'small', value: this.state.r, onChange: function onChange(e) {
              return _this2.change('r', e);
            } }),
          _react2.default.createElement(_Input2.default, { size: 'small', value: this.state.g, onChange: function onChange(e) {
              return _this2.change('g', e);
            } }),
          _react2.default.createElement(_Input2.default, { size: 'small', value: this.state.b, onChange: function onChange(e) {
              return _this2.change('b', e);
            } })
        )
      );
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(props) {
      this.setState((0, _mapValues2.default)(props.current.rgb().object(), Math.round));
    }
  }, {
    key: 'change',
    value: function change(part, _ref) {
      var _this3 = this;

      var value = _ref.target.value;

      this.setState(_defineProperty({}, part, value), function () {
        var color = safeColor(_this3.state);
        if (color) {
          _this3.props.onChange(color);
        }
      });
    }
  }]);

  return ColorPickerConverterRgb;
}(_WixComponent3.default);

ColorPickerConverterRgb.propTypes = {
  current: _propTypes.object.isRequired,
  onChange: _propTypes.func.isRequired
};
exports.default = ColorPickerConverterRgb;


function safeColor(input) {
  try {
    return (0, _color2.default)(input);
  } catch (error) {
    return null;
  }
}
