// wix-style-react/dist/src/ColorPicker/color-picker-history.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _colorPickerHistory = require('./color-picker-history.scss');

var _colorPickerHistory2 = _interopRequireDefault(_colorPickerHistory);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ColorPickerHistory = function ColorPickerHistory(_ref) {
  var show = _ref.show,
      current = _ref.current,
      previous = _ref.previous,
      _onClick = _ref.onClick;

  if (show) {
    return _react2.default.createElement(
      'div',
      { className: _colorPickerHistory2.default.root, 'data-hook': 'color-picker-history' },
      _react2.default.createElement('div', {
        'data-hook': 'color-picker-history-previous',
        style: { background: previous.hex() },
        onClick: function onClick() {
          return _onClick(previous);
        }
      }),
      _react2.default.createElement('div', {
        'data-hook': 'color-picker-history-current',
        style: { background: current.hex() }
      })
    );
  }
  return null;
};

ColorPickerHistory.propTypes = {
  show: _propTypes.bool.isRequired,
  previous: _propTypes.object.isRequired,
  current: _propTypes.object.isRequired,
  onClick: _propTypes.func.isRequired
};

exports.default = ColorPickerHistory;
