// wix-style-react/dist/stories/MessageBox/CustomModalExamples/FullScreenModal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _MessageBoxFunctionalLayout = require('wix-style-react/MessageBox/MessageBoxFunctionalLayout');

var _MessageBoxFunctionalLayout2 = _interopRequireDefault(_MessageBoxFunctionalLayout);

var _Backoffice = require('wix-style-react/Backoffice');

var _Modal = require('wix-style-react/Modal');

var _Modal2 = _interopRequireDefault(_Modal);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var FullScreenModal = function (_Component) {
  _inherits(FullScreenModal, _Component);

  function FullScreenModal() {
    _classCallCheck(this, FullScreenModal);

    var _this = _possibleConstructorReturn(this, (FullScreenModal.__proto__ || Object.getPrototypeOf(FullScreenModal)).call(this));

    _this.state = {
      isOpenFullScreenModal: false
    };
    return _this;
  }

  _createClass(FullScreenModal, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var setState = function setState(state) {
        return function () {
          return _this2.setState(state);
        };
      };
      var closeFullScreenModal = setState({ isOpenFullScreenModal: false });
      var openFullScreenModal = setState({ isOpenFullScreenModal: true });
      return _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
          _Backoffice.Button,
          {
            dataHook: 'open-full-screen-modal-button',
            onClick: openFullScreenModal
          },
          'Open Full Screen Modal'
        ),
        _react2.default.createElement(
          _Modal2.default,
          {
            isOpen: this.state.isOpenFullScreenModal,
            onRequestClose: closeFullScreenModal,
            contentLabel: 'Full screen modal example'
          },
          _react2.default.createElement(
            _MessageBoxFunctionalLayout2.default,
            {
              cancelText: 'Cancel',
              confirmText: 'OK',
              dataHook: 'fullscreen-modal',
              fullscreen: true,
              onCancel: closeFullScreenModal,
              onOk: closeFullScreenModal,
              theme: 'blue',
              title: 'Full screen modal'
            },
            'I\'m full screen modal!'
          )
        )
      );
    }
  }]);

  return FullScreenModal;
}(_react.Component);

exports.default = function () {
  return _react2.default.createElement(FullScreenModal, null);
};
