// wix-style-react/dist/src/Card/Header/Header.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Header = require('./Header.scss');

var _Header2 = _interopRequireDefault(_Header);

var _WixComponent2 = require('../../BaseComponents/WixComponent');

var _WixComponent3 = _interopRequireDefault(_WixComponent2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Header = function (_WixComponent) {
  _inherits(Header, _WixComponent);

  function Header() {
    _classCallCheck(this, Header);

    return _possibleConstructorReturn(this, (Header.__proto__ || Object.getPrototypeOf(Header)).apply(this, arguments));
  }

  _createClass(Header, [{
    key: 'render',
    value: function render() {
      var _classNames;

      var _props = this.props,
          title = _props.title,
          subtitle = _props.subtitle,
          withoutDivider = _props.withoutDivider,
          suffix = _props.suffix;


      var headerClasses = (0, _classnames2.default)((_classNames = {}, _defineProperty(_classNames, _Header2.default.header, true), _defineProperty(_classNames, _Header2.default.withDivider, !withoutDivider), _classNames));

      var titleElement = _react2.default.createElement(
        'div',
        { 'data-hook': 'title', className: _Header2.default.title },
        _react2.default.createElement(
          'span',
          null,
          title
        )
      );

      var subtitleElement = subtitle ? _react2.default.createElement(
        'div',
        { 'data-hook': 'subtitle', className: _Header2.default.subtitle },
        subtitle
      ) : null;

      var suffixElement = suffix ? _react2.default.createElement(
        'div',
        { 'data-hook': 'suffix' },
        suffix
      ) : null;

      return _react2.default.createElement(
        'div',
        { className: headerClasses },
        _react2.default.createElement(
          'div',
          { className: _Header2.default.container },
          titleElement,
          subtitleElement
        ),
        suffixElement
      );
    }
  }]);

  return Header;
}(_WixComponent3.default);

Header.propTypes = {
  title: _propTypes.node.isRequired,
  subtitle: _propTypes.node,
  withoutDivider: _propTypes.bool,
  suffix: _propTypes.node
};
Header.defaultProps = {
  subtitle: null,
  suffix: null,
  withoutDivider: false
};
exports.default = Header;
