// wix-style-react/dist/src/Backoffice/TextLink/TextLink.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _WixComponent2 = require('../../BaseComponents/WixComponent');

var _WixComponent3 = _interopRequireDefault(_WixComponent2);

var _TextLink = require('../../BaseComponents/TextLink');

var _TextLink2 = _interopRequireDefault(_TextLink);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var TextLink = function (_WixComponent) {
  _inherits(TextLink, _WixComponent);

  function TextLink() {
    _classCallCheck(this, TextLink);

    return _possibleConstructorReturn(this, (TextLink.__proto__ || Object.getPrototypeOf(TextLink)).apply(this, arguments));
  }

  _createClass(TextLink, [{
    key: 'render',
    value: function render() {
      return _react2.default.createElement(_TextLink2.default, this.props);
    }
  }]);

  return TextLink;
}(_WixComponent3.default);

TextLink.propTypes = _extends({}, _TextLink2.default.propTypes, {
  link: _propTypes2.default.string,
  disabled: _propTypes2.default.bool,
  download: _propTypes2.default.bool,
  rel: _propTypes2.default.string,
  target: _propTypes2.default.string,
  ariaLabel: _propTypes2.default.string
});
TextLink.defaultProps = _extends({}, _TextLink2.default.defaultProps, {
  disabled: false,
  download: false,
  rel: null,
  target: null
});
exports.default = TextLink;
