// wix-ui-backoffice/dist/src/components/FloatingHelper/FloatingHelperContent/FloatingHelperContent.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var prop_types_1 = require("prop-types");
var FloatingHelperContent_st_css_1 = require("./FloatingHelperContent.st.css");
var Text_1 = require("../../../components/Text");
var DataHooks_1 = require("./DataHooks");
var Button_1 = require("../../Button");
var constants_1 = require("./constants");
var themeToButtonProps = (_a = {},
    _a[constants_1.ActionButtonTheme.white] = { skin: Button_1.ButtonSkin.white, priority: Button_1.ButtonPriority.secondary },
    _a[constants_1.ActionButtonTheme.premium] = { skin: Button_1.ButtonSkin.premium, priority: Button_1.ButtonPriority.primary },
    _a);
exports.FloatingHelperContent = function (props) {
    var title = props.title, body = props.body, actionText = props.actionText, onActionClick = props.onActionClick, actionTheme = props.actionTheme, image = props.image;
    return (React.createElement("div", __assign({}, FloatingHelperContent_st_css_1.default('root', { hasBody: !!props.body }, props)),
        React.createElement("div", null,
            title && (React.createElement("div", { className: FloatingHelperContent_st_css_1.default.title },
                React.createElement(Text_1.Text, { "data-hook": DataHooks_1.DataHooks.title, bold: true, light: true }, title))),
            body && (React.createElement("div", { className: FloatingHelperContent_st_css_1.default.body },
                React.createElement(Text_1.Text, { "data-hook": DataHooks_1.DataHooks.body, light: true }, body))),
            actionText && onActionClick &&
                actionText.length > 0 && (React.createElement(Button_1.Button, { className: FloatingHelperContent_st_css_1.default.action, "data-hook": DataHooks_1.DataHooks.actionButton, skin: themeToButtonProps[actionTheme].skin, priority: themeToButtonProps[actionTheme].priority, size: Button_1.ButtonSize.small, onClick: onActionClick }, actionText))),
        image && React.createElement("div", { "data-hook": DataHooks_1.DataHooks.image, className: FloatingHelperContent_st_css_1.default.image }, image)));
};
exports.FloatingHelperContent.propTypes = {
    title: prop_types_1.string,
    body: prop_types_1.string.isRequired,
    actionText: prop_types_1.string,
    onActionClick: prop_types_1.func,
    image: prop_types_1.node
};
exports.FloatingHelperContent.defaultProps = {
    actionTheme: constants_1.ActionButtonTheme.white
};
var _a;
