// wix-style-react/dist/src/PageHeader/PageHeader.driver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var titleElement = function titleElement(element) {
  return element.querySelector('[data-hook="page-header-title"] h1');
};
var subtitleElement = function subtitleElement(element) {
  return element.querySelector('[data-hook="page-header-subtitle"] span');
};
var breadcrumbsElement = function breadcrumbsElement(element) {
  return element.querySelector('[data-hook="page-header-breadcrumbs"]');
};
var actionBarElement = function actionBarElement(element) {
  return element.querySelector('[data-hook="page-header-actionbar"]');
};
var backButtonElement = function backButtonElement(element) {
  return element.querySelector('[data-hook="page-header-backbutton"]');
};

exports.default = function (_ref) {
  var element = _ref.element;
  return {
    titleText: function titleText() {
      return titleElement(element).textContent;
    },
    isTitleExists: function isTitleExists() {
      return !!titleElement(element);
    },
    subtitleText: function subtitleText() {
      return subtitleElement(element).textContent;
    },
    isSubtitleExists: function isSubtitleExists() {
      return !!subtitleElement(element);
    },
    isBreadcrumbsExists: function isBreadcrumbsExists() {
      return !!breadcrumbsElement(element);
    },
    breadcrumbsText: function breadcrumbsText() {
      return breadcrumbsElement(element).textContent;
    },
    isActionBarExists: function isActionBarExists() {
      return !!actionBarElement(element);
    },
    isBackButtonExists: function isBackButtonExists() {
      return !!backButtonElement(element);
    }
  };
};
