// wix-style-react/dist/stories/Dropdown/ExampleWithCustomValue.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Dropdown = require('wix-style-react/Dropdown');

var _Dropdown2 = _interopRequireDefault(_Dropdown);

var _Checkbox = require('wix-style-react/Checkbox');

var _Checkbox2 = _interopRequireDefault(_Checkbox);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var style = {
  display: 'inline-block',
  padding: '0 5px 0',
  width: '200px',
  lineHeight: '22px',
  marginBottom: '350px'
};

var CustomValuesInDropdown = function (_React$Component) {
  _inherits(CustomValuesInDropdown, _React$Component);

  function CustomValuesInDropdown(props) {
    _classCallCheck(this, CustomValuesInDropdown);

    var _this = _possibleConstructorReturn(this, (CustomValuesInDropdown.__proto__ || Object.getPrototypeOf(CustomValuesInDropdown)).call(this, props));

    _this.state = { checkboxChecked: false };
    return _this;
  }

  _createClass(CustomValuesInDropdown, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var valueParser = function valueParser(option) {
        return typeof option.value === 'string' ? option.value : option.value.props.children[0].props.children;
      };

      var onSelect = function onSelect(option) {
        return console.log('Selected ', valueParser(option));
      };

      var onChange = function onChange() {
        return _this2.setState({ checkboxChecked: !_this2.state.checkboxChecked });
      };

      var customValue = _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
          'span',
          null,
          'Custom Value'
        ),
        _react2.default.createElement(
          'span',
          { style: { marginLeft: '5px' }, onClick: function onClick(e) {
              return e.stopPropagation();
            } },
          _react2.default.createElement(_Checkbox2.default, { onChange: onChange, checked: this.state.checkboxChecked })
        )
      );

      var options = [{ id: 1, value: 'Option 1' }, { id: 2, value: 'Option 2' }, { id: 3, value: 'Option 3' }, { id: 4, value: 'Option 4', disabled: true }, { id: 5, value: 'Option 5' }, { id: 6, value: customValue }, { id: 7, value: customValue }];

      return _react2.default.createElement(_Dropdown2.default, {
        options: options,
        selectedId: 6,
        placeholder: 'Choose an option',
        valueParser: valueParser,
        onSelect: onSelect
      });
    }
  }]);

  return CustomValuesInDropdown;
}(_react2.default.Component);

exports.default = function () {
  return _react2.default.createElement(
    'div',
    { className: 'ltr', style: style },
    _react2.default.createElement(CustomValuesInDropdown, null)
  );
};
