// wix-style-react/dist/stories/Dropdown/ExampleControlled.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Dropdown = require('wix-style-react/Dropdown');

var _Dropdown2 = _interopRequireDefault(_Dropdown);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var style = {
  display: 'inline-block',
  padding: '0 5px 0',
  width: '200px',
  lineHeight: '22px',
  marginBottom: '160px'
};

var options = [{ id: 1, value: 'Option 1' }, { id: 2, value: 'Option 2' }, { id: 3, value: 'Option 3' }, { id: 4, value: 'Option 4', disabled: true }, { id: 5, value: 'Option 5' }];

var ControlledDropdown = function (_React$Component) {
  _inherits(ControlledDropdown, _React$Component);

  function ControlledDropdown(props) {
    _classCallCheck(this, ControlledDropdown);

    var _this = _possibleConstructorReturn(this, (ControlledDropdown.__proto__ || Object.getPrototypeOf(ControlledDropdown)).call(this, props));

    _this.onSelect = _this.onSelect.bind(_this);
    return _this;
  }

  _createClass(ControlledDropdown, [{
    key: 'onSelect',
    value: function onSelect(option) {
      console.log('Option ' + JSON.stringify(option) + ' selected');
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(_Dropdown2.default, {
        dataHook: 'story-dropdown-controlled',
        options: options,
        onSelect: this.onSelect,
        placeholder: 'Choose an option'
      });
    }
  }]);

  return ControlledDropdown;
}(_react2.default.Component);

exports.default = function () {
  return _react2.default.createElement(
    'div',
    { className: 'ltr', style: style },
    _react2.default.createElement(ControlledDropdown, null)
  );
};
