// wix-style-react/dist/src/SectionHelper/SectionHelper.driver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _vanilla = require('wix-ui-test-utils/vanilla');

var _Button = require('../Backoffice/Button/Button.driver');

var _Button2 = _interopRequireDefault(_Button);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var buttonTestkitFactory = (0, _vanilla.testkitFactoryCreator)(_Button2.default);

var sectionHelperDriverFactory = function sectionHelperDriverFactory(_ref) {
  var element = _ref.element;

  var classExists = function classExists(className) {
    return element.classList.contains(className);
  };
  var actionButtonDriver = function actionButtonDriver() {
    return buttonTestkitFactory({
      wrapper: element,
      dataHook: 'sectionhelper-action-btn'
    });
  };

  var closeButtonDriver = function closeButtonDriver() {
    return buttonTestkitFactory({
      wrapper: element,
      dataHook: 'sectionhelper-close-btn'
    });
  };

  return {
    exists: function exists() {
      return !!element;
    },
    titleText: function titleText() {
      return element.querySelector('[data-hook="sectionhelper-title"]').textContent;
    },
    actionText: function actionText() {
      return actionButtonDriver().getButtonTextContent();
    },
    clickAction: function clickAction() {
      return actionButtonDriver().click();
    },
    clickClose: function clickClose() {
      return closeButtonDriver().click();
    },
    textContent: function textContent() {
      return element.textContent;
    },
    isWarning: function isWarning() {
      return classExists('warning');
    },
    isStandard: function isStandard() {
      return classExists('standard');
    },
    isDanger: function isDanger() {
      return classExists('danger');
    },
    isSuccess: function isSuccess() {
      return classExists('success');
    },
    isPremium: function isPremium() {
      return classExists('premium');
    }
  };
};

exports.default = sectionHelperDriverFactory;
