// wix-style-react/dist/src/PopoverMenuItem/PopoverMenuItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _WixComponent2 = require('../BaseComponents/WixComponent');

var _WixComponent3 = _interopRequireDefault(_WixComponent2);

var _Text = require('../Text');

var _Text2 = _interopRequireDefault(_Text);

var _PopoverMenuItem = require('./PopoverMenuItem.scss');

var _PopoverMenuItem2 = _interopRequireDefault(_PopoverMenuItem);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var PopoverMenuItem = function (_WixComponent) {
  _inherits(PopoverMenuItem, _WixComponent);

  function PopoverMenuItem() {
    _classCallCheck(this, PopoverMenuItem);

    return _possibleConstructorReturn(this, (PopoverMenuItem.__proto__ || Object.getPrototypeOf(PopoverMenuItem)).apply(this, arguments));
  }

  _createClass(PopoverMenuItem, [{
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        'li',
        { className: (0, _classnames3.default)(_PopoverMenuItem2.default.root, _defineProperty({}, _PopoverMenuItem2.default.large, this.props.size === 'large')) },
        _react2.default.createElement(
          'button',
          {
            type: 'button',
            className: _PopoverMenuItem2.default.button,
            onClick: this.props.onClick
          },
          this.props.icon && _react2.default.createElement(
            'div',
            { className: _PopoverMenuItem2.default.icon },
            this.props.icon
          ),
          _react2.default.createElement(
            'div',
            { className: _PopoverMenuItem2.default.text },
            _react2.default.createElement(
              _Text2.default,
              { dataHook: 'menu-item-text', size: this.props.size === 'normal' ? 'small' : 'medium' },
              this.props.text
            )
          )
        )
      );
    }
  }]);

  return PopoverMenuItem;
}(_WixComponent3.default);

PopoverMenuItem.propTypes = {
  icon: _propTypes2.default.node,
  text: _propTypes2.default.string,
  onClick: _propTypes2.default.func,
  size: _propTypes2.default.oneOf(['normal', 'large'])
};
PopoverMenuItem.defaultProps = {
  size: 'normal'
};
exports.default = PopoverMenuItem;
