// wix-style-react/dist/stories/DropdownLayout/ExampleControlledWithButtons.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _DropdownLayout = require('wix-style-react/DropdownLayout');

var _DropdownLayout2 = _interopRequireDefault(_DropdownLayout);

var _ChevronDown = require('wix-style-react/new-icons/ChevronDown');

var _ChevronDown2 = _interopRequireDefault(_ChevronDown);

var _Backoffice = require('wix-style-react/Backoffice');

var _ExampleControlledWithButtons = require('./ExampleControlledWithButtons.scss');

var _ExampleControlledWithButtons2 = _interopRequireDefault(_ExampleControlledWithButtons);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var longOptions = [{ id: 1, value: 'Option 1' }, { id: 2, value: 'Option 2' }, { id: 3, value: '-' }, { id: 4, value: 'Option 3' }, { id: 5, value: 'Disabled', disabled: true }, { id: 6, value: 'Option 4' }, { id: 7, value: 'Option 1' }, { id: 8, value: 'Option 2' }, { id: 9, value: '-' }, { id: 10, value: 'Option 3' }, { id: 11, value: 'Disabled', disabled: true }, { id: 12, value: 'Option 4' }, { id: 13, value: 'Option 1' }, { id: 14, value: 'Option 2' }, { id: 15, value: '-' }, { id: 16, value: 'Option 3' }, { id: 17, value: 'Disabled', disabled: true }, { id: 18, value: 'Option 4' }, { id: 19, value: 'Option 1' }, { id: 20, value: 'Option 2' }, { id: 21, value: '-' }, { id: 22, value: 'Option 3' }, { id: 23, value: 'Disabled', disabled: true }, { id: 24, value: 'Option 4' }];

var options = [{ id: 1, value: 'Option 1' }, { id: 2, value: 'Option 2' }, { id: 0, value: '-' }, { id: 3, value: 'Option 3' }, { id: 'disabled', value: 'Disabled', disabled: true }, { id: 4, value: 'Option 4' }];

var ControlledExample = function (_Component) {
  _inherits(ControlledExample, _Component);

  function ControlledExample() {
    _classCallCheck(this, ControlledExample);

    var _this = _possibleConstructorReturn(this, (ControlledExample.__proto__ || Object.getPrototypeOf(ControlledExample)).call(this));

    _this.state = { visible: -1 };
    return _this;
  }

  _createClass(ControlledExample, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var onClose = function onClose() {
        return _this2.setState({ visible: -1 });
      };
      var onClick = function onClick(button) {
        return function () {
          return _this2.setState({ visible: button });
        };
      };

      return _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
          'div',
          { className: _ExampleControlledWithButtons2.default.buttons_holder },
          _react2.default.createElement(
            'div',
            { className: _ExampleControlledWithButtons2.default.button },
            _react2.default.createElement(
              _Backoffice.Button,
              {
                withNewIcons: true,
                height: 'medium',
                suffixIcon: _react2.default.createElement(_ChevronDown2.default, null),
                onClick: onClick(1),
                theme: 'fullblue'
              },
              'Click On Me'
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _ExampleControlledWithButtons2.default.button },
            _react2.default.createElement(
              _Backoffice.Button,
              {
                withNewIcons: true,
                height: 'medium',
                suffixIcon: _react2.default.createElement(_ChevronDown2.default, null),
                onClick: onClick(2),
                theme: 'fullblue'
              },
              'Click On Me Too'
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _ExampleControlledWithButtons2.default.button },
            _react2.default.createElement(
              _Backoffice.Button,
              {
                withNewIcons: true,
                height: 'medium',
                suffixIcon: _react2.default.createElement(_ChevronDown2.default, null),
                onClick: onClick(3),
                theme: 'fullblue'
              },
              'Click On Me Now'
            )
          )
        ),
        _react2.default.createElement(
          'div',
          { className: 'ltr ' + _ExampleControlledWithButtons2.default.dropdown_a },
          _react2.default.createElement(_DropdownLayout2.default, { visible: this.state.visible === 1, options: options, maxHeightPixels: 100, onClose: onClose, onClickOutside: onClose })
        ),
        _react2.default.createElement(
          'div',
          { className: 'ltr ' + _ExampleControlledWithButtons2.default.dropdown_b },
          _react2.default.createElement(_DropdownLayout2.default, { visible: this.state.visible === 2, options: longOptions, maxHeightPixels: 1000, onClose: onClose, onClickOutside: onClose })
        ),
        _react2.default.createElement(
          'div',
          { className: 'ltr ' + _ExampleControlledWithButtons2.default.dropdown_c },
          _react2.default.createElement(_DropdownLayout2.default, { visible: this.state.visible === 3, options: options, onClose: onClose, onClickOutside: onClose })
        )
      );
    }
  }]);

  return ControlledExample;
}(_react.Component);

exports.default = function () {
  return _react2.default.createElement(ControlledExample, null);
};
