// wix-style-react/dist/stories/LanguagePicker/Example.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _LanguagePicker = require('wix-style-react/LanguagePicker');

var _LanguagePicker2 = _interopRequireDefault(_LanguagePicker);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var languageLocaleMap = {
  en: 'English',
  cs: 'Čeština',
  da: 'Dansk',
  de: 'Deutsch',
  fr: 'Français',
  es: 'Español',
  it: 'Italiano',
  no: 'Norsk',
  pl: 'Polski',
  pt: 'Português',
  ru: 'Русский',
  sv: 'Svenska',
  ja: '日本語',
  ko: '한국어',
  tr: 'Türkçe',
  nl: 'Nederlands',
  hi: 'हिन्दी',
  zh: 'Chinese',
  th: 'Thai'
};

var mockDictionary = {
  en: 'Hello',
  cs: 'Ahoj',
  da: 'Hej',
  de: 'Hallo',
  fr: 'Bonjour',
  es: 'Hola',
  it: 'Ciao',
  no: 'Hallo',
  pl: 'cześć',
  pt: 'Olá',
  ru: 'Здравствуйте',
  sv: 'Hallå',
  ja: 'こんにちは',
  ko: '안녕하세요',
  tr: 'Merhaba',
  nl: 'Hallo',
  hi: 'नमस्ते',
  zh: '你好',
  th: 'สวัสดี'
};

var Example = function (_React$Component) {
  _inherits(Example, _React$Component);

  function Example(props) {
    _classCallCheck(this, Example);

    var _this = _possibleConstructorReturn(this, (Example.__proto__ || Object.getPrototypeOf(Example)).call(this, props));

    _this.state = { languageKey: 'en' };
    return _this;
  }

  _createClass(Example, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      return _react2.default.createElement(
        'div',
        { style: { display: 'flex', alignItems: 'center', marginLeft: '30px' }, className: 'ltr' },
        _react2.default.createElement(
          _LanguagePicker2.default,
          { dataHook: 'story-languagePicker', selectedId: this.state.languageKey, onSelect: function onSelect(_ref) {
              var id = _ref.id;
              return _this2.setState({ languageKey: id });
            } },
          Object.keys(languageLocaleMap).map(function (languageKey) {
            return _react2.default.createElement(
              _LanguagePicker2.default.Option,
              { key: languageKey, languageKey: languageKey },
              languageLocaleMap[languageKey]
            );
          })
        ),
        _react2.default.createElement(
          'div',
          { id: 'languagePicker-output', style: { paddingBottom: '22px', marginLeft: '40px' } },
          _react2.default.createElement(
            'h1',
            null,
            mockDictionary[this.state.languageKey]
          )
        )
      );
    }
  }]);

  return Example;
}(_react2.default.Component);

exports.default = Example;
