// wix-style-react/dist/stories/LanguagePicker/index.story.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _CodeExample = require('wix-storybook-utils/CodeExample');

var _CodeExample2 = _interopRequireDefault(_CodeExample);

var _LanguagePicker = require('wix-style-react/LanguagePicker');

var _LanguagePicker2 = _interopRequireDefault(_LanguagePicker);

var _Example = require('./Example');

var _Example2 = _interopRequireDefault(_Example);

var _Example3 = require('!raw-loader!./Example');

var _Example4 = _interopRequireDefault(_Example3);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var LOCALES = {
  cs: 'Čeština',
  da: 'Dansk',
  de: 'Deutsch',
  en: 'English',
  es: 'Español',
  fr: 'Français',
  hi: 'हिन्दी',
  it: 'Italiano',
  ja: '日本語',
  ko: '한국어',
  nl: 'Nederlands',
  no: 'Norsk',
  pl: 'Polski',
  pt: 'Português',
  ru: 'Русский',
  sv: 'Svenska',
  th: 'Thai',
  tr: 'Türkçe',
  zh: 'Chinese'
};

exports.default = {
  category: 'Deprecated',
  component: _LanguagePicker2.default,
  componentPath: '../../src/LanguagePicker',

  componentProps: {
    children: Object.keys(LOCALES).map(function (lang) {
      return _react2.default.createElement(
        _LanguagePicker2.default.Option,
        { key: lang, languageKey: lang },
        LOCALES[lang]
      );
    })
  },

  exampleProps: {
    onSelect: function onSelect(_ref) {
      var value = _ref.value;
      return value;
    }
  },

  // TODO: update e2e to use generated example and remove `examples` below
  examples: _react2.default.createElement(
    _CodeExample2.default,
    { title: 'Example', code: _Example4.default },
    _react2.default.createElement(_Example2.default, null)
  )
};
