// wix-style-react/dist/src/Grid/Grid.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Card = exports.Col = exports.AutoAdjustedRow = exports.AutoAdjustedColumns = exports.Row = exports.Columns = exports.RawContainer = exports.Container = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Grid = require('./Grid.scss');

var _Grid2 = _interopRequireDefault(_Grid);

var _Card = require('../Card');

var _Card2 = _interopRequireDefault(_Card);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var childrenProp = { children: _propTypes2.default.node };

var RawContainer = function RawContainer(_ref) {
  var children = _ref.children;
  return _react2.default.createElement('div', {
    className: _Grid2.default.rawContainer,
    children: children
  });
};

RawContainer.propTypes = childrenProp;

var Container = function Container(_ref2) {
  var children = _ref2.children;
  return _react2.default.createElement('div', {
    className: _Grid2.default.wixContainer,
    children: children
  });
};

Container.propTypes = childrenProp;

var Columns = function (_Component) {
  _inherits(Columns, _Component);

  function Columns() {
    _classCallCheck(this, Columns);

    return _possibleConstructorReturn(this, (Columns.__proto__ || Object.getPrototypeOf(Columns)).apply(this, arguments));
  }

  _createClass(Columns, [{
    key: 'render',
    value: function render() {
      var _classNames;

      var _props = this.props,
          className = _props.className,
          rtl = _props.rtl,
          stretchViewsVertically = _props.stretchViewsVertically,
          dataHook = _props.dataHook,
          children = _props.children;


      var rowClasses = (0, _classnames2.default)(_Grid2.default.row, className, (_classNames = {}, _defineProperty(_classNames, _Grid2.default.rtl, rtl), _defineProperty(_classNames, _Grid2.default.stretch_vertically_row, stretchViewsVertically), _classNames));

      return _react2.default.createElement('div', {
        className: rowClasses,
        'data-hook': dataHook,
        children: children
      });
    }
  }]);

  return Columns;
}(_react.Component);

Columns.propTypes = {
  children: _propTypes2.default.node,
  className: _propTypes2.default.string,
  rtl: _propTypes2.default.bool,
  stretchViewsVertically: _propTypes2.default.bool,
  dataHook: _propTypes2.default.string
};
Columns.defaultProps = {
  stretchViewsVertically: false
};

var AutoAdjustedColumns = function (_Component2) {
  _inherits(AutoAdjustedColumns, _Component2);

  function AutoAdjustedColumns() {
    var _ref3;

    var _temp, _this2, _ret;

    _classCallCheck(this, AutoAdjustedColumns);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this2 = _possibleConstructorReturn(this, (_ref3 = AutoAdjustedColumns.__proto__ || Object.getPrototypeOf(AutoAdjustedColumns)).call.apply(_ref3, [this].concat(args))), _this2), _this2.DEFAULT_MAX_SPAN = 12, _temp), _possibleConstructorReturn(_this2, _ret);
  }

  _createClass(AutoAdjustedColumns, [{
    key: 'render',
    value: function render() {
      var children = this.props.children;
      var cols = Array.isArray(children) ? children : [children];
      var spanSize = Math.floor(this.DEFAULT_MAX_SPAN / cols.length);

      return _react2.default.createElement(
        'div',
        { className: (0, _classnames2.default)(_Grid2.default.row, _Grid2.default.flexContainer) },
        cols.map(function (col, index) {
          return _react2.default.createElement(Col, {
            span: spanSize,
            key: index,
            children: col
          });
        })
      );
    }
  }]);

  return AutoAdjustedColumns;
}(_react.Component);

AutoAdjustedColumns.propTypes = {
  children: _propTypes2.default.node
};

var Col = function (_Component3) {
  _inherits(Col, _Component3);

  function Col() {
    _classCallCheck(this, Col);

    return _possibleConstructorReturn(this, (Col.__proto__ || Object.getPrototypeOf(Col)).apply(this, arguments));
  }

  _createClass(Col, [{
    key: 'isVisibleHidden',
    value: function isVisibleHidden(str) {
      return str === 'hidden' || str === 'visible';
    }
  }, {
    key: 'isLegalCol',
    value: function isLegalCol(numStr) {
      if (numStr && !this.isVisibleHidden(numStr)) {
        var num = Number(numStr);
        return Number.isInteger(num) && num > 0 && num <= 12;
      }
      return false;
    }
  }, {
    key: 'render',
    value: function render() {
      var _classNames2;

      var _props2 = this.props,
          children = _props2.children,
          className = _props2.className,
          span = _props2.span,
          rtl = _props2.rtl,
          xs = _props2.xs,
          sm = _props2.sm,
          md = _props2.md,
          lg = _props2.lg,
          xl = _props2.xl,
          dataHook = _props2.dataHook;


      var columnClasses = (0, _classnames2.default)(className, _Grid2.default.column, (_classNames2 = {}, _defineProperty(_classNames2, _Grid2.default.rtl, rtl), _defineProperty(_classNames2, _Grid2.default['colXs' + span], this.isLegalCol(span)), _defineProperty(_classNames2, _Grid2.default['colXs' + xs], this.isLegalCol(xs)), _defineProperty(_classNames2, _Grid2.default['colSm' + sm], this.isLegalCol(sm)), _defineProperty(_classNames2, _Grid2.default['colMd' + md], this.isLegalCol(md)), _defineProperty(_classNames2, _Grid2.default['colLg' + lg], this.isLegalCol(lg)), _defineProperty(_classNames2, _Grid2.default['colXl' + xl], this.isLegalCol(xl)), _defineProperty(_classNames2, _Grid2.default[xs + 'Xs'], this.isVisibleHidden(xs)), _defineProperty(_classNames2, _Grid2.default[sm + 'Sm'], this.isVisibleHidden(sm)), _defineProperty(_classNames2, _Grid2.default[md + 'Md'], this.isVisibleHidden(md)), _defineProperty(_classNames2, _Grid2.default[lg + 'Lg'], this.isVisibleHidden(lg)), _classNames2));

      return _react2.default.createElement('div', {
        className: columnClasses,
        'data-hook': dataHook,
        children: children
      });
    }
  }]);

  return Col;
}(_react.Component);

Col.propTypes = {
  children: _propTypes2.default.node,
  className: _propTypes2.default.string,
  span: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  rtl: _propTypes2.default.bool,
  xs: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  sm: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  md: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  lg: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  xl: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  dataHook: _propTypes2.default.string
};
Col.defaultProps = {
  span: 12
};
exports.Container = Container;
exports.RawContainer = RawContainer;
exports.Columns = Columns;
exports.Row = Columns;
exports.AutoAdjustedColumns = AutoAdjustedColumns;
exports.AutoAdjustedRow = AutoAdjustedColumns;
exports.Col = Col;
exports.Card = _Card2.default;
