// wix-style-react/dist/src/SideMenu/core/navigation/LinkLayout.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _styles = require('./styles.scss');

var _styles2 = _interopRequireDefault(_styles);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var LinkLayout = function LinkLayout(_ref) {
  var _classnames;

  var children = _ref.children,
      isDiminishedHover = _ref.isDiminishedHover,
      isActive = _ref.isActive,
      disabled = _ref.disabled,
      rest = _objectWithoutProperties(_ref, ['children', 'isDiminishedHover', 'isActive', 'disabled']);

  return _react2.default.createElement(
    'span',
    _extends({
      className: (0, _classnames3.default)(_styles2.default.linkLayout, (_classnames = {}, _defineProperty(_classnames, _styles2.default.linkActive, isActive), _defineProperty(_classnames, _styles2.default.linkDiminishedHover, isDiminishedHover), _defineProperty(_classnames, _styles2.default.linkDisabled, disabled), _classnames)),
      'data-hook': 'menu-navigation-link-wrapper'
    }, rest),
    children
  );
};

LinkLayout.propTypes = {
  children: _propTypes2.default.node,
  isActive: _propTypes2.default.bool,
  isDiminishedHover: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool
};

exports.default = LinkLayout;
