// wix-style-react/dist/src/Text/Text.driver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _testUtils = require('stylable/test-utils');

var _TextSt = require('./Text.st.css');

var _TextSt2 = _interopRequireDefault(_TextSt);

var _deprecationLog = require('../utils/deprecationLog');

var _deprecationLog2 = _interopRequireDefault(_deprecationLog);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var textDriverFactory = function textDriverFactory(_ref) {
  var element = _ref.element;

  var stylableDOMUtil = new _testUtils.StylableDOMUtil(_TextSt2.default);

  return {
    exists: function exists() {
      return !!element;
    },
    getTagName: function getTagName() {
      return element.tagName.toLowerCase();
    },
    getText: function getText() {
      return element.innerHTML;
    },
    getSize: function getSize() {
      return stylableDOMUtil.getStyleState(element, 'size');
    },
    getSkin: function getSkin() {
      return stylableDOMUtil.getStyleState(element, 'skin');
    },
    getWeight: function getWeight() {
      return stylableDOMUtil.getStyleState(element, 'weight');
    },
    isLight: function isLight() {
      return stylableDOMUtil.hasStyleState(element, 'light');
    },
    isSecondary: function isSecondary() {
      return stylableDOMUtil.hasStyleState(element, 'secondary');
    },

    // Deprecated
    isBold: function isBold() {
      (0, _deprecationLog2.default)('Text testkit method "isBold" is deprecated, use "getWeight" method instead');
      return stylableDOMUtil.hasStyleState(element, 'bold');
    }
  };
};

exports.default = textDriverFactory;
