// wix-style-react/dist/stories/MessageBox/AlertExamples/FootNote.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _MessageBox = require('wix-style-react/MessageBox');

var _Text = require('wix-style-react/Text');

var _Text2 = _interopRequireDefault(_Text);

var _TextLink = require('wix-style-react/TextLink');

var _TextLink2 = _interopRequireDefault(_TextLink);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable react/prop-types */
exports.default = function () {
  return _react2.default.createElement(
    _MessageBox.MessageBoxFunctionalLayout,
    {
      title: 'Interuption Message',
      confirmText: 'Main',
      cancelText: 'Secondary',
      theme: 'blue',
      dataHook: 'alert-footnote',
      footerBottomChildren: _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
          _Text2.default,
          { size: 'small' },
          'By sending an invite, you agree to the '
        ),
        _react2.default.createElement(
          _TextLink2.default,
          { size: 'small' },
          'Wix Terms of Use.'
        )
      )
    },
    'This is a generic message. No harm done, but really needed to interrupt you.'
  );
};
