// wix-style-react/dist/src/RichTextArea/RichTextAreaToolbar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _WixComponent2 = require('../BaseComponents/WixComponent');

var _WixComponent3 = _interopRequireDefault(_WixComponent2);

var _RichTextAreaButton = require('./RichTextAreaButton');

var _RichTextAreaButton2 = _interopRequireDefault(_RichTextAreaButton);

var _RichTextAreaLinkButton = require('./RichTextAreaLinkButton');

var _RichTextAreaLinkButton2 = _interopRequireDefault(_RichTextAreaLinkButton);

var _RichTextAreaToolbar = require('./RichTextAreaToolbar.scss');

var _RichTextAreaToolbar2 = _interopRequireDefault(_RichTextAreaToolbar);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var RichTextAreaToolbar = function (_WixComponent) {
  _inherits(RichTextAreaToolbar, _WixComponent);

  function RichTextAreaToolbar() {
    _classCallCheck(this, RichTextAreaToolbar);

    return _possibleConstructorReturn(this, (RichTextAreaToolbar.__proto__ || Object.getPrototypeOf(RichTextAreaToolbar)).apply(this, arguments));
  }

  _createClass(RichTextAreaToolbar, [{
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        'div',
        { className: _RichTextAreaToolbar2.default.container },
        this.renderButton('mark', 'bold'),
        this.renderButton('mark', 'italic'),
        this.renderButton('mark', 'underline'),
        this.renderLinkButton(),
        this.renderButton('block', 'unordered-list'),
        this.renderButton('block', 'ordered-list'),
        this.props.onImageButtonClick ? this.renderImageButton() : null
      );
    }
  }, {
    key: 'renderButton',
    value: function renderButton(action, type) {
      var _props = this.props,
          _onClick = _props.onClick,
          hasMark = _props.hasMark,
          hasListBlock = _props.hasListBlock,
          disabled = _props.disabled;

      var isActive = action === 'mark' ? hasMark : hasListBlock;

      return _react2.default.createElement(
        'div',
        { className: (0, _classnames2.default)(_RichTextAreaToolbar2.default.button, _defineProperty({}, _RichTextAreaToolbar2.default.disabled, disabled)) },
        _react2.default.createElement(_RichTextAreaButton2.default, {
          disabled: disabled,
          onClick: function onClick() {
            return _onClick(action, type);
          },
          type: type,
          isActive: isActive(type)
        })
      );
    }
  }, {
    key: 'renderLinkButton',
    value: function renderLinkButton() {
      var _props2 = this.props,
          onLinkButtonClick = _props2.onLinkButtonClick,
          hasLink = _props2.hasLink,
          disabled = _props2.disabled,
          isSelectionExpanded = _props2.isSelectionExpanded;


      return _react2.default.createElement(
        'div',
        { className: (0, _classnames2.default)(_RichTextAreaToolbar2.default.button, _defineProperty({}, _RichTextAreaToolbar2.default.disabled, disabled)) },
        _react2.default.createElement(_RichTextAreaLinkButton2.default, {
          selection: this.props.selection,
          disabled: disabled,
          onClick: onLinkButtonClick,
          type: 'link',
          isActive: hasLink(),
          isSelectionExpanded: !isSelectionExpanded
        })
      );
    }
  }, {
    key: 'renderImageButton',
    value: function renderImageButton() {
      var _props3 = this.props,
          onImageButtonClick = _props3.onImageButtonClick,
          disabled = _props3.disabled;


      return _react2.default.createElement(
        'div',
        { className: (0, _classnames2.default)(_RichTextAreaToolbar2.default.button, _defineProperty({}, _RichTextAreaToolbar2.default.disabled, disabled)) },
        _react2.default.createElement(_RichTextAreaButton2.default, {
          disabled: disabled,
          onClick: onImageButtonClick,
          type: 'image',
          isActive: false
        })
      );
    }
  }]);

  return RichTextAreaToolbar;
}(_WixComponent3.default);

RichTextAreaToolbar.propTypes = {
  selection: _propTypes2.default.string,
  onClick: _propTypes2.default.func,
  onLinkButtonClick: _propTypes2.default.func,
  onImageButtonClick: _propTypes2.default.func,
  hasMark: _propTypes2.default.func.isRequired,
  hasListBlock: _propTypes2.default.func.isRequired,
  hasLink: _propTypes2.default.func.isRequired,
  disabled: _propTypes2.default.bool,
  isSelectionExpanded: _propTypes2.default.bool
};

exports.default = RichTextAreaToolbar;
