// wix-style-react/dist/stories/Page/index.story.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _Page = require('wix-style-react/Page');

var _Page2 = _interopRequireDefault(_Page);

var _Breadcrumbs = require('./Breadcrumbs');

var _Breadcrumbs2 = _interopRequireDefault(_Breadcrumbs);

var _storybookConfig = require('./storybookConfig');

var _PageChildren = require('./PageChildren');

require('./Page.scss');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  category: _storybookConfig.storybookConfig.category,
  storyName: _storybookConfig.storybookConfig.storyName,
  name: 'Page',
  component: _Page2.default,
  componentPath: '../../src/Page',

  componentProps: {
    children: [(0, _PageChildren.header)(_Breadcrumbs2.default), _PageChildren.tail, (0, _PageChildren.content)(false)],
    dataHook: 'story-page',
    gradientClassName: 'background-gradient',
    gradientCoverTail: true,
    backgroundImageUrl: 'https://static.wixstatic.com/media/f0548921c53940ec803dfb1c203e96fe.jpg/v1/fill/w_400,h_100/f0548921c53940ec803dfb1c203e96fe.jpg'
  },

  exampleProps: {
    children: [{
      label: 'header, tail & content',
      value: [(0, _PageChildren.header)(_Breadcrumbs2.default), _PageChildren.tail, (0, _PageChildren.content)(false)]
    }, {
      label: 'header & content',
      value: [(0, _PageChildren.header)(_Breadcrumbs2.default), (0, _PageChildren.content)(false)]
    }, {
      label: 'just content',
      value: [(0, _PageChildren.content)(false)]
    }, {
      label: 'header, tail, fixed-content & content',
      value: [(0, _PageChildren.header)(_Breadcrumbs2.default), _PageChildren.tail, _PageChildren.fixedContent, (0, _PageChildren.content)(false)]
    }],
    backgroundImageUrl: [{
      label: 'https://some-host.com/image-path.jpg',
      value: 'https://static.wixstatic.com/media/f0548921c53940ec803dfb1c203e96fe.jpg/v1/fill/w_400,h_100/f0548921c53940ec803dfb1c203e96fe.jpg'
    }]
  }
};
