// wix-style-react/dist/stories/Page/PageChildren.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fixedContent = exports.tail = exports.content = exports.header = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Page = require('wix-style-react/Page');

var _Page2 = _interopRequireDefault(_Page);

var _Button = require('wix-style-react/Button');

var _Button2 = _interopRequireDefault(_Button);

var _SomeContentComponent = require('./SomeContentComponent');

var _SomeContentComponent2 = _interopRequireDefault(_SomeContentComponent);

var _SomeTailComponent = require('./SomeTailComponent');

var _SomeTailComponent2 = _interopRequireDefault(_SomeTailComponent);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var header = exports.header = function header(breadcrumbs) {
  return _react2.default.createElement(_Page2.default.Header, {
    breadcrumbs: breadcrumbs,
    title: 'Page Title',
    subtitle: 'Page subtitle',
    showBackButton: true,
    onBackClicked: function onBackClicked() {},
    actionsBar: _react2.default.createElement(
      _Button2.default,
      null,
      'Action'
    )
  });
};

var content = exports.content = function content(showScss) {
  return _react2.default.createElement(
    _Page2.default.Content,
    null,
    _react2.default.createElement(_SomeContentComponent2.default, { showScss: showScss })
  );
};

var tail = exports.tail = _react2.default.createElement(
  _Page2.default.Tail,
  null,
  _react2.default.createElement(_SomeTailComponent2.default, null)
);

var fixedContent = exports.fixedContent = _react2.default.createElement(
  _Page2.default.FixedContent,
  null,
  _react2.default.createElement(
    'div',
    { style: { padding: '10px 0px', backgroundColor: 'white' } },
    'This is a fixedContent'
  )
);
