// wix-style-react/dist/stories/Text/index.story.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Text = require('wix-style-react/Text');

var _Text2 = _interopRequireDefault(_Text);

var _Text3 = require('../../src/Text/Text');

var _CodeExample = require('wix-storybook-utils/CodeExample');

var _CodeExample2 = _interopRequireDefault(_CodeExample);

var _ExampleTextTypography = require('./ExampleTextTypography');

var _ExampleTextTypography2 = _interopRequireDefault(_ExampleTextTypography);

var _ExampleTextTypography3 = require('!raw-loader!./ExampleTextTypography');

var _ExampleTextTypography4 = _interopRequireDefault(_ExampleTextTypography3);

var _ExampleMultiline = require('./ExampleMultiline');

var _ExampleMultiline2 = _interopRequireDefault(_ExampleMultiline);

var _ExampleMultiline3 = require('!raw-loader!./ExampleMultiline');

var _ExampleMultiline4 = _interopRequireDefault(_ExampleMultiline3);

var _ExampleEllipsis = require('./ExampleEllipsis');

var _ExampleEllipsis2 = _interopRequireDefault(_ExampleEllipsis);

var _ExampleEllipsis3 = require('!raw-loader!./ExampleEllipsis');

var _ExampleEllipsis4 = _interopRequireDefault(_ExampleEllipsis3);

var _ExampleH1TagName = require('./ExampleH1TagName');

var _ExampleH1TagName2 = _interopRequireDefault(_ExampleH1TagName);

var _ExampleH1TagName3 = require('!raw-loader!./ExampleH1TagName');

var _ExampleH1TagName4 = _interopRequireDefault(_ExampleH1TagName3);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  category: '1. Foundation',
  storyName: '1.2 Text',
  component: _Text2.default,
  componentPath: '../../src/Text/Text.js',

  componentProps: {
    children: 'Some text',
    light: false,
    dataHook: 'storybook-text',
    size: _Text3.SIZES.medium,
    secondary: false,
    skin: _Text3.SKINS.standard,
    weight: _Text3.WEIGHTS.thin,
    tagName: 'span',
    ellipsis: false
  },

  examples: _react2.default.createElement(
    'div',
    null,
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Multiline Example', code: _ExampleMultiline4.default },
      _react2.default.createElement(_ExampleMultiline2.default, null)
    ),
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Ellipsis Example', code: _ExampleEllipsis4.default },
      _react2.default.createElement(_ExampleEllipsis2.default, null)
    ),
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Custom TagName Example', code: _ExampleH1TagName4.default },
      _react2.default.createElement(_ExampleH1TagName2.default, null)
    ),
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Typography Examples', code: _ExampleTextTypography4.default },
      _react2.default.createElement(_ExampleTextTypography2.default, null)
    )
  )
};
