// wix-style-react/dist/src/providers/WixStyleTpaProvider.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _Wix = require('Wix');

var _Wix2 = _interopRequireDefault(_Wix);

var _WixStyleProvider = require('./WixStyleProvider');

var _WixStyleProvider2 = _interopRequireDefault(_WixStyleProvider);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var WixStyleTpaProvider = function (_React$Component) {
  _inherits(WixStyleTpaProvider, _React$Component);

  function WixStyleTpaProvider(props) {
    _classCallCheck(this, WixStyleTpaProvider);

    var _this = _possibleConstructorReturn(this, (WixStyleTpaProvider.__proto__ || Object.getPrototypeOf(WixStyleTpaProvider)).call(this, props));

    _this.update = _this.update.bind(_this);
    _this.state = {
      theme: props.themeCreator(),
      events: props.events.filter(function (event) {
        return _Wix2.default.Events[event] !== undefined;
      })
    };
    return _this;
  }

  _createClass(WixStyleTpaProvider, [{
    key: 'update',
    value: function update(data) {
      this.setState({ theme: this.props.themeCreator(data) });
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      this.state.events.forEach(function (event) {
        return _Wix2.default.addEventListener(_Wix2.default.Events[event], _this2.update);
      });
    }
  }, {
    key: 'componentWillUnmout',
    value: function componentWillUnmout() {
      var _this3 = this;

      this.state.events.forEach(function (event) {
        return _Wix2.default.removeEventListener(_Wix2.default.Events[event], _this3.update);
      });
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        _WixStyleProvider2.default,
        { theme: this.state.theme },
        this.props.children
      );
    }
  }]);

  return WixStyleTpaProvider;
}(_react2.default.Component);

exports.default = WixStyleTpaProvider;


WixStyleTpaProvider.propTypes = {
  themeCreator: _propTypes.func,
  children: _propTypes.any,
  events: _propTypes.array
};

WixStyleTpaProvider.defaultProps = {
  events: ['STYLE_PARAMS_CHANGE']
};
