// wix-style-react/dist/src/MessageBox/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MessageBoxLayout2 = exports.MessageBoxFunctionalLayout = exports.FooterLayout1 = exports.FooterLayout = exports.HeaderLayout1 = exports.HeaderLayout = exports.MessageBoxLayout1 = exports.MessageBoxMarketerialLayout = undefined;

var _MessageBoxMarketerialLayout = require('./MessageBoxMarketerialLayout');

var _MessageBoxMarketerialLayout2 = _interopRequireDefault(_MessageBoxMarketerialLayout);

var _MessageBoxFunctionalLayout = require('./MessageBoxFunctionalLayout');

var _MessageBoxFunctionalLayout2 = _interopRequireDefault(_MessageBoxFunctionalLayout);

var _HeaderLayout = require('./HeaderLayout');

var _HeaderLayout2 = _interopRequireDefault(_HeaderLayout);

var _FooterLayout = require('./FooterLayout');

var _FooterLayout2 = _interopRequireDefault(_FooterLayout);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.MessageBoxMarketerialLayout = _MessageBoxMarketerialLayout2.default;
exports.MessageBoxLayout1 = _MessageBoxMarketerialLayout2.default;
exports.HeaderLayout = _HeaderLayout2.default;
exports.HeaderLayout1 = _HeaderLayout2.default;
exports.FooterLayout = _FooterLayout2.default;
exports.FooterLayout1 = _FooterLayout2.default;
exports.MessageBoxFunctionalLayout = _MessageBoxFunctionalLayout2.default;
exports.MessageBoxLayout2 = _MessageBoxFunctionalLayout2.default;
