// wix-style-react/dist/src/MessageBox/MessageBoxFunctionalLayout.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _HeaderLayout = require('./HeaderLayout');

var _HeaderLayout2 = _interopRequireDefault(_HeaderLayout);

var _FooterLayout = require('./FooterLayout');

var _FooterLayout2 = _interopRequireDefault(_FooterLayout);

var _WixComponent2 = require('../BaseComponents/WixComponent');

var _WixComponent3 = _interopRequireDefault(_WixComponent2);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _throttle = require('lodash/throttle');

var _throttle2 = _interopRequireDefault(_throttle);

var _MessageBoxFunctionalLayout = require('./MessageBoxFunctionalLayout.scss');

var _MessageBoxFunctionalLayout2 = _interopRequireDefault(_MessageBoxFunctionalLayout);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var MessageBoxFunctionalLayout = function (_WixComponent) {
  _inherits(MessageBoxFunctionalLayout, _WixComponent);

  function MessageBoxFunctionalLayout(props) {
    _classCallCheck(this, MessageBoxFunctionalLayout);

    var _this = _possibleConstructorReturn(this, (MessageBoxFunctionalLayout.__proto__ || Object.getPrototypeOf(MessageBoxFunctionalLayout)).call(this, props));

    _this._initializeMessageBoxRef = function (el) {
      if (el && el.scrollHeight > el.clientHeight) {
        _this.setState({ hasScroll: true });

        _this.messageBoxRef = el;
        _this.messageBoxRef.addEventListener('scroll', _this._handleMessageBoxScroll);
      }
    };

    _this._handleMessageBoxScroll = (0, _throttle2.default)(function () {
      var scrolledToBottom = _this.messageBoxRef.scrollTop + _this.messageBoxRef.clientHeight === _this.messageBoxRef.scrollHeight;

      if (scrolledToBottom !== _this.state.scrolledToBottom) {
        _this.setState({ scrolledToBottom: scrolledToBottom });
      }
    }, 16);


    _this.state = {
      hasScroll: false,
      scrolledToBottom: false
    };
    _this.messageBoxRef = null;
    return _this;
  }

  _createClass(MessageBoxFunctionalLayout, [{
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      if (this.state.hasScroll) {
        this.messageBoxRef.removeEventListener('scroll', this._handleMessageBoxScroll);
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _classNames;

      var _props = this.props,
          title = _props.title,
          onCancel = _props.onCancel,
          onOk = _props.onOk,
          onClose = _props.onClose,
          confirmText = _props.confirmText,
          cancelText = _props.cancelText,
          children = _props.children,
          buttonsHeight = _props.buttonsHeight,
          hideFooter = _props.hideFooter,
          footerBottomChildren = _props.footerBottomChildren,
          theme = _props.theme,
          closeButton = _props.closeButton,
          disableConfirmation = _props.disableConfirmation,
          disableCancel = _props.disableCancel,
          width = _props.width,
          noBodyPadding = _props.noBodyPadding,
          maxHeight = _props.maxHeight,
          fullscreen = _props.fullscreen;
      var _state = this.state,
          hasScroll = _state.hasScroll,
          scrolledToBottom = _state.scrolledToBottom;


      var messageBoxBodyClassNames = (0, _classnames2.default)(_MessageBoxFunctionalLayout2.default.body, (_classNames = {}, _defineProperty(_classNames, _MessageBoxFunctionalLayout2.default.scrollable, typeof maxHeight !== 'undefined'), _defineProperty(_classNames, _MessageBoxFunctionalLayout2.default.noPadding, noBodyPadding), _defineProperty(_classNames, _MessageBoxFunctionalLayout2.default.fullscreenBody, fullscreen), _defineProperty(_classNames, _MessageBoxFunctionalLayout2.default.noFooter, hideFooter), _defineProperty(_classNames, _MessageBoxFunctionalLayout2.default.footerBorder, hasScroll && !scrolledToBottom), _classNames));
      var messageBoxBodyStyle = {
        maxHeight: maxHeight
      };

      var contentClassName = (0, _classnames2.default)(_MessageBoxFunctionalLayout2.default.content, _defineProperty({}, _MessageBoxFunctionalLayout2.default.fullscreenContent, fullscreen));

      return _react2.default.createElement(
        'div',
        { className: contentClassName, style: { width: width } },
        _react2.default.createElement(_HeaderLayout2.default, { title: title, onCancel: onClose ? onClose : onCancel, theme: theme, closeButton: closeButton }),
        _react2.default.createElement(
          'div',
          {
            'data-hook': 'message-box-body',
            className: messageBoxBodyClassNames,
            style: messageBoxBodyStyle,
            ref: this._initializeMessageBoxRef
          },
          children
        ),
        !hideFooter ? _react2.default.createElement(_FooterLayout2.default, {
          bottomChildren: footerBottomChildren,
          enableCancel: !disableCancel,
          enableOk: !disableConfirmation,
          buttonsHeight: buttonsHeight,
          confirmText: confirmText,
          cancelText: cancelText,
          onCancel: onCancel,
          onOk: onOk,
          theme: theme
        }) : null
      );
    }
  }]);

  return MessageBoxFunctionalLayout;
}(_WixComponent3.default);

MessageBoxFunctionalLayout.propTypes = {
  hideFooter: _propTypes2.default.bool,
  confirmText: _propTypes2.default.node,
  cancelText: _propTypes2.default.node,
  theme: _propTypes2.default.string,
  onOk: _propTypes2.default.func,
  onCancel: _propTypes2.default.func,
  onClose: _propTypes2.default.func,
  width: _propTypes2.default.string,
  title: _propTypes2.default.node,
  children: _propTypes2.default.any,
  maxHeight: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]),
  buttonsHeight: _propTypes2.default.string,
  closeButton: _propTypes2.default.bool,
  disableCancel: _propTypes2.default.bool,
  disableConfirmation: _propTypes2.default.bool,
  noBodyPadding: _propTypes2.default.bool,
  footerBottomChildren: _propTypes2.default.node,
  fullscreen: _propTypes2.default.bool
};

MessageBoxFunctionalLayout.defaultProps = {
  buttonsHeight: 'small',
  disableCancel: false,
  disableConfirmation: false,
  noBodyPadding: false,
  fullscreen: false
};

exports.default = MessageBoxFunctionalLayout;
