// playable/dist/src/develop.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var index_1 = tslib_1.__importDefault(require("./index"));
var hls_1 = tslib_1.__importDefault(require("./adapters/hls"));
var dash_1 = tslib_1.__importDefault(require("./adapters/dash"));
var subtitles_1 = tslib_1.__importDefault(require("./modules/ui/subtitles/subtitles"));
var types_1 = require("./modules/playback-engine/types");
var DEFAULT_URLS = {
    DASH: 'https://dash.akamaized.net/akamai/bbb_30fps/bbb_30fps.mpd',
    HLS: 'https://files.wixstatic.com/files/video/64b2fa_039e5c16db504dbaad166ba28d377744/repackage/hls',
    MP4: 'https://wixmp-01bd43eabd844aac9eab64f5.wixmp.com/videos/output/720p/Highest Peak.mp4',
    'MP4-VERTICAL': 'https://storage.googleapis.com/video-player-media-server-static/videoplayback.mp4',
};
index_1.default.registerModule('subtitles', subtitles_1.default);
index_1.default.registerPlaybackAdapter(hls_1.default);
index_1.default.registerPlaybackAdapter(dash_1.default);
Object.defineProperty(window, 'Playable', {
    value: index_1.default,
});
document.addEventListener('DOMContentLoaded', function () {
    var player = index_1.default.create({
        preload: types_1.PreloadTypes.METADATA,
        size: {
            width: 800,
            height: 600,
        },
        playInline: true,
        overlay: false,
        logo: {
            src: '',
        },
    });
    Object.defineProperty(window, 'player', {
        value: player,
    });
    var selectVideo = function (type) {
        player.setSrc({
            type: type,
            url: DEFAULT_URLS[type],
        });
        player.setTitle(type + " format");
    };
    selectVideo('MP4');
    document.getElementById('types').addEventListener('click', function (event) {
        var type = event.target.dataset.type;
        if (!type) {
            return;
        }
        if (type === 'MP4-VERTICAL') {
            player.setSrc({
                type: 'MP4',
                url: DEFAULT_URLS[type],
            });
            player.setTitle("" + type);
        }
        else {
            selectVideo(type);
        }
    });
    document.getElementById('theme-switcher').addEventListener('click', function (event) {
        var color = event.target.dataset.color;
        if (!color) {
            return;
        }
        player.updateTheme({ progressColor: color });
    });
    player.attachToElement(document.getElementById('player-wrapper'));
});
