// wix-style-react/dist/stories/EndorseContentLayout/examples/error.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _EndorseContentLayout = require('wix-style-react/EndorseContentLayout');

var _EndorseContentLayout2 = _interopRequireDefault(_EndorseContentLayout);

var _Backoffice = require('wix-style-react/Backoffice');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var secondaryCtaStyle = {
  fontSize: '14px',
  color: '#2b81cb',
  textDecoration: 'none'
};

exports.default = function () {
  return _react2.default.createElement(_EndorseContentLayout2.default, {
    head: 'Oops\u2026 Something Went Wrong',
    content: 'We\u2019re sorry, the application is experiencing problems. Please refresh the page or give it a try later. If the problem persists please contact Support.',
    primaryCta: _react2.default.createElement(
      _Backoffice.Button,
      { theme: 'fullred', onClick: function onClick() {
          return alert('im trying!');
        } },
      'Try Again'
    ),
    secondaryCta: _react2.default.createElement(
      'a',
      { href: '#', style: secondaryCtaStyle },
      'Cancel'
    )
  });
};
