// wix-style-react/dist/src/BaseComponents/TextLinkLayout/TextLinkLayout.driver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _testUtils = require('react-dom/test-utils');

var _testUtils2 = _interopRequireDefault(_testUtils);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var textLinkLayoutDriverFactory = function textLinkLayoutDriverFactory(_ref) {
  var element = _ref.element,
      wrapper = _ref.wrapper,
      component = _ref.component;


  return {
    exists: function exists() {
      return !!element;
    },
    getContent: function getContent() {
      return element.textContent;
    },
    doesComponentHasClass: function doesComponentHasClass(className) {
      return element.className.includes(className);
    },
    isDarkBackground: function isDarkBackground() {
      return element.style._values.color === 'rgb(240, 244, 247)';
    },
    isGreyScale: function isGreyScale() {
      return element.style._values.color === 'rgb(22, 45, 61)';
    },
    hover: function hover() {
      return _testUtils2.default.Simulate.mouseEnter(element);
    },
    isUnderline: function isUnderline() {
      return element.style._values['text-decoration'] === 'underline';
    },
    getDisplay: function getDisplay() {
      return element.style._values.display;
    },
    isLightBackground: function isLightBackground() {
      return element.style._values.color === 'rgb(56, 153, 236)';
    },
    getSize: function getSize() {
      return element.classList.contains('t1_3') ? 'medium' : element.classList.contains('t3_3') ? 'small' : 'unknown';
    },
    setProps: function setProps(props) {
      var ClonedWithProps = _react2.default.cloneElement.apply(_react2.default, [component, Object.assign({}, component.props, props)].concat(_toConsumableArray(component.props.children || [])));
      _reactDom2.default.render(_react2.default.createElement(
        'div',
        { ref: function ref(r) {
            return element = r;
          } },
        ClonedWithProps
      ), wrapper);
    }
  };
};

exports.default = textLinkLayoutDriverFactory;
