// wix-style-react/dist/stories/PageHeader/index.story.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _PageHeader = require('wix-style-react/PageHeader');

var _PageHeader2 = _interopRequireDefault(_PageHeader);

var _Button = require('wix-style-react/Button');

var _Button2 = _interopRequireDefault(_Button);

var _Breadcrumbs = require('../Page/Breadcrumbs');

var _Breadcrumbs2 = _interopRequireDefault(_Breadcrumbs);

require('./PageHeader.scss');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  category: '2. Layout',
  storyName: '2.5 + PageHeader',
  component: _PageHeader2.default,
  componentPath: '../../src/PageHeader',

  componentProps: {
    onBackClicked: function onBackClicked() {},
    title: 'Page Header',
    dataHook: 'story-page-header'
  },

  exampleProps: {
    breadcrumbs: [{ label: 'Breadcrumbs', value: _Breadcrumbs2.default }],
    actionsBar: [{ label: 'Button', value: _react2.default.createElement(
        _Button2.default,
        null,
        'Action'
      ) }, {
      label: 'Two buttons',
      value: _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
          _Button2.default,
          null,
          'Button #1'
        ),
        _react2.default.createElement(
          _Button2.default,
          null,
          'Button #2'
        )
      )
    }]
  }
};
