// wix-style-react/dist/stories/Input/ExampleRefs.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Input = require('wix-style-react/Input');

var _Input2 = _interopRequireDefault(_Input);

var _TextButton = require('wix-storybook-utils/TextButton');

var _TextButton2 = _interopRequireDefault(_TextButton);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var CommandsExample = function (_React$Component) {
  _inherits(CommandsExample, _React$Component);

  function CommandsExample() {
    _classCallCheck(this, CommandsExample);

    return _possibleConstructorReturn(this, (CommandsExample.__proto__ || Object.getPrototypeOf(CommandsExample)).apply(this, arguments));
  }

  _createClass(CommandsExample, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var handleClick1 = function handleClick1() {
        return _this2.refs.inputtest.focus();
      };

      var handleClick2 = function handleClick2() {
        _this2.refs.inputtest.focus();
        setTimeout(function () {
          return _this2.refs.inputtest.blur();
        }, 1000);
      };

      var handleClick3 = function handleClick3() {
        _this2.refs.inputtest.focus();
        _this2.refs.inputtest.select();
      };

      return _react2.default.createElement(
        'div',
        { className: 'ltr' },
        _react2.default.createElement(
          'div',
          { style: { width: '400px' } },
          _react2.default.createElement(_Input2.default, { theme: this.props.theme, ref: 'inputtest' })
        ),
        _react2.default.createElement(
          _TextButton2.default,
          { onClick: handleClick1 },
          'Focus'
        ),
        _react2.default.createElement(
          _TextButton2.default,
          { onClick: handleClick2 },
          'Focus & blur 1 second later'
        ),
        _react2.default.createElement(
          _TextButton2.default,
          { onClick: handleClick3 },
          'Select text'
        )
      );
    }
  }]);

  return CommandsExample;
}(_react2.default.Component);

CommandsExample.propTypes = {
  theme: _propTypes2.default.string
};

var Example = function Example(_ref) {
  var theme = _ref.theme;
  return _react2.default.createElement(
    'div',
    null,
    _react2.default.createElement(CommandsExample, { theme: theme })
  );
};

Example.propTypes = {
  theme: _propTypes2.default.string
};

exports.default = Example;
