// wix-style-react/dist/stories/Input/ExampleStandard.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Input = require('wix-style-react/Input');

var _Input2 = _interopRequireDefault(_Input);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var style = {
  display: 'inline-block',
  padding: '0 5px',
  width: '200px',
  lineHeight: '22px'
};

var Example = function Example(_ref) {
  var theme = _ref.theme;
  return _react2.default.createElement(
    'div',
    null,
    _react2.default.createElement(
      'div',
      { className: 'ltr', style: style },
      'Input',
      _react2.default.createElement('br', null),
      _react2.default.createElement(_Input2.default, { theme: theme })
    ),
    _react2.default.createElement(
      'div',
      { className: 'ltr', style: style },
      'Focus',
      _react2.default.createElement(_Input2.default, { theme: theme, forceFocus: true })
    ),
    _react2.default.createElement(
      'div',
      { className: 'ltr', style: style },
      'Hover',
      _react2.default.createElement(_Input2.default, { theme: theme, forceHover: true })
    ),
    _react2.default.createElement(
      'div',
      { className: 'ltr', style: style },
      'With placeholder',
      _react2.default.createElement(_Input2.default, { theme: theme, placeholder: 'Search...' })
    ),
    _react2.default.createElement(
      'div',
      { className: 'ltr', style: style },
      'Disabled',
      _react2.default.createElement('br', null),
      _react2.default.createElement(_Input2.default, { theme: theme, disabled: true, placeholder: 'disabled' })
    ),
    _react2.default.createElement(
      'div',
      { className: 'ltr', style: style },
      'Disabled with search',
      _react2.default.createElement('br', null),
      _react2.default.createElement(_Input2.default, { theme: theme, magnifyingGlass: true, disabled: true, placeholder: 'disabled with search' })
    ),
    theme === 'amaterial' && _react2.default.createElement(
      'div',
      { className: 'ltr', style: style },
      _react2.default.createElement(_Input2.default, { title: 'With Title', placeholder: 'Placeholder...', theme: theme })
    )
  );
};

Example.propTypes = {
  theme: _propTypes2.default.string
};

exports.default = Example;
