// wix-style-react/dist/stories/Input/ExampleError.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Input = require('wix-style-react/Input');

var _Input2 = _interopRequireDefault(_Input);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var style = {
  display: 'inline-block',
  padding: '0 5px',
  width: '200px',
  lineHeight: '22px'
};

var Example = function Example(_ref) {
  var theme = _ref.theme;
  return _react2.default.createElement(
    'div',
    null,
    _react2.default.createElement(
      'div',
      { className: 'ltr', style: style },
      'Left to right',
      _react2.default.createElement(_Input2.default, { theme: theme, status: 'error' })
    ),
    _react2.default.createElement(
      'div',
      { className: 'rtl', style: style },
      'Right to left',
      _react2.default.createElement(_Input2.default, { theme: theme, rtl: true, status: 'error' })
    ),
    _react2.default.createElement(
      'div',
      { className: 'ltr', style: style },
      'Error with tooltip',
      _react2.default.createElement(_Input2.default, { theme: theme, status: 'error', statusMessage: 'I am a message' })
    )
  );
};

Example.propTypes = {
  theme: _propTypes2.default.string
};

exports.default = Example;
