// wix-style-react/dist/stories/Input/ExampleSizes.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Input = require('wix-style-react/Input');

var _Input2 = _interopRequireDefault(_Input);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var style = {
  display: 'inline-block',
  padding: '0 5px',
  width: '200px',
  lineHeight: '22px',
  verticalAlign: 'top'
};

var defaultProps = {
  size: 'normal',
  magnifyingGlass: true,
  placeholder: 'They did not know it was impossible, so they did it!',
  unit: '$'
};

var Example = function Example(_ref) {
  var theme = _ref.theme;
  return _react2.default.createElement(
    'div',
    { className: 'ltr' },
    _react2.default.createElement(
      'div',
      { style: style },
      'Small',
      _react2.default.createElement(_Input2.default, _extends({ theme: theme }, defaultProps, { size: 'small' })),
      _react2.default.createElement('br', null),
      _react2.default.createElement(_Input2.default, _extends({ theme: theme }, defaultProps, { size: 'small', error: true }))
    ),
    _react2.default.createElement(
      'div',
      { style: style },
      'Normal',
      _react2.default.createElement('br', null),
      _react2.default.createElement(_Input2.default, _extends({ theme: theme }, defaultProps, { size: 'normal' })),
      _react2.default.createElement('br', null),
      _react2.default.createElement(_Input2.default, _extends({ theme: theme }, defaultProps, { size: 'normal', error: true }))
    ),
    _react2.default.createElement(
      'div',
      { style: style },
      'Large',
      _react2.default.createElement('br', null),
      _react2.default.createElement(_Input2.default, _extends({ theme: theme }, defaultProps, { size: 'large' })),
      _react2.default.createElement('br', null),
      _react2.default.createElement(_Input2.default, _extends({ theme: theme }, defaultProps, { size: 'large', error: true }))
    )
  );
};

Example.propTypes = {
  theme: _propTypes2.default.string
};

exports.default = Example;
