// wix-ui-backoffice/dist/src/components/TimePicker/TimePicker.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var propTypes = require("prop-types");
var omit = require("lodash/omit");
var FormFieldSpinnerUp_1 = require("wix-ui-icons-common/system/FormFieldSpinnerUp");
var FormFieldSpinnerDown_1 = require("wix-ui-icons-common/system/FormFieldSpinnerDown");
var TimePicker_1 = require("wix-ui-core/TimePicker");
var constants_1 = require("wix-ui-core/dist/src/components/TimePicker/constants");
var TimePicker_st_css_1 = require("./TimePicker.st.css");
var constants_2 = require("./constants");
var defaultProps = {
    size: constants_2.Size.medium,
    disableAmPm: false,
    error: false,
    style: {},
    tickerUpIcon: React.createElement(FormFieldSpinnerUp_1.default, null),
    tickerDownIcon: React.createElement(FormFieldSpinnerDown_1.default, null),
    value: TimePicker_1.TimePicker.defaultProps.value
};
function getInputWidthState(timePickerWidth, size, disableAmPm) {
    if (timePickerWidth) {
        return '';
    }
    var states = [size];
    if (!disableAmPm) {
        states.push('ampm');
    }
    return states.join('_');
}
var TimePicker = /** @class */ (function (_super) {
    __extends(TimePicker, _super);
    function TimePicker() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    TimePicker.prototype.focus = function () {
        this.coreTimePickerRef.focus();
    };
    TimePicker.prototype.blur = function () {
        this.coreTimePickerRef.blur();
    };
    TimePicker.prototype.render = function () {
        var _this = this;
        var _a = this.props, disableAmPm = _a.disableAmPm, size = _a.size, error = _a.error, width = _a.width, disabled = _a.disabled, inlineStyle = _a.style;
        var styleObj = width ? __assign({}, inlineStyle, { width: width }) : inlineStyle;
        var coreTimePickerProps = __assign({}, omit(this.props, 'size', 'disableAmPm', 'width', 'style'), { useAmPm: disableAmPm ? constants_1.AmPmOptions.None : constants_1.AmPmOptions.Uppercase });
        return (React.createElement(TimePicker_1.TimePicker, __assign({ ref: function (ref) { return _this.coreTimePickerRef = ref; } }, coreTimePickerProps, TimePicker_st_css_1.default('root', { size: size, error: error, disabled: disabled, inputWidth: getInputWidthState(width, size, disableAmPm) }, coreTimePickerProps), { style: styleObj })));
    };
    TimePicker.displayName = 'TimePickerBackoffice';
    TimePicker.defaultProps = defaultProps;
    TimePicker.propTypes = __assign({}, omit(TimePicker_1.TimePicker.propTypes, 'useAmPm'), { size: propTypes.oneOf(Object.keys(constants_2.Size)), disableAmPm: propTypes.bool, width: propTypes.string, style: propTypes.object });
    return TimePicker;
}(React.PureComponent));
exports.TimePicker = TimePicker;
