// wix-style-react/dist/stories/MultiSelect/ExampleStandard.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.valueParser = exports.options = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _MultiSelect = require('wix-style-react/MultiSelect');

var _MultiSelect2 = _interopRequireDefault(_MultiSelect);

var _ExampleStandard = require('./ExampleStandard.scss');

var _ExampleStandard2 = _interopRequireDefault(_ExampleStandard);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var options = exports.options = [{ value: 'Alabama', id: 'Alabama', tag: { label: 'Alabama' } }, { value: 'Alaska', id: 'Alaska' }, { value: _react2.default.createElement(
    'div',
    { className: _ExampleStandard2.default.option },
    _react2.default.createElement(
      'div',
      null,
      'Arizona'
    ),
    _react2.default.createElement('div', { className: _ExampleStandard2.default.thumb })
  ), id: 'Arizona', tag: { label: 'Arizona', thumb: _react2.default.createElement('div', { className: _ExampleStandard2.default.thumb }) } }, { value: 'Arkansas', id: 'Arkansas' }, { value: 'California', id: 'California' }, { value: 'California2', id: 'California2' }, { value: 'California3', id: 'California3' }, { value: 'California4', id: 'California4' }, { value: 'California5', id: 'California5' }, { value: 'California6', id: 'California6' }, { value: 'California7', id: 'California7' }, { id: 'Divider', value: '-' }, { value: 'Two words', id: 'Two words' }];

var valueParser = exports.valueParser = function valueParser(option) {
  return option.tag ? option.tag.label : option.value;
};

var ExampleStandard = function (_React$Component) {
  _inherits(ExampleStandard, _React$Component);

  function ExampleStandard(props) {
    _classCallCheck(this, ExampleStandard);

    var _this = _possibleConstructorReturn(this, (ExampleStandard.__proto__ || Object.getPrototypeOf(ExampleStandard)).call(this, props));

    _this.getValue = function (option) {
      return typeof option.value === 'string' ? option.value : option.value.props.children[0].props.children;
    };

    _this.handleOnSelect = function (tags) {
      return Array.isArray(tags) ? _this.setState({ tags: [].concat(_toConsumableArray(_this.state.tags), _toConsumableArray(tags)) }) : _this.setState({ tags: [].concat(_toConsumableArray(_this.state.tags), [tags]) });
    };

    _this.handleOnRemoveTag = function (tagId) {
      return _this.setState({ tags: _this.state.tags.filter(function (currTag) {
          return currTag.id !== tagId;
        }) });
    };

    _this.handleOnChange = function (event) {
      return _this.setState({ inputValue: event.target.value });
    };

    _this.predicate = function (option) {
      return _this.getValue(option).toLowerCase().includes(_this.state.inputValue.toLowerCase());
    };

    _this.state = {
      tags: [],
      options: options,
      inputValue: ''
    };
    return _this;
  }

  _createClass(ExampleStandard, [{
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        'div',
        { className: 'ltr' },
        _react2.default.createElement(
          'div',
          { className: _ExampleStandard2.default.main },
          _react2.default.createElement(_MultiSelect2.default, {
            tags: this.state.tags,
            onSelect: this.handleOnSelect,
            onRemoveTag: this.handleOnRemoveTag,
            onChange: this.handleOnChange,
            onManuallyInput: function onManuallyInput() {
              return console.log('NOW');
            },
            options: this.state.options,
            value: this.state.inputValue,
            predicate: this.predicate,
            valueParser: valueParser
          })
        )
      );
    }
  }]);

  return ExampleStandard;
}(_react2.default.Component);

exports.default = ExampleStandard;
