// wix-style-react/dist/stories/Tooltip/Core/ExampleMove.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Tooltip = require('wix-style-react/Tooltip');

var _Tooltip2 = _interopRequireDefault(_Tooltip);

var _Example = require('./Example.scss');

var _Example2 = _interopRequireDefault(_Example);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function () {
  return _react2.default.createElement(
    _Tooltip2.default,
    {
      active: true,
      placement: 'right',
      alignment: 'top',
      textAlign: 'left',
      content: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. In vel erat eu sem rutrum tristique. Sed id odio non magna sodales egestas.',
      showTrigger: 'custom',
      hideTrigger: 'custom',
      theme: 'dark',
      moveBy: { x: -40, y: 40 },
      maxWidth: '400px',
      moveArrowTo: 20
    },
    _react2.default.createElement(
      'div',
      { className: _Example2.default.box },
      _react2.default.createElement(
        'p',
        null,
        'Lorem ipsum dolor sit amet, consectetur adipiscing elit. In vel erat eu sem rutrum tristique. Sed id odio non magna sodales egestas. Sed a magna elit. Suspendisse potenti. Quisque vehicula magna commodo eros sodales, eu volutpat tellus laoreet. Ut euismod elementum libero ac lobortis. Phasellus efficitur odio nec lorem sollicitudin, id maximus nunc porta. Integer ut felis eleifend, finibus neque eget, rutrum odio. Proin ultrices a quam in finibus. Cras sit amet rutrum velit. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos.'
      )
    )
  );
};
