// wix-ui-backoffice/dist/src/components/CounterBadge/CounterBadge.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var prop_types_1 = require("prop-types");
var wix_ui_theme_1 = require("wix-ui-theme");
var Badge_1 = require("wix-ui-core/Badge");
var theme_1 = require("./theme");
var constants_1 = require("./constants");
var maxContentLength = 2;
var counterStyle = {
    fontFamily: 'HelveticaNeueW01-65Medi, HelveticaNeueW02-65Medi, HelveticaNeueW10-65Medi, Helvetica Neue, Helvetica, Arial, sans-serif',
    fontSize: '10px',
    lineHeight: '12px',
    letterSpacing: '1px',
    margin: '0 -1px 0 0'
};
var CounterBadge = /** @class */ (function (_super) {
    __extends(CounterBadge, _super);
    function CounterBadge() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    CounterBadge.prototype.render = function () {
        var _a = this.props, skin = _a.skin, children = _a.children, coreProps = __rest(_a, ["skin", "children"]);
        var content = (React.Children.map(children, function (child) { return child; }))[0];
        var isIcon = typeof content !== 'string' && typeof content !== 'number';
        var canGrow = false;
        if (!isIcon) {
            var contentLength = content.toString().length;
            if (contentLength > maxContentLength) {
                throw new Error("CounterBadge children max length can not be more than " + maxContentLength);
            }
            if (contentLength === maxContentLength) {
                canGrow = true;
            }
        }
        return (React.createElement(wix_ui_theme_1.ThemedComponent, __assign({}, { theme: theme_1.theme, skin: skin, canGrow: canGrow }),
            React.createElement(Badge_1.Badge, __assign({}, coreProps), isIcon ?
                content :
                React.createElement("span", { style: counterStyle, "data-class": "badge-content" }, content))));
    };
    CounterBadge.displayName = 'CounterBadge';
    CounterBadge.propTypes = {
        /** Type of the badge */
        skin: prop_types_1.oneOf(['default', 'standard', 'urgent', 'success']),
        /** Content of the badge */
        children: prop_types_1.node
    };
    CounterBadge.defaultProps = {
        skin: constants_1.Skin.DEFAULT,
        children: ''
    };
    return CounterBadge;
}(React.PureComponent));
exports.CounterBadge = CounterBadge;
