// wix-style-react/dist/src/Tag/Tag.driver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _testUtils = require('react-dom/test-utils');

var _testUtils2 = _interopRequireDefault(_testUtils);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _utils = require('../../test/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var tagDriverFactory = function tagDriverFactory(_ref) {
  var element = _ref.element,
      wrapper = _ref.wrapper,
      component = _ref.component;


  var removeButton = element.querySelector('a');
  var thumb = element.querySelector('span');
  var contentWithoutThumb = element.querySelector('span');

  return {
    exists: function exists() {
      return !!element;
    },
    isLarge: function isLarge() {
      return (0, _utils.isClassExists)(element, 'large');
    },
    isStandardTheme: function isStandardTheme() {
      return (0, _utils.isClassExists)(element, 'standardTheme');
    },
    isWarningTheme: function isWarningTheme() {
      return (0, _utils.isClassExists)(element, 'warningTheme');
    },
    isErrorTheme: function isErrorTheme() {
      return (0, _utils.isClassExists)(element, 'errorTheme');
    },
    isRemovable: function isRemovable() {
      return (0, _utils.isClassExists)(removeButton, 'tagRemoveButton');
    },
    removeTag: function removeTag() {
      return _testUtils2.default.Simulate.click(removeButton);
    },
    click: function click() {
      return _testUtils2.default.Simulate.click(element);
    },
    isThumbExists: function isThumbExists() {
      return (0, _utils.isClassExists)(thumb, 'thumb');
    },
    isWrapped: function isWrapped() {
      return (0, _utils.isClassExists)(element, 'tagWrap') && (0, _utils.isClassExists)(contentWithoutThumb, 'innerTagWrap');
    },
    isDisabled: function isDisabled() {
      return (0, _utils.isClassExists)(element, 'disabled');
    },
    getLabel: function getLabel() {
      return element.textContent;
    },
    getTitle: function getTitle() {
      return element.title;
    },
    setProps: function setProps(props) {
      var ClonedWithProps = _react2.default.cloneElement.apply(_react2.default, [component, Object.assign({}, component.props, props)].concat(_toConsumableArray(component.props.children || [])));
      _reactDom2.default.render(_react2.default.createElement(
        'div',
        { ref: function ref(r) {
            return element = r;
          } },
        ClonedWithProps
      ), wrapper);
    }
  };
};

exports.default = tagDriverFactory;
