// wix-style-react/dist/stories/StatsWidget/ExampleStatsWidgetWithFilters.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _StatsWidget = require('../../src/StatsWidget');

var _StatsWidget2 = _interopRequireDefault(_StatsWidget);

var _ExampleStatsWidget = require('./ExampleStatsWidget.scss');

var _ExampleStatsWidget2 = _interopRequireDefault(_ExampleStatsWidget);

var _ButtonWithOptions = require('../../src/ButtonWithOptions');

var _ButtonWithOptions2 = _interopRequireDefault(_ButtonWithOptions);

var _Grid = require('../../src/Grid');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var statistics = [{
  title: '$10',
  subtitle: 'Revenue'
}, {
  title: '2',
  subtitle: 'Products'
}, {
  title: '1',
  subtitle: 'Transactions'
}, {
  title: '$5',
  subtitle: 'Profit'
}];

var dropdownOption = [{ id: 0, value: 'This month' }, { id: 1, value: 'This week' }];

var onFilterChange = function onFilterChange() {
  alert('hi');
};

var optionsArray = dropdownOption.map(function (option) {
  var value = option.value,
      props = _objectWithoutProperties(option, ['value']);

  return _react2.default.createElement(
    _ButtonWithOptions2.default.Option,
    _extends({ key: option.id }, props),
    value
  );
});

var ButtonWithOptionsProps = {
  theme: 'no-border',
  dataHook: 'StatsWidgetFilter',
  selectedId: 1
};

exports.default = function () {
  return _react2.default.createElement(
    _Grid.Container,
    null,
    _react2.default.createElement(
      'div',
      { 'data-hook': 'card-example', className: _ExampleStatsWidget2.default.statsWidgetWrapper },
      _react2.default.createElement(
        _StatsWidget2.default,
        { title: 'Let\'s see what\'s going on with your store', statistics: statistics },
        _react2.default.createElement(
          _StatsWidget2.default.Filter,
          _extends({ selectedId: 1, onSelect: onFilterChange }, ButtonWithOptionsProps),
          _react2.default.createElement(_ButtonWithOptions2.default.Button, null),
          optionsArray
        ),
        _react2.default.createElement(
          _StatsWidget2.default.Filter,
          _extends({ selectedId: 1, onSelect: onFilterChange }, ButtonWithOptionsProps),
          _react2.default.createElement(_ButtonWithOptions2.default.Button, null),
          optionsArray
        )
      )
    )
  );
};
