// wix-style-react/dist/stories/DataTable/ExampleWithAnimatedRowDetails.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _DataTable = require('wix-style-react/DataTable');

var _DataTable2 = _interopRequireDefault(_DataTable);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

require('./Example.scss');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var style = {
  width: '966px'
};

var baseData = [{ firstName: 'Meghan', lastName: 'Bishop' }, { firstName: 'Sara', lastName: 'Porter' }, { firstName: 'Deborah', lastName: 'Rhodes' }, { firstName: 'Walter', lastName: 'Jenning' }];

var generateData = function generateData() {
  var data = [];
  for (var i = 0; i < 2; i++) {
    data = data.concat(baseData);
  }
  return data;
};

var MyRowDetailsComponent = function MyRowDetailsComponent(props) {
  return _react2.default.createElement(
    'div',
    { style: { padding: '9px' } },
    _react2.default.createElement(
      'h2',
      null,
      'User Details'
    ),
    _react2.default.createElement(
      'p',
      null,
      'First name: ',
      props.firstName
    ),
    _react2.default.createElement(
      'p',
      null,
      'Last name: ',
      props.lastName
    )
  );
};

MyRowDetailsComponent.propTypes = {
  firstName: _propTypes2.default.string.isRequired,
  lastName: _propTypes2.default.string.isRequired
};

var ExampleWithAnimatedRowDetails = function (_React$Component) {
  _inherits(ExampleWithAnimatedRowDetails, _React$Component);

  function ExampleWithAnimatedRowDetails() {
    _classCallCheck(this, ExampleWithAnimatedRowDetails);

    return _possibleConstructorReturn(this, (ExampleWithAnimatedRowDetails.__proto__ || Object.getPrototypeOf(ExampleWithAnimatedRowDetails)).apply(this, arguments));
  }

  _createClass(ExampleWithAnimatedRowDetails, [{
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        'div',
        { style: style },
        _react2.default.createElement(_DataTable2.default, {
          dataHook: 'story-data-table',
          data: generateData(),
          rowDetails: function rowDetails(row) {
            return _react2.default.createElement(MyRowDetailsComponent, row);
          },
          rowDetailsAnimation: true,
          newDesign: true,
          allowMultiDetailsExpansion: true,
          columns: [{ title: 'Row Number', render: function render(row, rowNum) {
              return '#' + (rowNum + 1);
            }, width: '20%', minWidth: '75px', important: true }, { title: 'First Name', render: function render(row) {
              return _react2.default.createElement(
                'span',
                null,
                row.firstName
              );
            }, width: '40%', minWidth: '100px' }, { title: 'Last Name', render: function render(row) {
              return _react2.default.createElement(
                'span',
                null,
                row.lastName
              );
            }, width: '40%', minWidth: '100px' }]
        })
      );
    }
  }]);

  return ExampleWithAnimatedRowDetails;
}(_react2.default.Component);

exports.default = ExampleWithAnimatedRowDetails;
