// wix-ui-backoffice/dist/src/testkit/enzyme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var enzyme_1 = require("wix-ui-test-utils/enzyme");
var CounterBadge_driver_1 = require("../components/CounterBadge/CounterBadge.driver");
exports.counterBadgeTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(CounterBadge_driver_1.counterBadgeDriverFactory);
var Badge_driver_1 = require("../components/Badge/Badge.driver");
exports.badgeTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Badge_driver_1.badgeDriverFactory);
var LinearProgressBar_driver_1 = require("../components/LinearProgressBar/LinearProgressBar.driver");
exports.linearProgressBarTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(LinearProgressBar_driver_1.linearProgressBarDriverFactory);
var CircularProgressBar_driver_1 = require("../components/CircularProgressBar/CircularProgressBar.driver");
exports.circularProgressBarTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(CircularProgressBar_driver_1.circularProgressBarDriverFactory);
var Heading_driver_1 = require("../components/Heading/Heading.driver");
exports.headingTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Heading_driver_1.headingDriverFactory);
var Button_driver_1 = require("../components/Button/Button.driver");
exports.buttonTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Button_driver_1.buttonDriverFactory);
var CloseButton_driver_1 = require("../components/CloseButton/CloseButton.driver");
exports.closeButtonTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(CloseButton_driver_1.closeButtonDriverFactory);
var Checkbox_driver_1 = require("../components/Checkbox/Checkbox.driver");
exports.checkboxTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Checkbox_driver_1.checkboxDriverFactory);
var Text_driver_1 = require("../components/core/CoreText/Text.driver");
exports.coreTextTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Text_driver_1.textDriverFactory);
var UIText_driver_1 = require("../components/StylableUIText/UIText.driver");
exports.uiTextTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(UIText_driver_1.uiTextDriverFactory);
var Text_driver_2 = require("../components/Text/Text.driver");
exports.textTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Text_driver_2.textDriverFactory);
var FullTextView_driver_1 = require("../components/FullTextView/FullTextView.driver");
exports.fullTextViewTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(FullTextView_driver_1.fullTextViewDriverFactory);
var Autocomplete_driver_1 = require("../components/Autocomplete/Autocomplete.driver");
exports.autocompleteTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Autocomplete_driver_1.autocompleteDriverFactory);
var ToggleSwitch_driver_1 = require("../components/ToggleSwitch/ToggleSwitch.driver");
exports.toggleSwitchTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(ToggleSwitch_driver_1.toggleSwitchDriverFactory);
var CounterBadge_driver_2 = require("../components/StylableCounterBadge/CounterBadge.driver");
exports.stylableCounterBadgeTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(CounterBadge_driver_2.counterBadgeDriverFactory);
var LabelWithOptions_driver_1 = require("../components/LabelWithOptions/LabelWithOptions.driver");
exports.labelWithOptionsTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(LabelWithOptions_driver_1.labelWithOptionsDriverFactory);
var Label_driver_1 = require("../components/Label/Label.driver");
exports.labelTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Label_driver_1.labelDriverFactory);
var FloatingHelper_driver_1 = require("../components/FloatingHelper/FloatingHelper.driver");
exports.floatingHelperTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(FloatingHelper_driver_1.floatingHelperDriverFactory);
var Thumbnail_driver_1 = require("../components/Thumbnail/Thumbnail.driver");
exports.thumbnailTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Thumbnail_driver_1.thumbnailDriverFactory);
var AddressInput_driver_1 = require("../components/AddressInput/AddressInput.driver");
exports.addressInputTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(AddressInput_driver_1.addressInputDriverFactory);
var TimePicker_driver_1 = require("../components/TimePicker/TimePicker.driver");
exports.timePickerTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(TimePicker_driver_1.timePickerDriverFactory);
