// wix-ui-backoffice/dist/src/testkit/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var vanilla_1 = require("wix-ui-test-utils/vanilla");
var CounterBadge_driver_1 = require("../components/CounterBadge/CounterBadge.driver");
exports.counterBadgeTestkitFactory = vanilla_1.testkitFactoryCreator(CounterBadge_driver_1.counterBadgeDriverFactory);
var Badge_driver_1 = require("../components/Badge/Badge.driver");
exports.badgeTestkitFactory = vanilla_1.testkitFactoryCreator(Badge_driver_1.badgeDriverFactory);
var LinearProgressBar_driver_1 = require("../components/LinearProgressBar/LinearProgressBar.driver");
exports.linearProgressBarTestkitFactory = vanilla_1.testkitFactoryCreator(LinearProgressBar_driver_1.linearProgressBarDriverFactory);
var CircularProgressBar_driver_1 = require("../components/CircularProgressBar/CircularProgressBar.driver");
exports.circularProgressBarTestkitFactory = vanilla_1.testkitFactoryCreator(CircularProgressBar_driver_1.circularProgressBarDriverFactory);
var Heading_driver_1 = require("../components/Heading/Heading.driver");
exports.headingTestkitFactory = vanilla_1.testkitFactoryCreator(Heading_driver_1.headingDriverFactory);
var Button_driver_1 = require("../components/Button/Button.driver");
exports.buttonTestkitFactory = vanilla_1.testkitFactoryCreator(Button_driver_1.buttonDriverFactory);
var CloseButton_driver_1 = require("../components/CloseButton/CloseButton.driver");
exports.closeButtonTestkitFactory = vanilla_1.testkitFactoryCreator(CloseButton_driver_1.closeButtonDriverFactory);
var Checkbox_driver_1 = require("../components/Checkbox/Checkbox.driver");
exports.checkboxTestkitFactory = vanilla_1.testkitFactoryCreator(Checkbox_driver_1.checkboxDriverFactory);
var Text_driver_1 = require("../components/core/CoreText/Text.driver");
exports.coreTextTestkitFactory = vanilla_1.testkitFactoryCreator(Text_driver_1.textDriverFactory);
var UIText_driver_1 = require("../components/StylableUIText/UIText.driver");
exports.uiTextTestkitFactory = vanilla_1.testkitFactoryCreator(UIText_driver_1.uiTextDriverFactory);
var Text_driver_2 = require("../components/Text/Text.driver");
exports.textTestkitFactory = vanilla_1.testkitFactoryCreator(Text_driver_2.textDriverFactory);
var FullTextView_driver_1 = require("../components/FullTextView/FullTextView.driver");
exports.fullTextViewTestkitFactory = vanilla_1.testkitFactoryCreator(FullTextView_driver_1.fullTextViewDriverFactory);
var Autocomplete_driver_1 = require("../components/Autocomplete/Autocomplete.driver");
exports.autocompleteTestkitFactory = vanilla_1.testkitFactoryCreator(Autocomplete_driver_1.autocompleteDriverFactory);
var ToggleSwitch_driver_1 = require("../components/ToggleSwitch/ToggleSwitch.driver");
exports.toggleSwitchTestkitFactory = vanilla_1.testkitFactoryCreator(ToggleSwitch_driver_1.toggleSwitchDriverFactory);
var CounterBadge_driver_2 = require("../components/StylableCounterBadge/CounterBadge.driver");
exports.stylableCounterBadgeTestkitFactory = vanilla_1.testkitFactoryCreator(CounterBadge_driver_2.counterBadgeDriverFactory);
var LabelWithOptions_driver_1 = require("../components/LabelWithOptions/LabelWithOptions.driver");
exports.labelWithOptionsTestkitFactory = vanilla_1.testkitFactoryCreator(LabelWithOptions_driver_1.labelWithOptionsDriverFactory);
var Label_driver_1 = require("../components/Label/Label.driver");
exports.labelTestkitFactory = vanilla_1.testkitFactoryCreator(Label_driver_1.labelDriverFactory);
var FloatingHelper_driver_1 = require("../components/FloatingHelper/FloatingHelper.driver");
exports.floatingHelperTestkitFactory = vanilla_1.testkitFactoryCreator(FloatingHelper_driver_1.floatingHelperDriverFactory);
var Thumbnail_driver_1 = require("../components/Thumbnail/Thumbnail.driver");
exports.thumbnailTestkitFactory = vanilla_1.testkitFactoryCreator(Thumbnail_driver_1.thumbnailDriverFactory);
var AddressInput_driver_1 = require("../components/AddressInput/AddressInput.driver");
exports.addressInputTestkitFactory = vanilla_1.testkitFactoryCreator(AddressInput_driver_1.addressInputDriverFactory);
var TimePicker_driver_1 = require("../components/TimePicker/TimePicker.driver");
exports.timePickerTestkitFactory = vanilla_1.testkitFactoryCreator(TimePicker_driver_1.timePickerDriverFactory);
