// wix-style-react/dist/src/themes/answers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.button = undefined;

var _utils = require('./utils');

var pallete = {
  main: '#3899ec', //B10
  mainHover: '#4eb7f5', //B20
  dividers: '#eaf7ff', //B50
  danger: '#d8504c', //R05
  dangerHover: '#ee5951', //R10
  white: '#ffffff', //D80
  attention: '#fb7d33',
  attentionHover: '#ff9a48'
};

var heightMap = {
  small: '30px',
  medium: '36px',
  large: '42px'
};

var paddingMap = {
  small: '0 17px',
  medium: '0 23px',
  large: '0 29px'
};

var fontSizeMap = {
  small: '14px',
  medium: '16px',
  large: '20px'
};

var lineHeightMap = {
  small: '18px',
  medium: '24px',
  large: '24px'
};

var skinToColorMap = {
  standard: pallete.white,
  danger: pallete.white,
  attention: pallete.white
};

var skinToBackgroundMap = {
  standard: pallete.main,
  danger: pallete.danger,
  attention: pallete.attention
};

var skinToBorderColorMap = {
  standard: pallete.main,
  danger: pallete.danger,
  attention: pallete.attention
};

var skinToHoverColorMap = {
  standard: pallete.white,
  danger: pallete.white,
  attention: pallete.white
};

var skinToHoverBackgroundMap = {
  standard: pallete.mainHover,
  danger: pallete.dangerHover,
  attention: pallete.attentionHover
};

var skinToHoverBorderColorMap = {
  standard: pallete.mainHover,
  danger: pallete.dangerHover,
  attention: pallete.attentionHover
};

var button = {
  fontSize: function fontSize(_ref) {
    var height = _ref.height;
    return fontSizeMap[height];
  },
  lineHeight: function lineHeight(_ref2) {
    var height = _ref2.height;
    return lineHeightMap[height];
  },

  height: function height(_ref3) {
    var _height = _ref3.height;
    return heightMap[_height];
  },
  padding: function padding(_ref4) {
    var height = _ref4.height;
    return paddingMap[height];
  },
  borderRadius: function borderRadius(_ref5) {
    var height = _ref5.height;
    return (0, _utils.pxDivide)(heightMap[height], 2);
  },

  color: function color(_ref6) {
    var skin = _ref6.skin;
    return skinToColorMap[skin];
  },
  backgroundColor: function backgroundColor(_ref7) {
    var skin = _ref7.skin;
    return skinToBackgroundMap[skin];
  },
  borderColor: function borderColor(_ref8) {
    var skin = _ref8.skin;
    return skinToBorderColorMap[skin];
  },

  hover: {
    color: function color(_ref9) {
      var skin = _ref9.skin;
      return skinToHoverColorMap[skin];
    },
    backgroundColor: function backgroundColor(_ref10) {
      var skin = _ref10.skin;
      return skinToHoverBackgroundMap[skin];
    },
    borderColor: function borderColor(_ref11) {
      var skin = _ref11.skin;
      return skinToHoverBorderColorMap[skin];
    }
  }
};

exports.button = button;
exports.default = {
  button: button
};
