// wix-style-react/dist/src/themes/backoffice.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.input = exports.button = undefined;

var _utils = require('./utils');

var pallete = {
  main: '#3899ec', //B10
  mainHover: '#4eb7f5', //B20
  notifications: '#c1e4fe', //B30
  dividers: '#eaf7ff', //B50
  danger: '#ee5951', //R10
  dangerHover: '#ff6666', //R20
  mainInputText: '#162d3d',
  disabled: '#cbd3dc', //D55
  white: '#ffffff' //D80
};

var heightMap = {
  small: '30px',
  medium: '36px',
  large: '42px'
};

var paddingMap = {
  small: '0 17px',
  medium: '0 23px',
  large: '0 29px'
};

var fontSizeMap = {
  small: '14px',
  medium: '16px',
  large: '20px'
};

var lineHeightMap = {
  small: '18px',
  medium: '24px',
  large: '24px'
};

var skinToColorMap = {
  standard: pallete.white,
  danger: pallete.white,
  emptyStandard: pallete.main
};

var skinToBackgroundMap = {
  standard: pallete.main,
  danger: pallete.danger,
  emptyStandard: pallete.white
};

var skinToBorderColorMap = {
  standard: pallete.main,
  danger: pallete.danger,
  emptyStandard: pallete.main
};

var skinToHoverColorMap = {
  standard: pallete.white,
  danger: pallete.white,
  emptyStandard: pallete.white
};

var skinToHoverBackgroundMap = {
  standard: pallete.mainHover,
  danger: pallete.dangerHover,
  emptyStandard: pallete.mainHover
};

var skinToHoverBorderColorMap = {
  standard: pallete.mainHover,
  danger: pallete.dangerHover,
  emptyStandard: pallete.mainHover
};

var button = {
  fontSize: function fontSize(_ref) {
    var height = _ref.height;
    return fontSizeMap[height];
  },
  lineHeight: function lineHeight(_ref2) {
    var height = _ref2.height;
    return lineHeightMap[height];
  },

  height: function height(_ref3) {
    var _height = _ref3.height;
    return heightMap[_height];
  },
  padding: function padding(_ref4) {
    var height = _ref4.height;
    return paddingMap[height];
  },
  borderRadius: function borderRadius(_ref5) {
    var height = _ref5.height;
    return (0, _utils.pxDivide)(heightMap[height], 2);
  },

  color: function color(_ref6) {
    var skin = _ref6.skin;
    return skinToColorMap[skin];
  },
  backgroundColor: function backgroundColor(_ref7) {
    var skin = _ref7.skin;
    return skinToBackgroundMap[skin];
  },
  borderColor: function borderColor(_ref8) {
    var skin = _ref8.skin;
    return skinToBorderColorMap[skin];
  },

  hover: {
    color: function color(_ref9) {
      var skin = _ref9.skin;
      return skinToHoverColorMap[skin];
    },
    backgroundColor: function backgroundColor(_ref10) {
      var skin = _ref10.skin;
      return skinToHoverBackgroundMap[skin];
    },
    borderColor: function borderColor(_ref11) {
      var skin = _ref11.skin;
      return skinToHoverBorderColorMap[skin];
    }
  },

  disabled: {
    color: 'black',
    backgroundColor: pallete.disabled,
    borderColor: pallete.disabled
  }
};

exports.button = button;
var input = exports.input = {
  color: pallete.mainInputText,
  borderColor: pallete.notifications,
  backgroundColor: pallete.white,
  borderRadius: function borderRadius(_ref12) {
    var roundInput = _ref12.roundInput;
    return roundInput ? '18px' : '6px';
  },
  fontSize: '16px',
  padding: '0 12px',
  height: '30px',

  hover: {
    backgroundColor: pallete.dividers
  },

  focus: {
    boxShadow: 'none',
    borderColor: pallete.mainHover
  }
};

exports.default = {
  button: button,
  input: input
};
