// wix-animations/dist/stories/Animations/Sequence/ExampleSameTransition.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Animator = require('../../../src/components/Animator');

var _Animator2 = _interopRequireDefault(_Animator);

var _Example = require('../Example.scss');

var css = _interopRequireWildcard(_Example);

var _AnimationTemplate = require('../AnimationTemplate');

var _AnimationTemplate2 = _interopRequireDefault(_AnimationTemplate);

var _StyledDiv = require('../StyledDiv');

var _StyledDiv2 = _interopRequireDefault(_StyledDiv);

var _OrangeDiv = require('../OrangeDiv');

var _OrangeDiv2 = _interopRequireDefault(_OrangeDiv);

var _Dropdown = require('wix-style-react/dist/src/Dropdown');

var _Dropdown2 = _interopRequireDefault(_Dropdown);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ExampleMixedChildren = function (_React$Component) {
  _inherits(ExampleMixedChildren, _React$Component);

  function ExampleMixedChildren(props) {
    _classCallCheck(this, ExampleMixedChildren);

    var _this = _possibleConstructorReturn(this, (ExampleMixedChildren.__proto__ || Object.getPrototypeOf(ExampleMixedChildren)).call(this, props));

    _this.state = {
      useSameTransition: true
    };

    _this.options = [{ id: true, value: '{useSameTransition: true}' }, { id: false, value: '{useSameTransition: false}' }];
    return _this;
  }

  _createClass(ExampleMixedChildren, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var show = this.props.show;

      var directionIn = show ? 'left' : 'right';
      var directionOut = show ? 'right' : 'left';

      return _react2.default.createElement(
        'div',
        { style: { display: 'flex' } },
        _react2.default.createElement(
          'div',
          { style: { flexGrow: 1 } },
          _react2.default.createElement(_Dropdown2.default, {
            width: '150px',
            selectedId: this.state.useSameTransition,
            onSelect: function onSelect(option) {
              return _this2.setState({ useSameTransition: option.id });
            },
            options: this.options
          })
        ),
        _react2.default.createElement(
          'div',
          { style: { overflow: 'hidden', height: '100px', width: '740px', position: 'relative' } },
          _react2.default.createElement(
            _Animator2.default,
            { childClassName: css.absolute, translate: { enter: { direction: directionIn, size: '100%' }, exit: { direction: directionOut, size: '100%' }, useSameTransition: this.state.useSameTransition } },
            show && _react2.default.createElement(
              _StyledDiv2.default,
              { style: { opacity: '0.5' } },
              'I am transparent so you can see the animation is smooth when using',
              _react2.default.createElement(
                'pre',
                { style: { margin: 0, display: 'inline' } },
                '{',
                'useSameTransition: true',
                '}'
              )
            ),
            !show && _react2.default.createElement(
              _OrangeDiv2.default,
              { style: { opacity: '0.5' } },
              'Check out the code to see how it is being done'
            )
          )
        )
      );
    }
  }]);

  return ExampleMixedChildren;
}(_react2.default.Component);

ExampleMixedChildren.propTypes = {
  show: _propTypes2.default.boolean
};

exports.default = function () {
  return _react2.default.createElement(
    _AnimationTemplate2.default,
    null,
    _react2.default.createElement(ExampleMixedChildren, null)
  );
};
