// wix-animations/dist/stories/Animations/Sequence/ExampleSequence.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Dropdown = require('wix-style-react/dist/src/Dropdown');

var _Dropdown2 = _interopRequireDefault(_Dropdown);

var _Animator = require('../../../src/components/Animator');

var _Animator2 = _interopRequireDefault(_Animator);

var _Example = require('../Example.scss');

var css = _interopRequireWildcard(_Example);

var _AnimationTemplate = require('../AnimationTemplate');

var _AnimationTemplate2 = _interopRequireDefault(_AnimationTemplate);

var _StyledDiv = require('../StyledDiv');

var _StyledDiv2 = _interopRequireDefault(_StyledDiv);

var _OrangeDiv = require('../OrangeDiv');

var _OrangeDiv2 = _interopRequireDefault(_OrangeDiv);

var _SequenceHelper = require('./SequenceHelper');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ExampleSequence = function (_React$Component) {
  _inherits(ExampleSequence, _React$Component);

  function ExampleSequence(props) {
    _classCallCheck(this, ExampleSequence);

    var _this = _possibleConstructorReturn(this, (ExampleSequence.__proto__ || Object.getPrototypeOf(ExampleSequence)).call(this, props));

    _this.state = {
      sequenceType: 'default'
    };

    _this.sequenceTypes = [{ id: 'default', value: 'default' }, { id: 'flip', value: 'flip' }, { id: 'reverse', value: 'reverse' }, { id: 'reverse-flip', value: 'reverse-flip' }];

    return _this;
  }

  _createClass(ExampleSequence, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var show = this.props.show;
      var type = this.state.sequenceType;

      return _react2.default.createElement(
        'div',
        { className: css.sequenceType, style: { display: 'flex', height: '200px' } },
        _react2.default.createElement(
          'div',
          { style: { marginRight: '30px' } },
          _react2.default.createElement(
            'div',
            { style: { marginBottom: '6px' } },
            'Choose Sequence Type'
          ),
          _react2.default.createElement(_Dropdown2.default, {
            width: '120px',
            selectedId: type,
            onSelect: function onSelect(option) {
              return _this2.setState({ sequenceType: option.id });
            },
            options: this.sequenceTypes
          })
        ),
        _react2.default.createElement(
          _Animator2.default,
          { show: show, sequence: this.state.sequenceType === 'default' ? true : this.state.sequenceType, height: true, className: css.sequenceExampleWrapper },
          _react2.default.createElement(
            _StyledDiv2.default,
            null,
            'We'
          ),
          _react2.default.createElement(
            _OrangeDiv2.default,
            null,
            'will'
          ),
          _react2.default.createElement(
            _StyledDiv2.default,
            null,
            'animate'
          ),
          _react2.default.createElement(
            _OrangeDiv2.default,
            null,
            'in'
          ),
          _react2.default.createElement(
            _StyledDiv2.default,
            null,
            'sequence'
          )
        ),
        _react2.default.createElement(_SequenceHelper.TextHelper, { type: type })
      );
    }
  }]);

  return ExampleSequence;
}(_react2.default.Component);

ExampleSequence.propTypes = {
  show: _propTypes2.default.bool
};

exports.default = function () {
  return _react2.default.createElement(
    _AnimationTemplate2.default,
    null,
    _react2.default.createElement(ExampleSequence, null)
  );
};
