// wix-style-react/dist/stories/Notification/ExampleStandard.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _ExampleStandard = require('./ExampleStandard.scss');

var _ExampleStandard2 = _interopRequireDefault(_ExampleStandard);

var _Notification = require('./Notification');

var _Notification2 = _interopRequireDefault(_Notification);

var _Notification3 = require('../../src/Notification');

var _Label = require('../../src/Label');

var _Label2 = _interopRequireDefault(_Label);

var _ToggleSwitch = require('../../src/ToggleSwitch');

var _ToggleSwitch2 = _interopRequireDefault(_ToggleSwitch);

var _RadioGroup = require('../../src/RadioGroup');

var _RadioGroup2 = _interopRequireDefault(_RadioGroup);

var _Input = require('../../src/Input');

var _Input2 = _interopRequireDefault(_Input);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ExampleStandard = function (_Component) {
  _inherits(ExampleStandard, _Component);

  function ExampleStandard() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ExampleStandard);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ExampleStandard.__proto__ || Object.getPrototypeOf(ExampleStandard)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      notification: {
        show: true,
        type: _Notification3.GLOBAL_NOTIFICATION,
        timeout: '',
        zIndex: 10000
      },
      actionButton: {
        type: 'button',
        text: 'Thanks',
        link: 'https://www.wix.com'
      }
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ExampleStandard, [{
    key: 'setComponentState',
    value: function setComponentState(componentName, obj) {
      var _this2 = this;

      this.setState(function (prevState) {
        prevState[componentName] = _extends({}, _this2.state[componentName], obj);
        Object.keys(prevState[componentName]).forEach(function (k) {
          return !prevState[componentName][k] && delete prevState[componentName][k];
        });
        return prevState;
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      var params = _extends({}, this.state);
      params.notification.theme = this.props.theme;
      return _react2.default.createElement(
        'form',
        { className: _ExampleStandard2.default.form },
        _react2.default.createElement(
          'div',
          { className: _ExampleStandard2.default.output },
          _react2.default.createElement(_Notification2.default, _extends({}, params, { onChange: this.props.onChange }))
        ),
        _react2.default.createElement(
          'div',
          { className: _ExampleStandard2.default.input },
          _react2.default.createElement(
            'div',
            { className: _ExampleStandard2.default.option },
            _react2.default.createElement(
              'div',
              { className: _ExampleStandard2.default.flex },
              _react2.default.createElement(
                _Label2.default,
                null,
                'This text will be covered by a local notification'
              )
            ),
            _react2.default.createElement('hr', null)
          ),
          _react2.default.createElement(
            'div',
            { className: _ExampleStandard2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Show Notification'
            ),
            _react2.default.createElement(
              'div',
              { className: _ExampleStandard2.default.flex },
              _react2.default.createElement(_ToggleSwitch2.default, {
                size: 'small',
                checked: this.state.notification.show,
                onChange: function onChange() {
                  return _this3.setComponentState('notification', { show: !_this3.state.notification.show });
                }
              })
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _ExampleStandard2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Type'
            ),
            _react2.default.createElement(
              'div',
              { className: _ExampleStandard2.default.flex },
              _react2.default.createElement(
                _RadioGroup2.default,
                {
                  display: 'horizontal',
                  value: this.state.notification.type,
                  onChange: function onChange(type) {
                    return _this3.setComponentState('notification', { type: type });
                  }
                },
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: _Notification3.GLOBAL_NOTIFICATION },
                  'Global (push the content)'
                ),
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: _Notification3.LOCAL_NOTIFICATION },
                  'Local (on top of the content)'
                ),
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: _Notification3.STICKY_NOTIFICATION },
                  'Sticky (at screen top)'
                )
              )
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _ExampleStandard2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Timeout in ms'
            ),
            _react2.default.createElement(
              'div',
              { className: _ExampleStandard2.default.column },
              _react2.default.createElement(_Input2.default, {
                placeholder: 'Set the timeout', size: 'small', type: 'number',
                value: this.state.notification.timeout,
                onChange: function onChange(e) {
                  return _this3.setComponentState('notification', { timeout: Number(e.target.value) });
                }
              })
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _ExampleStandard2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Button Type'
            ),
            _react2.default.createElement(
              'div',
              { className: _ExampleStandard2.default.flex },
              _react2.default.createElement(
                _RadioGroup2.default,
                {
                  display: 'horizontal',
                  value: this.state.actionButton.type,
                  onChange: function onChange(type) {
                    return _this3.setComponentState('actionButton', { type: type });
                  }
                },
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'button' },
                  'Button'
                ),
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'textLink' },
                  'TextLink'
                ),
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'none' },
                  'None'
                )
              )
            )
          ),
          this.state.actionButton.type === 'none' ? null : _react2.default.createElement(
            'div',
            null,
            this.state.actionButton.type !== 'textLink' ? null : _react2.default.createElement(
              'div',
              { className: _ExampleStandard2.default.option },
              _react2.default.createElement(
                _Label2.default,
                null,
                'Link'
              ),
              _react2.default.createElement(
                'div',
                { className: _ExampleStandard2.default.flex },
                _react2.default.createElement(_Input2.default, {
                  value: this.state.actionButton.link, size: 'small',
                  onChange: function onChange(event) {
                    return _this3.setComponentState('actionButton', { link: event.target.value });
                  }
                })
              )
            ),
            _react2.default.createElement(
              'div',
              { className: _ExampleStandard2.default.option },
              _react2.default.createElement(
                _Label2.default,
                null,
                'Text'
              ),
              _react2.default.createElement(
                'div',
                { className: _ExampleStandard2.default.flex },
                _react2.default.createElement(_Input2.default, {
                  value: this.state.actionButton.text, size: 'small',
                  onChange: function onChange(event) {
                    return _this3.setComponentState('actionButton', { text: event.target.value });
                  }
                })
              )
            )
          )
        ),
        _react2.default.createElement(
          'div',
          { className: _ExampleStandard2.default.option },
          _react2.default.createElement(
            _Label2.default,
            null,
            'z-index (optional)'
          ),
          _react2.default.createElement(
            'div',
            { className: _ExampleStandard2.default.column },
            _react2.default.createElement(_Input2.default, {
              placeholder: 'optional z-index', size: 'small', type: 'number',
              value: this.state.notification.zIndex,
              onChange: function onChange(e) {
                return _this3.setComponentState('notification', { zIndex: Number(e.target.value) });
              }
            })
          )
        )
      );
    }
  }]);

  return ExampleStandard;
}(_react.Component);

ExampleStandard.propTypes = {
  onChange: _propTypes2.default.func,
  theme: _propTypes2.default.string
};
exports.default = ExampleStandard;
