// wix-ui-backoffice/dist/src/components/LinearProgressBar/LinearProgressBar.spec.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var driver_factory_1 = require("wix-ui-test-utils/driver-factory");
var LinearProgressBar_driver_1 = require("./LinearProgressBar.driver");
var LinearProgressBar_1 = require("./LinearProgressBar");
var testkit_1 = require("../../testkit");
var enzyme_1 = require("../../testkit/enzyme");
var testkitTests_1 = require("../../common/testkitTests");
describe('LinearProgressBar', function () {
    var createDriver = driver_factory_1.createDriverFactory(LinearProgressBar_driver_1.linearProgressBarDriverFactory);
    var defaultProps = {
        value: 40
    };
    describe('on error', function () {
        var errorProps = {
            error: true,
            errorMessage: 'No soup for you',
            showProgressIndication: true
        };
        it('should display tooltip text only on hover', function () {
            var driver = createDriver(React.createElement(LinearProgressBar_1.LinearProgressBar, __assign({}, defaultProps, errorProps)));
            expect(driver.isTooltipShown()).toBe(false);
            driver.getTooltip().mouseEnter();
            expect(driver.isTooltipShown()).toBe(true);
            expect(driver.getTooltip().getContentElement().innerHTML).toContain(errorProps.errorMessage);
        });
        it('should display error icon', function () {
            var driver = createDriver(React.createElement(LinearProgressBar_1.LinearProgressBar, __assign({}, defaultProps, errorProps)));
            expect(driver.isErrorIconShown()).toBe(true);
        });
    });
    describe('on completion', function () {
        var successProps = {
            value: 100,
            showProgressIndication: true,
        };
        it('should display success icon', function () {
            var driver = createDriver(React.createElement(LinearProgressBar_1.LinearProgressBar, __assign({}, successProps)));
            expect(driver.isSuccessIconShown()).toBe(true);
        });
    });
    testkitTests_1.runTestkitExistsSuite({
        Element: React.createElement(LinearProgressBar_1.LinearProgressBar, __assign({}, defaultProps)),
        testkitFactory: testkit_1.linearProgressBarTestkitFactory,
        enzymeTestkitFactory: enzyme_1.linearProgressBarTestkitFactory
    });
});
