// wix-style-react/dist/stories/Breadcrumbs/index.story.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _CodeExample = require('wix-storybook-utils/CodeExample');

var _CodeExample2 = _interopRequireDefault(_CodeExample);

var _Breadcrumbs = require('wix-style-react/Breadcrumbs');

var _Breadcrumbs2 = _interopRequireDefault(_Breadcrumbs);

var _ExampleStandard = require('!raw-loader!../Breadcrumbs/ExampleStandard');

var _ExampleStandard2 = _interopRequireDefault(_ExampleStandard);

var _ExampleStandard3 = require('../Breadcrumbs/ExampleStandard');

var _ExampleStandard4 = _interopRequireDefault(_ExampleStandard3);

var _ExampleSizes = require('!raw-loader!../Breadcrumbs/ExampleSizes');

var _ExampleSizes2 = _interopRequireDefault(_ExampleSizes);

var _ExampleSizes3 = require('../Breadcrumbs/ExampleSizes');

var _ExampleSizes4 = _interopRequireDefault(_ExampleSizes3);

var _ExampleThemes = require('!raw-loader!../Breadcrumbs/ExampleThemes');

var _ExampleThemes2 = _interopRequireDefault(_ExampleThemes);

var _ExampleThemes3 = require('../Breadcrumbs/ExampleThemes');

var _ExampleThemes4 = _interopRequireDefault(_ExampleThemes3);

var _ExampleWithNodes = require('!raw-loader!../Breadcrumbs/ExampleWithNodes');

var _ExampleWithNodes2 = _interopRequireDefault(_ExampleWithNodes);

var _ExampleWithNodes3 = require('../Breadcrumbs/ExampleWithNodes');

var _ExampleWithNodes4 = _interopRequireDefault(_ExampleWithNodes3);

var _ExampleWithOnClick = require('!raw-loader!../Breadcrumbs/ExampleWithOnClick');

var _ExampleWithOnClick2 = _interopRequireDefault(_ExampleWithOnClick);

var _ExampleWithOnClick3 = require('../Breadcrumbs/ExampleWithOnClick');

var _ExampleWithOnClick4 = _interopRequireDefault(_ExampleWithOnClick3);

var _ExampleWithChosenActiveElement = require('!raw-loader!../Breadcrumbs/ExampleWithChosenActiveElement');

var _ExampleWithChosenActiveElement2 = _interopRequireDefault(_ExampleWithChosenActiveElement);

var _ExampleWithChosenActiveElement3 = require('../Breadcrumbs/ExampleWithChosenActiveElement');

var _ExampleWithChosenActiveElement4 = _interopRequireDefault(_ExampleWithChosenActiveElement3);

var _ExampleUsingURL = require('!raw-loader!../Breadcrumbs/ExampleUsingURL');

var _ExampleUsingURL2 = _interopRequireDefault(_ExampleUsingURL);

var _ExampleUsingURL3 = require('../Breadcrumbs/ExampleUsingURL');

var _ExampleUsingURL4 = _interopRequireDefault(_ExampleUsingURL3);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var items = [{
  id: '1',
  value: 'First item'
}, {
  id: '2',
  link: 'http://www.wix.com',
  value: 'Linked item'
}, {
  id: '3',
  value: 'Third item'
}];

exports.default = {
  category: '6. Navigation',
  storyName: '6.2 Breadcrumbs',
  component: _Breadcrumbs2.default,
  componentPath: './../../src/Breadcrumbs',
  componentProps: {
    activeId: '1',
    items: items,
    size: 'medium'
  },
  exampleProps: {
    activeId: ['1', '2', '3'],
    items: [{ label: 'One item', value: [{ id: 0, value: 'Homepage' }] }, { label: 'Three items', value: items }]
  },
  examples: _react2.default.createElement(
    'div',
    null,
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Standard', code: _ExampleStandard2.default },
      _react2.default.createElement(_ExampleStandard4.default, null)
    ),
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Sizes', code: _ExampleSizes2.default },
      _react2.default.createElement(_ExampleSizes4.default, null)
    ),
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Example using path factory', code: _ExampleUsingURL2.default },
      _react2.default.createElement(_ExampleUsingURL4.default, null)
    ),
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Themes', code: _ExampleThemes2.default },
      _react2.default.createElement(_ExampleThemes4.default, null)
    ),
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Html node items', code: _ExampleWithNodes2.default },
      _react2.default.createElement(_ExampleWithNodes4.default, null)
    ),
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'On click callback', code: _ExampleWithOnClick2.default },
      _react2.default.createElement(_ExampleWithOnClick4.default, null)
    ),
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Chosen active element', code: _ExampleWithChosenActiveElement2.default },
      _react2.default.createElement(_ExampleWithChosenActiveElement4.default, null)
    )
  )
};
