// wix-style-react/dist/stories/FloatingHelperContent/FloatingHelperContent.story.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _FloatingHelperContent = require('../../src/FloatingHelper/FloatingHelperContent');

var _FloatingHelperContent2 = _interopRequireDefault(_FloatingHelperContent);

var _StorySettings = require('./StorySettings');

var _Image = require('wix-ui-icons-common/Image');

var _Image2 = _interopRequireDefault(_Image);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

var image = React.createElement(_Image2.default, { width: '102', height: '102' });

var componentProps = {
  title: 'This is the title',
  body: 'This is the a long text which is passed in the `body` propterty',
  actionText: 'Ok, Take Me There!',
  onActionClick: function onActionClick() {
    return null;
  }
};

exports.default = {
  category: _StorySettings.storySettings.kind,
  storyName: _StorySettings.storySettings.story,
  component: _FloatingHelperContent2.default,
  componentPath: '../../node_modules/wix-ui-backoffice/src/components/FloatingHelper/FloatingHelperContent/FloatingHelperContent.tsx',
  displayName: 'FloatingHelper.Content',
  componentProps: componentProps,
  exampleProps: {
    onActionClick: function onActionClick() {/** see interactive preview*/},
    actionTheme: ['white', 'premium'],
    image: [{ label: 'with image', value: image }]
  },
  exampleImport: 'import ... - do not import directly, use FloatingHelper.Content'
};
