// wix-style-react/dist/src/CloseButton/CloseButton.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _Button = require('../Button');

var _Button2 = _interopRequireDefault(_Button);

var _Close = require('../new-icons/system/Close');

var _Close2 = _interopRequireDefault(_Close);

var _CloseLarge = require('../new-icons/system/CloseLarge');

var _CloseLarge2 = _interopRequireDefault(_CloseLarge);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
  For internal usage,
  TODO should be refactored later, together with buttons cleanup
*/

var CloseButton = function CloseButton(_ref) {
  var arialLabel = _ref.arialLabel,
      dataHook = _ref.dataHook,
      size = _ref.size,
      theme = _ref.theme,
      onClick = _ref.onClick;

  return _react2.default.createElement(
    _Button2.default,
    {
      'aria-label': arialLabel,
      dataHook: dataHook,
      height: size === 'small' ? 'medium' : 'large',
      theme: theme,
      onClick: onClick
    },
    size === 'small' ? _react2.default.createElement(_Close2.default, null) : _react2.default.createElement(_CloseLarge2.default, null)
  );
};

CloseButton.propTypes = {
  arialLabel: _propTypes.string,
  dataHook: _propTypes.string,
  size: (0, _propTypes.oneOf)(['small', 'large']),
  theme: _propTypes.string,
  onClick: _propTypes.func
};

CloseButton.defaultProps = {
  arialLabel: 'close button',
  theme: 'close-transparent',
  size: 'large'
};

CloseButton.displayName = 'CloseButton';

exports.default = CloseButton;
