// wix-ui-backoffice/dist/src/components/AddressInput/AddressInput.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var prop_types_1 = require("prop-types");
var AddressInput_1 = require("wix-ui-core/AddressInput");
var AddressInput_st_css_1 = require("./AddressInput.st.css");
var Location_1 = require("wix-ui-icons-common/Location");
var omit = require("lodash/omit");
var Search_1 = require("wix-ui-icons-common/Search");
var THROTTLE_INTERVAL = 150;
var excludePropsArray = ['forceContentElementVisibility', 'forceOptions', 'throttleInterval', 'locationIcon'];
var AddressInput = /** @class */ (function (_super) {
    __extends(AddressInput, _super);
    function AddressInput() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    AddressInput.prototype.focus = function () {
        this.addressInputRef.focus();
    };
    AddressInput.prototype.blur = function () {
        this.addressInputRef.blur();
    };
    AddressInput.prototype.render = function () {
        var _this = this;
        var _a = this.props, size = _a.size, showLocationIcon = _a.showLocationIcon, magnifyingGlass = _a.magnifyingGlass, rest = __rest(_a, ["size", "showLocationIcon", "magnifyingGlass"]);
        return (React.createElement(AddressInput_1.AddressInput, __assign({}, rest, AddressInput_st_css_1.default('root', { size: size }, rest), { forceContentElementVisibility: false, forceOptions: null, throttleInterval: THROTTLE_INTERVAL, locationIcon: showLocationIcon && React.createElement(Location_1.default, null), suffix: magnifyingGlass && React.createElement(Search_1.default, { className: AddressInput_st_css_1.default.search }), ref: function (ref) { return _this.addressInputRef = ref; } })));
    };
    AddressInput.displayName = 'AddressInput';
    AddressInput.propTypes = __assign({}, omit(AddressInput_1.AddressInput.propTypes, excludePropsArray), { size: prop_types_1.oneOf(['large', 'medium', 'small']), showLocationIcon: prop_types_1.bool, magnifyingGlass: prop_types_1.bool });
    AddressInput.defaultProps = {
        size: 'medium',
        magnifyingGlass: true
    };
    return AddressInput;
}(React.PureComponent));
exports.AddressInput = AddressInput;
