// wix-style-react/dist/src/RadioGroup/RadioGroup.protractor.driver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _protractor = require('wix-ui-test-utils/protractor');

var _RadioButtonProtractor = require('./RadioButton/RadioButton.protractor.driver');

var _RadioButtonProtractor2 = _interopRequireDefault(_RadioButtonProtractor);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var radioGroupDriverFactory = function radioGroupDriverFactory(component) {
  var getRadioButtonLabel = function getRadioButtonLabel(index) {
    return component.$$('div [data-hook="radio-label"]').get(index);
  };
  var getRadioButtonRoot = function getRadioButtonRoot(index) {
    return component.all(by.xpath('./div')).get(index);
  };

  return {
    getButtonDriver: function getButtonDriver(index) {
      return (0, _RadioButtonProtractor2.default)(getRadioButtonRoot(index));
    },
    getRadioAtIndex: function getRadioAtIndex(index) {
      return getRadioButtonLabel(index);
    },
    selectByIndex: function selectByIndex(index) {
      return getRadioButtonLabel(index).click();
    },
    isRadioChecked: function isRadioChecked(index) {
      return component.$$('div input').get(index).isSelected();
    },
    isRadioDisabled: function isRadioDisabled(index) {
      return !!component.$$('div input').get(index).getAttribute('disabled');
    },
    /**
     * @deprecated
     * @see getButtonDriver
     */
    isRadioFocused: function isRadioFocused(index) {
      return (0, _protractor.isFocused)(component.$$('div label').get(index).$('[tabindex="0"]'));
    },
    element: function element() {
      return component;
    }
  };
};

exports.default = radioGroupDriverFactory;
