// wix-style-react/dist/stories/ImageViewer/ExampleStandard.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _ImageViewer = require('../../src/ImageViewer');

var _ImageViewer2 = _interopRequireDefault(_ImageViewer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var style = {
  display: 'inline-block',
  verticalAlign: 'middle',
  padding: '0 25px 25px'
};

var Example = function Example() {
  return _react2.default.createElement(
    'div',
    null,
    _react2.default.createElement(
      'div',
      { className: 'ltr', style: style },
      'Without image',
      _react2.default.createElement('br', null),
      _react2.default.createElement('br', null),
      _react2.default.createElement(_ImageViewer2.default, { dataHook: 'empty-image-viewer', onAddImage: function onAddImage() {} })
    ),
    _react2.default.createElement(
      'div',
      { className: 'ltr', style: style },
      'With image',
      _react2.default.createElement('br', null),
      _react2.default.createElement('br', null),
      _react2.default.createElement(_ImageViewer2.default, { onRemoveImage: function onRemoveImage() {}, onUpdateImage: function onUpdateImage() {}, imageUrl: 'https://upload.wikimedia.org/wikipedia/commons/d/dd/New_Mela_Ramanputhur_Holy_Family_Church.jpg' })
    ),
    _react2.default.createElement(
      'div',
      { className: 'ltr', style: style },
      'With error',
      _react2.default.createElement('br', null),
      _react2.default.createElement('br', null),
      _react2.default.createElement(_ImageViewer2.default, { dataHook: 'empty-image-viewer', error: true, errorMessage: 'Just a demo', onAddImage: function onAddImage() {} })
    )
  );
};

Example.propTypes = {
  theme: _propTypes2.default.string
};

exports.default = Example;
