// wix-style-react/dist/stories/Search/index.story.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Search = require('wix-style-react/Search');

var _Search2 = _interopRequireDefault(_Search);

var _CodeExample = require('wix-storybook-utils/CodeExample');

var _CodeExample2 = _interopRequireDefault(_CodeExample);

var _expandable = require('!raw-loader!./expandable.example');

var _expandable2 = _interopRequireDefault(_expandable);

var _expandable3 = require('./expandable.example');

var _expandable4 = _interopRequireDefault(_expandable3);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var createOption = function createOption(value) {
  var id = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;
  return { id: id, value: value };
};

var options = ['The quick', 'brown', 'fox', 'jumps over', 'the lazy', 'dog', 'Option1', 'Option2', 'Option3', 'Option4', 'Option5', 'last Option'].map(createOption);

var settings = {
  category: '3. Inputs',
  storyName: '3.9 Search',
  dataHook: 'storybook-search',
  options: options
};

exports.default = {
  category: settings.category,
  storyName: settings.storyName,
  component: _Search2.default,
  componentPath: '../../src/Search',

  componentProps: function componentProps(setState) {
    return {
      dataHook: settings.dataHook,
      value: '',
      options: settings.options,
      showOptionsIfEmptyInput: false,
      closeOnSelect: false,

      onChange: function onChange(e) {
        return setState({ value: e.target.value });
      },

      onSelect: function onSelect(option) {
        return setState({ value: option.value });
      }
    };
  },

  exampleProps: {
    onSelect: function onSelect(option) {
      return option.value;
    },
    onChange: function onChange(e) {
      return e.target.value;
    },
    options: [{ label: 'One option', value: [createOption('Just me :)')] }, { label: options.length + ' options', value: options }]
  },

  examples: _react2.default.createElement(
    'div',
    null,
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Expandable', code: _expandable2.default },
      _react2.default.createElement(_expandable4.default, null)
    )
  )
};
