// wix-style-react/dist/src/DataTable/DataTable.protractor.driver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var rowSelector = 'tbody tr';
var rowByIdx = function rowByIdx(component, index) {
  return component.$$(rowSelector).get(index);
};
var scrollIntoView = function scrollIntoView(el) {
  return browser.executeScript(function (element) {
    element.scrollIntoView();
  }, el.getWebElement());
};

var dataTableDriverFactory = function dataTableDriverFactory(component) {
  return {
    rowsCount: function rowsCount() {
      return component.$$(rowSelector).count();
    },
    clickRowByIndex: function clickRowByIndex(index) {
      return rowByIdx(component, index).click();
    },
    getRowTextByIndex: function getRowTextByIndex(index) {
      return rowByIdx(component, index).getText();
    },
    scrollToRowByIdx: function scrollToRowByIdx(index) {
      return scrollIntoView(rowByIdx(component, index));
    },
    element: function element() {
      return component;
    }
  };
};

exports.default = dataTableDriverFactory;
