// wix-style-react/dist/stories/FloatingHelper/ControlledExample.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ControlledExample = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _FloatingHelper = require('wix-style-react/FloatingHelper');

var _FloatingHelper2 = _interopRequireDefault(_FloatingHelper);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ControlledExample = exports.ControlledExample = function (_React$Component) {
  _inherits(ControlledExample, _React$Component);

  function ControlledExample() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ControlledExample);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ControlledExample.__proto__ || Object.getPrototypeOf(ControlledExample)).call.apply(_ref, [this].concat(args))), _this), _this.state = { open: true }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ControlledExample, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      return React.createElement(
        'div',
        null,
        React.createElement(
          'div',
          null,
          React.createElement(
            'button',
            { onClick: function onClick() {
                return _this2.setState({ open: true });
              } },
            'Click to open'
          ),
          React.createElement('br', null),
          React.createElement(
            'button',
            { onClick: function onClick() {
                return _this2.setState({ open: false });
              } },
            'Click to close'
          )
        ),
        React.createElement(_FloatingHelper2.default, {
          opened: this.state.open,
          target: React.createElement(
            'span',
            null,
            'I am a FloatingHelper target'
          ),
          content: React.createElement(_FloatingHelper2.default.Content, {
            title: 'Don\u2019t forget to setup payments',
            body: 'In order to sell your music you need to choose a payment method.'
          }),
          placement: 'right'
        })
      );
    }
  }]);

  return ControlledExample;
}(React.Component);
