// wix-style-react/dist/stories/FloatingHelper/FullExample.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FullExample = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _FloatingHelper = require('wix-style-react/FloatingHelper');

var _FloatingHelper2 = _interopRequireDefault(_FloatingHelper);

var _Image = require('wix-ui-icons-common/Image');

var _Image2 = _interopRequireDefault(_Image);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var FullExample = exports.FullExample = function (_React$Component) {
  _inherits(FullExample, _React$Component);

  function FullExample() {
    _classCallCheck(this, FullExample);

    return _possibleConstructorReturn(this, (FullExample.__proto__ || Object.getPrototypeOf(FullExample)).apply(this, arguments));
  }

  _createClass(FullExample, [{
    key: 'render',
    value: function render() {
      return React.createElement(_FloatingHelper2.default, {
        target: React.createElement(
          'span',
          null,
          'I am a FloatingHelper target'
        ),
        content: React.createElement(_FloatingHelper2.default.Content, {
          title: 'Don\u2019t forget to setup payments',
          body: 'In order to sell your music you need to choose a payment method.',
          actionText: 'Ok, Take Me There',
          onActionClick: function onActionClick() {
            return null;
          },
          image: React.createElement(_Image2.default, { width: '102', height: '102', viewBox: '4 4 18 18' })
        }),
        placement: 'right'
      });
    }
  }]);

  return FullExample;
}(React.Component);
