// wix-ui-core/dist/src/components/Thumbnail/Thumbnail.driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var test_utils_1 = require("stylable/test-utils");
var Thumbnail_st_css_1 = require("./Thumbnail.st.css");
exports.thumbnailDriverFactory = function (_a) {
    var element = _a.element, eventTrigger = _a.eventTrigger;
    var stylableDOMUtil = new test_utils_1.StylableDOMUtil(Thumbnail_st_css_1.default);
    var selectedIcon = element.querySelector('[data-hook="selected-icon"]');
    return {
        /** checks if the element exists */
        exists: function () { return !!element; },
        /** checks if the selected icon element exists */
        hasSelectedIcon: function () { return window.getComputedStyle(selectedIcon).getPropertyValue('display') !== 'none'; },
        /** triggers a click event on the element */
        click: function () { return eventTrigger.click(element); },
        /** check if component is selected */
        isSelected: function () { return stylableDOMUtil.hasStyleState(element, 'selected'); },
        /** returns the components's children */
        getContent: function () { return element.childNodes[0]; },
        /** returns the components's selected icon */
        getSelectedIcon: function () { return selectedIcon.childNodes[0]; },
        /** check if component is disabled */
        isDisabled: function () { return stylableDOMUtil.hasStyleState(element, 'disabled'); },
    };
};
