// wix-style-react/dist/stories/InputWithOptions/ExampleNoDropdownIfEmptyInput.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _InputWithOptions = require('wix-style-react/InputWithOptions');

var _InputWithOptions2 = _interopRequireDefault(_InputWithOptions);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var style = {
  display: 'inline-block',
  padding: '0 5px 0',
  width: '200px',
  lineHeight: '22px'
};

var options = [{ id: '0', value: 'First option' }, { id: '1', value: 'Second option' }, { id: '2', value: 'Third option', disabled: true }, { id: '3', value: 'Fourth option' }, { id: '4', value: 'Fifth option' }, { id: '5', value: 'Very long option text jldlkasj ldk jsalkdjsal kdjaklsjdlkasj dklasj' }];

var NoDropdownIfEmptyInput = function (_React$Component) {
  _inherits(NoDropdownIfEmptyInput, _React$Component);

  function NoDropdownIfEmptyInput() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, NoDropdownIfEmptyInput);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = NoDropdownIfEmptyInput.__proto__ || Object.getPrototypeOf(NoDropdownIfEmptyInput)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      value: ''
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(NoDropdownIfEmptyInput, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      return _react2.default.createElement(_InputWithOptions2.default, {
        showOptionsIfEmptyInput: false,
        options: options,
        value: this.state.value,
        onChange: function onChange(e) {
          return _this2.setState({ value: e.target.value });
        }
      });
    }
  }]);

  return NoDropdownIfEmptyInput;
}(_react2.default.Component);

exports.default = function () {
  return _react2.default.createElement(
    'div',
    { className: 'ltr', style: style },
    _react2.default.createElement(NoDropdownIfEmptyInput, null)
  );
};
