// wix-style-react/dist/stories/TextLink/index.story.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _TextLink = require('wix-style-react/TextLink');

var _TextLink2 = _interopRequireDefault(_TextLink);

var _CodeExample = require('wix-storybook-utils/CodeExample');

var _CodeExample2 = _interopRequireDefault(_CodeExample);

var _ExampleStandard = require('./ExampleStandard');

var _ExampleStandard2 = _interopRequireDefault(_ExampleStandard);

var _ExampleStandard3 = require('!raw-loader!./ExampleStandard');

var _ExampleStandard4 = _interopRequireDefault(_ExampleStandard3);

var _ExampleDark = require('./ExampleDark');

var _ExampleDark2 = _interopRequireDefault(_ExampleDark);

var _ExampleDark3 = require('!raw-loader!./ExampleDark');

var _ExampleDark4 = _interopRequireDefault(_ExampleDark3);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  category: '5. Buttons',
  storyName: '5.8 Text Link',
  component: _TextLink2.default,
  componentPath: '../../src/Backoffice/TextLink',

  componentProps: {
    link: 'https://wix.com',
    underlineStyle: 'hover',
    children: 'Click to visit wix.com'
  },

  example: _react2.default.createElement(
    'div',
    null,
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Standard', code: _ExampleStandard4.default },
      _react2.default.createElement(_ExampleStandard2.default, null)
    ),
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Dark', code: _ExampleDark4.default },
      _react2.default.createElement(_ExampleDark2.default, null)
    )
  )
};
