// wix-style-react/dist/src/Input/InputErrorSuffix.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _FormFieldErrorFilled = require('wix-ui-icons-common/system/FormFieldErrorFilled');

var _FormFieldErrorFilled2 = _interopRequireDefault(_FormFieldErrorFilled);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Tooltip = require('../Tooltip');

var _Tooltip2 = _interopRequireDefault(_Tooltip);

var _Input = require('./Input.scss');

var _Input2 = _interopRequireDefault(_Input);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var InputErrorSuffix = function (_React$Component) {
  _inherits(InputErrorSuffix, _React$Component);

  function InputErrorSuffix() {
    _classCallCheck(this, InputErrorSuffix);

    return _possibleConstructorReturn(this, (InputErrorSuffix.__proto__ || Object.getPrototypeOf(InputErrorSuffix)).apply(this, arguments));
  }

  _createClass(InputErrorSuffix, [{
    key: 'render',
    value: function render() {
      var classes = (0, _classnames2.default)(_Input2.default.exclamation, _defineProperty({}, _Input2.default.narrow, this.props.narrow));
      return _react2.default.createElement(
        _Tooltip2.default,
        {
          dataHook: 'input-tooltip',
          disabled: this.props.errorMessage.length === 0,
          placement: this.props.tooltipPlacement,
          alignment: 'center',
          textAlign: 'left',
          content: this.props.errorMessage,
          overlay: '',
          theme: 'dark',
          maxWidth: '230px',
          hideDelay: 150,
          zIndex: 10000
        },
        _react2.default.createElement(
          'div',
          { className: classes },
          _react2.default.createElement(_FormFieldErrorFilled2.default, null)
        )
      );
    }
  }]);

  return InputErrorSuffix;
}(_react2.default.Component);

InputErrorSuffix.propTypes = {
  theme: _propTypes2.default.oneOf(['normal', 'paneltitle', 'material', 'amaterial']),
  errorMessage: _propTypes2.default.string.isRequired,
  focused: _propTypes2.default.bool,
  narrow: _propTypes2.default.bool,
  tooltipPlacement: _propTypes2.default.string
};

exports.default = InputErrorSuffix;
