// wix-style-react/dist/src/Input/InputHelpSuffix.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Tooltip = require('../Tooltip');

var _Tooltip2 = _interopRequireDefault(_Tooltip);

var _InfoCircle = require('../new-icons/InfoCircle');

var _InfoCircle2 = _interopRequireDefault(_InfoCircle);

var _Input = require('./Input.scss');

var _Input2 = _interopRequireDefault(_Input);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var InputHelpSuffix = function (_React$Component) {
  _inherits(InputHelpSuffix, _React$Component);

  function InputHelpSuffix() {
    _classCallCheck(this, InputHelpSuffix);

    return _possibleConstructorReturn(this, (InputHelpSuffix.__proto__ || Object.getPrototypeOf(InputHelpSuffix)).apply(this, arguments));
  }

  _createClass(InputHelpSuffix, [{
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        _Tooltip2.default,
        {
          dataHook: 'input-tooltip',
          disabled: this.props.helpMessage.length === 0,
          maxWidth: '230px',
          placement: 'right',
          alignment: 'center',
          textAlign: 'left',
          hideDelay: 100,
          content: this.props.helpMessage,
          overlay: ''
        },
        _react2.default.createElement(
          'div',
          { className: _Input2.default.help },
          _react2.default.createElement(_InfoCircle2.default, null)
        )
      );
    }
  }]);

  return InputHelpSuffix;
}(_react2.default.Component);

InputHelpSuffix.propTypes = {
  theme: _propTypes2.default.oneOf(['normal', 'paneltitle', 'material', 'amaterial']),
  helpMessage: _propTypes2.default.string.isRequired,
  help: _propTypes2.default.bool
};

exports.default = InputHelpSuffix;
