// wix-style-react/dist/src/Input/ThemedInputErrorSuffix.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _get = function get(object, property, receiver) { if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { return get(parent, property, receiver); } } else if ("value" in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _InputErrorSuffix2 = require('./InputErrorSuffix');

var _InputErrorSuffix3 = _interopRequireDefault(_InputErrorSuffix2);

var _Tooltip = require('../Tooltip');

var _Tooltip2 = _interopRequireDefault(_Tooltip);

var _FormFieldErrorFilled = require('../new-icons/system/FormFieldErrorFilled');

var _FormFieldErrorFilled2 = _interopRequireDefault(_FormFieldErrorFilled);

var _Input = require('./Input.scss');

var _Input2 = _interopRequireDefault(_Input);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var placementToMoveBy = {
  right: { x: 2, y: -10 },
  left: { x: -2, y: -10 },
  top: { x: 0, y: -10 },
  bottom: { x: 0, y: -8 }
};

var AmaterialErrorSuffix = function AmaterialErrorSuffix(_ref) {
  var focused = _ref.focused,
      error = _ref.error,
      errorMessage = _ref.errorMessage,
      placement = _ref.placement,
      onShow = _ref.onShow;
  return focused ? null : _react2.default.createElement(
    _Tooltip2.default,
    {
      dataHook: 'input-tooltip',
      disabled: !error && !errorMessage,
      placement: placement,
      maxWidth: '230px',
      hideDelay: 5,
      showDelay: 5,
      active: focused,
      moveBy: placementToMoveBy[placement],
      alignment: 'center',
      content: errorMessage,
      overlay: '',
      textAlign: 'left',
      onShow: onShow
    },
    _react2.default.createElement(
      'div',
      { className: (0, _classnames2.default)(_Input2.default.errorIcon, _Input2.default.suffix) },
      _react2.default.createElement(_FormFieldErrorFilled2.default, { size: '1.5em' })
    )
  );
};

AmaterialErrorSuffix.propTypes = {
  errorMessage: _propTypes2.default.node.isRequired,
  error: _propTypes2.default.bool,
  focused: _propTypes2.default.bool,
  placement: _propTypes2.default.oneOf(['left', 'right', 'top', 'bottom']),
  onShow: _propTypes2.default.func
};

AmaterialErrorSuffix.defaultProps = {
  placement: 'right'
};

var ThemedInputErrorSuffix = function (_InputErrorSuffix) {
  _inherits(ThemedInputErrorSuffix, _InputErrorSuffix);

  function ThemedInputErrorSuffix() {
    _classCallCheck(this, ThemedInputErrorSuffix);

    return _possibleConstructorReturn(this, (ThemedInputErrorSuffix.__proto__ || Object.getPrototypeOf(ThemedInputErrorSuffix)).apply(this, arguments));
  }

  _createClass(ThemedInputErrorSuffix, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          theme = _props.theme,
          focused = _props.focused,
          error = _props.error,
          errorMessage = _props.errorMessage,
          tooltipPlacement = _props.tooltipPlacement,
          onTooltipShow = _props.onTooltipShow;

      return theme === 'amaterial' ? _react2.default.createElement(AmaterialErrorSuffix, { focused: focused, error: error, errorMessage: errorMessage, placement: tooltipPlacement, onShow: onTooltipShow }) : _get(ThemedInputErrorSuffix.prototype.__proto__ || Object.getPrototypeOf(ThemedInputErrorSuffix.prototype), 'render', this).call(this);
    }
  }]);

  return ThemedInputErrorSuffix;
}(_InputErrorSuffix3.default);

ThemedInputErrorSuffix.propTypes = {
  tooltipPlacement: _propTypes2.default.oneOf(['left', 'right', 'top', 'bottom']),
  onTooltipShow: _propTypes2.default.func
};

exports.default = ThemedInputErrorSuffix;
