// wix-style-react/dist/stories/Tooltip/Composite/ExamplePopoverMenu.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _RadioGroup = require('wix-style-react/RadioGroup');

var _RadioGroup2 = _interopRequireDefault(_RadioGroup);

var _Label = require('wix-style-react/Label');

var _Label2 = _interopRequireDefault(_Label);

var _Input = require('wix-style-react/Input');

var _Input2 = _interopRequireDefault(_Input);

var _PopoverMenuBuilder = require('./PopoverMenuBuilder');

var _PopoverMenuBuilder2 = _interopRequireDefault(_PopoverMenuBuilder);

var _PopoverMenuTemplate = require('./PopoverMenuTemplate');

var _PopoverMenuTemplate2 = _interopRequireDefault(_PopoverMenuTemplate);

var _Example = require('./Example.scss');

var _Example2 = _interopRequireDefault(_Example);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ExamplePopoverMenu = function (_Component) {
  _inherits(ExamplePopoverMenu, _Component);

  function ExamplePopoverMenu() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ExamplePopoverMenu);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ExamplePopoverMenu.__proto__ || Object.getPrototypeOf(ExamplePopoverMenu)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      size: 'normal',
      placement: 'top',
      menuItems: [{ iconName: 'Edit', text: 'Edit' }, { iconName: 'Hidden', text: 'Hide' }, { iconName: 'Delete', text: 'Delete' }],
      buttonTheme: 'icon-greybackground',
      maxWidth: '378px'
    }, _this.addRow = function () {
      return _this.setState({ menuItems: [].concat(_toConsumableArray(_this.state.menuItems), [{ iconName: '', text: '' }]) });
    }, _this.updateRowIcon = function (iconName, i) {
      return _this.setState({
        menuItems: [].concat(_toConsumableArray(_this.state.menuItems.slice(0, i).map(function (obj) {
          return Object.assign({}, obj);
        })), [{ iconName: iconName, text: _this.state.menuItems[i].text }], _toConsumableArray(_this.state.menuItems.slice(i + 1).map(function (obj) {
          return Object.assign({}, obj);
        })))
      });
    }, _this.updateRowText = function (text, i) {
      return _this.setState({
        menuItems: [].concat(_toConsumableArray(_this.state.menuItems.slice(0, i).map(function (obj) {
          return Object.assign({}, obj);
        })), [{ iconName: _this.state.menuItems[i].iconName, text: text }], _toConsumableArray(_this.state.menuItems.slice(i + 1).map(function (obj) {
          return Object.assign({}, obj);
        })))
      });
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ExamplePopoverMenu, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      return _react2.default.createElement(
        'form',
        { className: _Example2.default.form },
        _react2.default.createElement(
          'div',
          { className: _Example2.default.input },
          _react2.default.createElement(
            'div',
            { className: _Example2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Size'
            ),
            _react2.default.createElement(
              'div',
              { className: _Example2.default.flex },
              _react2.default.createElement(
                _RadioGroup2.default,
                {
                  display: 'horizontal',
                  value: this.state.size,
                  onChange: function onChange(size) {
                    return _this2.setState({ size: size });
                  }
                },
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'normal' },
                  'Normal'
                ),
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'large' },
                  'Large'
                )
              )
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _Example2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Direction'
            ),
            _react2.default.createElement(
              'div',
              { className: _Example2.default.flex },
              _react2.default.createElement(
                _RadioGroup2.default,
                {
                  display: 'horizontal',
                  value: this.state.placement,
                  onChange: function onChange(placement) {
                    return _this2.setState({ placement: placement });
                  }
                },
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'top' },
                  'Top'
                ),
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'right' },
                  'Right'
                ),
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'bottom' },
                  'Bottom'
                ),
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'left' },
                  'Left'
                )
              )
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _Example2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Button Theme'
            ),
            _react2.default.createElement(
              'div',
              { className: _Example2.default.flex },
              _react2.default.createElement(
                _RadioGroup2.default,
                {
                  display: 'vertical',
                  value: this.state.buttonTheme,
                  onChange: function onChange(buttonTheme) {
                    return _this2.setState({ buttonTheme: buttonTheme });
                  }
                },
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'icon-greybackground' },
                  'Icon Grey Background'
                ),
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'icon-standard' },
                  'Icon Blue'
                ),
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'icon-standardsecondary' },
                  'Icon Blue Secondary'
                ),
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'icon-white' },
                  'Icon White'
                ),
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'icon-whitesecondary' },
                  'Icon White Secondary'
                )
              )
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _Example2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Max Width'
            ),
            _react2.default.createElement(
              'div',
              { className: _Example2.default.flex },
              _react2.default.createElement(_Input2.default, {
                size: 'small',
                value: this.state.maxWidth,
                onChange: function onChange(e) {
                  return _this2.setState({ maxWidth: e.target.value });
                }
              })
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _Example2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Text'
            ),
            _react2.default.createElement(
              'div',
              { className: _Example2.default.flex },
              _react2.default.createElement(_PopoverMenuBuilder2.default, {
                menuItems: this.state.menuItems,
                updateRowText: this.updateRowText,
                addRow: this.addRow,
                updateRowIcon: this.updateRowIcon
              })
            )
          )
        ),
        _react2.default.createElement(
          'div',
          { className: _Example2.default.output },
          _react2.default.createElement(
            'div',
            { className: _Example2.default.exampleWrapper },
            _react2.default.createElement(_PopoverMenuTemplate2.default, {
              menuItems: this.state.menuItems,
              size: this.state.size,
              placement: this.state.placement,
              onChange: this.props.onChange,
              buttonTheme: this.state.buttonTheme,
              maxWidth: this.state.maxWidth
            })
          )
        )
      );
    }
  }]);

  return ExamplePopoverMenu;
}(_react.Component);

ExamplePopoverMenu.propTypes = {
  onChange: _propTypes2.default.func
};
exports.default = ExamplePopoverMenu;
