// wix-style-react/dist/stories/Tooltip/Composite/ExamplePopover.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Template = require('./Template');

var _Template2 = _interopRequireDefault(_Template);

var _RadioGroup = require('../../../src/RadioGroup');

var _RadioGroup2 = _interopRequireDefault(_RadioGroup);

var _Label = require('../../../src/Label');

var _Label2 = _interopRequireDefault(_Label);

var _Input = require('../../../src/Input');

var _Input2 = _interopRequireDefault(_Input);

var _Example = require('./Example.scss');

var _Example2 = _interopRequireDefault(_Example);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ExamplePopover = function (_Component) {
  _inherits(ExamplePopover, _Component);

  function ExamplePopover() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ExamplePopover);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ExamplePopover.__proto__ || Object.getPrototypeOf(ExamplePopover)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      placement: 'top',
      text: 'Popover appears on click',
      maxWidth: ''
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ExamplePopover, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      return _react2.default.createElement(
        'form',
        { className: _Example2.default.form },
        _react2.default.createElement(
          'div',
          { className: _Example2.default.input },
          _react2.default.createElement(
            'div',
            { className: _Example2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Direction'
            ),
            _react2.default.createElement(
              'div',
              { className: _Example2.default.flex },
              _react2.default.createElement(
                _RadioGroup2.default,
                {
                  display: 'horizontal',
                  value: this.state.placement,
                  onChange: function onChange(placement) {
                    return _this2.setState({ placement: placement });
                  }
                },
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'top' },
                  'Top'
                ),
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'right' },
                  'Right'
                ),
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'bottom' },
                  'Bottom'
                ),
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'left' },
                  'Left'
                )
              )
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _Example2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Text'
            ),
            _react2.default.createElement(
              'div',
              { className: _Example2.default.flex },
              _react2.default.createElement(_Input2.default, {
                size: 'small',
                value: this.state.text,
                onChange: function onChange(e) {
                  return _this2.setState({ text: e.target.value });
                }
              })
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _Example2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Max Width'
            ),
            _react2.default.createElement(
              'div',
              { className: _Example2.default.flex },
              _react2.default.createElement(_Input2.default, {
                size: 'small',
                value: this.state.maxWidth,
                onChange: function onChange(e) {
                  return _this2.setState({ maxWidth: e.target.value });
                }
              })
            )
          )
        ),
        _react2.default.createElement(
          'div',
          { className: _Example2.default.output },
          _react2.default.createElement(
            'div',
            { className: _Example2.default.exampleWrapper },
            _react2.default.createElement(_Template2.default, {
              theme: 'light',
              placement: this.state.placement,
              tooltipContent: this.state.text,
              showTrigger: 'click',
              hideTrigger: 'click',
              type: 'popover',
              maxWidth: this.state.maxWidth,
              onChange: this.props.onChange
            })
          )
        )
      );
    }
  }]);

  return ExamplePopover;
}(_react.Component);

exports.default = ExamplePopover;


ExamplePopover.propTypes = {
  onChange: _propTypes2.default.func.isRequired
};
