// wix-style-react/dist/src/Input/Unit/testkit/Unit.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.unitDriverFactory = exports.componentFactory = exports.unitTestkitFactory = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _Unit = require('../Unit');

var _Unit2 = _interopRequireDefault(_Unit);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var unitDriverFactory = function unitDriverFactory(_ref) {
  var element = _ref.element;

  return {
    isEmpty: function isEmpty() {
      return element.children.length === 0;
    },
    hasChild: function hasChild(style) {
      return !!element.querySelector(style);
    }
  };
};

var componentFactory = function componentFactory() {
  var props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  var element = void 0;
  var wrapperDiv = document.createElement('div');
  _reactDom2.default.render(_react2.default.createElement(
    'div',
    { ref: function ref(r) {
        return element = r;
      } },
    _react2.default.createElement(_Unit2.default, props)
  ), wrapperDiv);
  return { element: element.childNodes[0], wrapper: wrapperDiv };
};

var unitTestkitFactory = function unitTestkitFactory(_ref2) {
  var wrapper = _ref2.wrapper,
      dataHook = _ref2.dataHook;

  var element = wrapper.querySelector('[data-hook=\'' + dataHook + '\']');
  return unitDriverFactory({ element: element, wrapper: wrapper });
};

exports.unitTestkitFactory = unitTestkitFactory;
exports.componentFactory = componentFactory;
exports.unitDriverFactory = unitDriverFactory;
