// wix-style-react/dist/src/Tabs/core/constants/tab-prop-types.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.type = exports.sideContent = exports.width = exports.activeId = exports.onClick = exports.items = exports.item = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _tabTypes = require('./tab-types');

var _tabTypes2 = _interopRequireDefault(_tabTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var stringOrNumber = _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]);

var item = exports.item = _propTypes2.default.shape({ id: stringOrNumber, title: _propTypes2.default.node, dataHook: _propTypes2.default.string });

var items = exports.items = _propTypes2.default.arrayOf(item);

var onClick = exports.onClick = _propTypes2.default.func;

var activeId = exports.activeId = stringOrNumber;

var width = exports.width = stringOrNumber;

var sideContent = exports.sideContent = _propTypes2.default.node;

var type = exports.type = _propTypes2.default.oneOf(_tabTypes2.default);
