// wix-ui-backoffice/dist/src/components/FloatingHelper/FloatingHelper.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var pick = require("lodash/pick");
var prop_types_1 = require("prop-types");
var ClosablePopover_1 = require("./ClosablePopover");
var FloatingHelper_st_css_1 = require("./FloatingHelper.st.css");
var DataHooks_1 = require("./DataHooks");
var CloseButton_1 = require("../CloseButton");
var FloatingHelperContent_1 = require("./FloatingHelperContent");
var pickedPropNames = ['initiallyOpened', 'opened', 'target', 'onClose', 'onOpen', 'placement', 'appendTo'];
var pickedPopoverPropTypes = pick.apply(void 0, [ClosablePopover_1.ClosablePopover.propTypes].concat(pickedPropNames));
var FloatingHelper = /** @class */ (function (_super) {
    __extends(FloatingHelper, _super);
    function FloatingHelper() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.open = function () { _this.closablePopoverRef.open(); };
        _this.close = function () { _this.closablePopoverRef.close(); };
        return _this;
    }
    FloatingHelper.prototype.isControlled = function () {
        return this.props.opened !== undefined;
    };
    FloatingHelper.prototype.getCloseButtonHandler = function (closableActions) {
        return this.isControlled() ?
            (this.props.onClose ? this.props.onClose : function () { return null; }) :
            closableActions.close;
    };
    FloatingHelper.prototype.renderContent = function (closableActions, _a) {
        var _this = this;
        var width = _a.width, content = _a.content;
        return (React.createElement("div", { "data-hook": DataHooks_1.DataHooks.contentWrapper, style: { width: width } },
            React.createElement(CloseButton_1.CloseButton, { className: FloatingHelper_st_css_1.default.closeButton, "data-hook": DataHooks_1.DataHooks.closeButton, onClick: function () { return _this.getCloseButtonHandler(closableActions)(); }, skin: CloseButton_1.CloseButtonSkin.white, size: CloseButton_1.CloseButtonSize.large }),
            React.createElement("div", { "data-hook": DataHooks_1.DataHooks.innerContent, className: FloatingHelper_st_css_1.default.innerContent }, content)));
    };
    FloatingHelper.prototype.render = function () {
        var _this = this;
        var _a = this.props, children = _a.children, width = _a.width, content = _a.content, rest = __rest(_a, ["children", "width", "content"]);
        var renderContent = function (closableActions) { return _this.renderContent(closableActions, { width: width, content: content }); };
        var closablePopoverProps = __assign({}, rest, { content: renderContent, showArrow: true, closeOnMouseLeave: false // TODO: We pass this as a workaround because with React-15 the mouse-enter does not propagate from the Portal to the React-Popper's Manager
         });
        return (React.createElement(ClosablePopover_1.ClosablePopover, __assign({}, closablePopoverProps, { ref: function (ref) { return _this.closablePopoverRef = ref; } }, FloatingHelper_st_css_1.default('root', {}, this.props))));
    };
    ;
    FloatingHelper.displayName = 'FloatingHelper';
    FloatingHelper.Content = FloatingHelperContent_1.FloatingHelperContent;
    FloatingHelper.defaultProps = {
        appendTo: 'window',
        width: '444px',
        initiallyOpened: true
    };
    FloatingHelper.propTypes = __assign({}, pickedPopoverPropTypes, { width: prop_types_1.oneOfType([prop_types_1.string, prop_types_1.number]), target: prop_types_1.node.isRequired, content: prop_types_1.node.isRequired // TODO: validate it is a <HelperContent>
     });
    return FloatingHelper;
}(React.Component));
exports.FloatingHelper = FloatingHelper;
