// wix-ui-core/dist/src/components/CircularProgressBar/CircularProgressBar.driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var test_utils_1 = require("stylable/test-utils");
var CircularProgressBar_st_css_1 = require("./CircularProgressBar.st.css");
exports.circularProgressBarDriverFactory = function (_a) {
    var element = _a.element;
    var stylableDOMUtil = new test_utils_1.StylableDOMUtil(CircularProgressBar_st_css_1.default);
    var getElement = function (dataHook) { return element.querySelector("[data-hook=\"" + dataHook + "\"]"); };
    var getValue = function () { return !element ? null : getElement('progress-indicator').innerHTML; };
    var driver = {
        exists: function () { return !!element; },
        isSuccessIconDisplayed: function () { return !!getElement('success-icon'); },
        isErrorIconDisplayed: function () { return !!getElement('error-icon'); },
        isPercentagesProgressDisplayed: function () { return !!getElement('progress-indicator'); },
        getValue: function () { return getValue(); },
        isCompleted: function () { return getValue() === '100'; },
        hasError: function () { return stylableDOMUtil.hasStyleState(element, 'error'); },
    };
    return driver;
};
