// wix-style-react/dist/stories/RichTextArea/RichTextAreaExample.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Label = require('../../src/Label');

var _Label2 = _interopRequireDefault(_Label);

var _TextField = require('../../src/TextField');

var _TextField2 = _interopRequireDefault(_TextField);

var _TextArea = require('../../src/TextArea');

var _TextArea2 = _interopRequireDefault(_TextArea);

var _Input = require('../../src/Input');

var _Input2 = _interopRequireDefault(_Input);

var _InputArea = require('../../src/InputArea');

var _InputArea2 = _interopRequireDefault(_InputArea);

var _ToggleSwitch = require('../../src/ToggleSwitch');

var _ToggleSwitch2 = _interopRequireDefault(_ToggleSwitch);

var _RichTextAreaTemplate = require('./RichTextAreaTemplate');

var _RichTextAreaTemplate2 = _interopRequireDefault(_RichTextAreaTemplate);

var _RichTextAreaExample = require('./RichTextAreaExample.scss');

var _RichTextAreaExample2 = _interopRequireDefault(_RichTextAreaExample);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var RichTextAreaExample = function (_Component) {
  _inherits(RichTextAreaExample, _Component);

  function RichTextAreaExample() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, RichTextAreaExample);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = RichTextAreaExample.__proto__ || Object.getPrototypeOf(RichTextAreaExample)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      error: false,
      errorMessage: '',
      resizable: false,
      onImageRequest: function onImageRequest(callback) {
        var src = window.prompt('Enter the URL of the image:');
        if (!src) {
          return;
        }
        callback(src);
      },
      placeholder: 'Rich placeholder',
      value: '<p>$$$ Rich text area</p><strong>bold text here</strong><ul><li>The amazing fox lives in <a href="wix.com">Lithuania</a></li></ul>'
    }, _this.handleRichTextAreaChange = function (value) {
      _this.setState({ value: value });
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(RichTextAreaExample, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      return _react2.default.createElement(
        'div',
        { className: _RichTextAreaExample2.default.form },
        _react2.default.createElement(
          'div',
          { className: _RichTextAreaExample2.default.input },
          _react2.default.createElement(
            'div',
            { className: _RichTextAreaExample2.default.option },
            _react2.default.createElement(
              _Label2.default,
              { 'for': 'placeholderInput' },
              'Placeholder'
            ),
            _react2.default.createElement(
              'div',
              { className: _RichTextAreaExample2.default.flex },
              _react2.default.createElement(_Input2.default, {
                id: 'placeholderInput',
                size: 'normal',
                theme: 'normal',
                value: this.state.placeholder,
                onChange: function onChange(event) {
                  return _this2.setState({ placeholder: event.target.value });
                }
              })
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _RichTextAreaExample2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Error: '
            ),
            _react2.default.createElement(_ToggleSwitch2.default, {
              size: 'small',
              checked: this.state.error,
              onChange: function onChange() {
                return _this2.setState({ error: !_this2.state.error });
              }
            })
          ),
          this.renderErrorMessageInput(),
          _react2.default.createElement(
            'div',
            { className: _RichTextAreaExample2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Disabled: '
            ),
            _react2.default.createElement(_ToggleSwitch2.default, {
              size: 'small',
              checked: this.state.disabled,
              onChange: function onChange() {
                return _this2.setState({ disabled: !_this2.state.disabled });
              }
            })
          ),
          _react2.default.createElement(
            'div',
            { className: _RichTextAreaExample2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Resizable: '
            ),
            _react2.default.createElement(_ToggleSwitch2.default, {
              size: 'small',
              checked: this.state.resizable,
              onChange: function onChange() {
                return _this2.setState({ resizable: !_this2.state.resizable });
              }
            })
          )
        ),
        _react2.default.createElement(
          'div',
          { className: _RichTextAreaExample2.default.output },
          _react2.default.createElement(_RichTextAreaTemplate2.default, _extends({
            onTemplateChange: this.props.onChange,
            onChange: this.handleRichTextAreaChange,
            disabled: this.state.disabled
          }, this.state)),
          _react2.default.createElement(
            'div',
            { className: _RichTextAreaExample2.default.output },
            _react2.default.createElement(
              _TextArea2.default,
              null,
              _react2.default.createElement(
                _Label2.default,
                { 'for': 'placeholderInput' },
                'Output value'
              ),
              _react2.default.createElement(_InputArea2.default, {
                id: 'placeholderInput',
                size: 'normal',
                theme: 'normal',
                value: this.state.value,
                resizable: true,
                readOnly: true
              })
            )
          )
        )
      );
    }
  }, {
    key: 'renderErrorMessageInput',
    value: function renderErrorMessageInput() {
      var _this3 = this;

      var _state = this.state,
          error = _state.error,
          errorMessage = _state.errorMessage;


      if (!error) {
        return null;
      }

      return _react2.default.createElement(
        'div',
        { className: _RichTextAreaExample2.default.output },
        _react2.default.createElement(
          _TextField2.default,
          null,
          _react2.default.createElement(
            _Label2.default,
            { 'for': 'errorMessageInput' },
            'Error message'
          ),
          _react2.default.createElement(_Input2.default, {
            id: 'errorMessageInput',
            size: 'normal',
            theme: 'normal',
            value: errorMessage,
            onChange: function onChange(event) {
              return _this3.setState({ errorMessage: event.target.value });
            }
          })
        )
      );
    }
  }]);

  return RichTextAreaExample;
}(_react.Component);

RichTextAreaExample.propTypes = {
  onChange: _propTypes2.default.func
};

exports.default = RichTextAreaExample;
