// wix-style-react/dist/src/Backoffice/ButtonLayout/ButtonLayout.driver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _utils = require('../../../test/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var buttonLayoutDriverFactory = function buttonLayoutDriverFactory(_ref) {
  var element = _ref.element,
      wrapper = _ref.wrapper,
      component = _ref.component;

  var getAttribute = function getAttribute(element, attribute) {
    return element.getAttribute(attribute);
  };

  return {
    exists: function exists() {
      return !!element;
    },
    doesComponentHasClass: function doesComponentHasClass(className) {
      return (0, _utils.isClassExists)(element, className);
    },
    getComponentAttribute: function getComponentAttribute(attribute) {
      return getAttribute(element, attribute);
    },
    setProps: function setProps(props) {
      var ClonedWithProps = _react2.default.cloneElement.apply(_react2.default, [component, Object.assign({}, component.props, props)].concat(_toConsumableArray(component.props.children || [])));
      _reactDom2.default.render(_react2.default.createElement(
        'div',
        { ref: function ref(r) {
            return element = r;
          } },
        ClonedWithProps
      ), wrapper);
    }
  };
};

exports.default = buttonLayoutDriverFactory;
