// wix-style-react/dist/stories/ButtonLayout.story.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _ButtonLayout = require('wix-style-react/ButtonLayout');

var _ButtonLayout2 = _interopRequireDefault(_ButtonLayout);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var children = _react2.default.createElement(
  'a',
  {
    href: 'http://www.wix.com',
    target: '_blank',
    rel: 'noopener noreferrer',
    style: { textDecoration: 'inherit' }
  },
  'Link Like Button'
);

exports.default = {
  category: '5. Buttons',
  storyName: '5.0 ButtonLayout',
  component: _ButtonLayout2.default,
  componentPath: '../src/ButtonLayout',
  componentProps: {
    children: children,
    theme: 'fullblue',
    height: 'medium'
  },

  exampleProps: {
    children: [{ label: 'anchor element', value: children }, { label: 'text in span', value: _react2.default.createElement(
        'span',
        null,
        'text in span'
      ) }]
  }
};
