// wix-style-react/dist/stories/Checkbox.story.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Checkbox = require('wix-style-react/Checkbox');

var _Checkbox2 = _interopRequireDefault(_Checkbox);

var _Languages = require('wix-style-react/new-icons/Languages');

var _Languages2 = _interopRequireDefault(_Languages);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var labelExamples = [{ label: 'Simple string', value: 'Hello World!' }, {
  label: 'Component',
  value: _react2.default.createElement(
    'span',
    { key: 0 },
    'Hello ',
    _react2.default.createElement(
      'strong',
      null,
      'World!'
    )
  )
}, {
  label: 'Component with icon',
  value: _react2.default.createElement(
    'span',
    { key: 1 },
    'Hello ',
    _react2.default.createElement(_Languages2.default, null)
  )
}];

exports.default = {
  category: '4. Selection',
  storyName: '4.2 Checkbox',
  component: _Checkbox2.default,
  componentPath: '../src/Checkbox',

  componentProps: function componentProps(setState) {
    return {
      children: labelExamples[0].value,
      onChange: function onChange(_ref) {
        var checked = _ref.target.checked;
        return setState({ checked: checked });
      },
      dataHook: 'storybook-checkbox'
    };
  },

  exampleProps: {
    children: labelExamples,
    onChange: function onChange(_ref2) {
      var checked = _ref2.target.checked;
      return checked ? 'Checked' : 'Unchecked';
    }
  }
};
