// wix-style-react/dist/stories/FormField.story.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _FormField = require('wix-style-react/FormField');

var _FormField2 = _interopRequireDefault(_FormField);

var _Input = require('wix-style-react/Input');

var _Input2 = _interopRequireDefault(_Input);

var _InputArea = require('wix-style-react/InputArea');

var _InputArea2 = _interopRequireDefault(_InputArea);

var _RichTextArea = require('wix-style-react/RichTextArea');

var _RichTextArea2 = _interopRequireDefault(_RichTextArea);

var _DatePicker = require('wix-style-react/DatePicker');

var _DatePicker2 = _interopRequireDefault(_DatePicker);

var _Dropdown = require('wix-style-react/Dropdown');

var _Dropdown2 = _interopRequireDefault(_Dropdown);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ID = 'formFieldId';

var placeholder = 'Default text goes here...';
var childrenExamples = [{ label: 'Input',
  value: _react2.default.createElement(_Input2.default, { placeholder: placeholder, id: ID })
}, { label: 'Input with char counter',
  // eslint-disable-next-line react/prop-types
  value: function value(_ref) {
    var setCharactersLeft = _ref.setCharactersLeft;
    return _react2.default.createElement(_Input2.default, {
      placeholder: placeholder,
      onChange: function onChange(e) {
        return setCharactersLeft(50 - e.target.value.length);
      },
      id: ID
    });
  }
}, { label: 'InputArea',
  value: _react2.default.createElement(_InputArea2.default, { placeholder: placeholder, id: ID })
}, { label: 'InputArea with char counter',
  // eslint-disable-next-line react/prop-types
  value: function value(_ref2) {
    var setCharactersLeft = _ref2.setCharactersLeft;
    return _react2.default.createElement(_InputArea2.default, {
      placeholder: placeholder,
      id: ID,
      onChange: function onChange(e) {
        return setCharactersLeft(50 - e.target.value.length);
      }
    });
  }
}, { label: 'RichTextArea',
  value: _react2.default.createElement(_RichTextArea2.default, { placeholder: placeholder })
}, { label: 'DatePicker',
  value: _react2.default.createElement(_DatePicker2.default, { value: new Date() })
}, { label: 'Dropdown',
  value: _react2.default.createElement(_Dropdown2.default, { placeholder: 'Select dominant hand', options: [{ id: 0, value: 'Left' }, { id: 1, value: 'Right' }, { id: 2, value: 'Ambidextrous' }] })
}];

exports.default = {
  category: 'Components',
  storyName: 'FormField',
  component: _FormField2.default,
  componentPath: '../src/FormField',

  componentProps: {
    dataHook: 'storybook-formfield',
    children: childrenExamples[0].value,
    label: 'This is an input:',
    required: true,
    infoContent: 'I help you to fill info',
    id: 'formFieldId'
  },

  exampleProps: {
    children: childrenExamples
  }
};
