// wix-style-react/dist/stories/RadioGroup.story.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NUM_OF_BUTTONS_IN_EXAMPLE = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _RadioGroup = require('wix-style-react/RadioGroup');

var _RadioGroup2 = _interopRequireDefault(_RadioGroup);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var exampleChildren = [{
  label: '4 radios',
  value: [1, 2, 3, 4].map(function (n) {
    return _react2.default.createElement(_RadioGroup2.default.Radio, { key: n, value: n, children: 'Option ' + n });
  })
}, {
  label: '2 radios',
  value: [_react2.default.createElement(
    _RadioGroup2.default.Radio,
    { key: 0, value: 1 },
    _react2.default.createElement(
      'div',
      null,
      'Option 1'
    ),
    _react2.default.createElement(
      'small',
      null,
      'best option'
    )
  ), _react2.default.createElement(
    _RadioGroup2.default.Radio,
    { key: 1, value: 2 },
    _react2.default.createElement(
      'div',
      null,
      'Option 2'
    ),
    _react2.default.createElement(
      'small',
      null,
      'Also pretty good option'
    )
  )]
}];

var NUM_OF_BUTTONS_IN_EXAMPLE = exports.NUM_OF_BUTTONS_IN_EXAMPLE = exampleChildren[0].length;

exports.default = {
  category: '4. Selection',
  storyName: '4.3 Radio Button Group',
  component: _RadioGroup2.default,
  componentPath: '../src/RadioGroup',

  componentProps: function componentProps(setState) {
    return {
      value: 1,
      hasError: false,
      size: 'medium',
      children: exampleChildren[0].value,
      onChange: function onChange(value) {
        return setState({ value: value });
      },
      dataHook: 'storybook-radiogroup'
    };
  },

  exampleProps: {
    children: exampleChildren,
    onChange: function onChange(value) {
      return value;
    }
  }
};
