// wix-animations/dist/stories/Animations/ExampleMockServer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Animator = require('../../src/components/Animator');

var _Animator2 = _interopRequireDefault(_Animator);

var _Example = require('./Example.scss');

var css = _interopRequireWildcard(_Example);

var _AnimationTemplate = require('./AnimationTemplate');

var _AnimationTemplate2 = _interopRequireDefault(_AnimationTemplate);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Item = function Item() {
  return _react2.default.createElement(
    'div',
    { className: css.basicDiv },
    'I am an Item'
  );
};

var ExampleMockServer = function (_React$Component) {
  _inherits(ExampleMockServer, _React$Component);

  function ExampleMockServer(props) {
    _classCallCheck(this, ExampleMockServer);

    var _this = _possibleConstructorReturn(this, (ExampleMockServer.__proto__ || Object.getPrototypeOf(ExampleMockServer)).call(this, props));

    _this.state = {
      items: [],
      isLoader: !props.show
    };

    _this.to = { in: 'top', out: 'bottom' };
    return _this;
  }

  _createClass(ExampleMockServer, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(_ref) {
      var show = _ref.show;

      if (show) {
        this.setState({
          items: new Array(5).fill(1).map(function (_, index) {
            return _react2.default.createElement(Item, { key: index });
          })
        });
      } else {
        this.setState({
          items: []
        });
      }

      this.setState({ isLoader: !show });
    }
  }, {
    key: 'render',
    value: function render() {

      return _react2.default.createElement(
        'div',
        { style: { height: '70px', display: 'flex' } },
        _react2.default.createElement(
          _Animator2.default,
          { opacity: true, sequence: true, translate: { to: { in: 'top', out: 'bottom' } }, className: '' + css.flexParent },
          this.state.isLoader && _react2.default.createElement(
            'div',
            { key: 99999, animatorChildStyle: { position: 'absolute' }, style: { fontSize: '25px', textAlign: 'center' } },
            'Loading....'
          ),
          this.state.items
        )
      );
    }
  }]);

  return ExampleMockServer;
}(_react2.default.Component);

ExampleMockServer.propTypes = {
  show: _propTypes2.default.bool
};

exports.default = function () {
  return _react2.default.createElement(
    _AnimationTemplate2.default,
    null,
    _react2.default.createElement(ExampleMockServer, null)
  );
};
