// wix-animations/dist/stories/Animations/ShowHideControls.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Button = require('wix-style-react/dist/src/Button');

var _Button2 = _interopRequireDefault(_Button);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Grid = require('wix-style-react/dist/src/Grid');

var _Example = require('./Example.scss');

var _Example2 = _interopRequireDefault(_Example);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ShowHideControls = function (_React$Component) {
  _inherits(ShowHideControls, _React$Component);

  _createClass(ShowHideControls, [{
    key: 'toggle',
    value: function toggle() {
      this.setState({ show: !this.state.show });
    }
  }, {
    key: 'onChange',
    value: function onChange() {
      clearInterval(this.state.intervalId);
      this.toggle();
    }
  }]);

  function ShowHideControls(props) {
    _classCallCheck(this, ShowHideControls);

    var _this = _possibleConstructorReturn(this, (ShowHideControls.__proto__ || Object.getPrototypeOf(ShowHideControls)).call(this, props));

    _this.state = {
      show: true
    };

    return _this;
  }

  _createClass(ShowHideControls, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      if (!this.props.interval) {
        return;
      }
      var intervalId = setInterval(function () {
        _this2.toggle();
      }, 5000);

      this.setState({ intervalId: intervalId });
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      clearInterval(this.state.intervalId);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      var _props = this.props,
          children = _props.children,
          height = _props.height;

      var styles = height ? { height: height } : {};
      var show = this.state.show;

      return _react2.default.createElement(
        'div',
        { className: _Example2.default.showHideWrapper, style: styles },
        _react2.default.createElement(
          _Grid.Container,
          null,
          _react2.default.createElement(
            _Grid.Row,
            null,
            _react2.default.createElement(
              _Grid.Col,
              { span: '2' },
              _react2.default.createElement(
                _Button2.default,
                { theme: show ? 'transparentblue' : 'fullblue', onClick: function onClick() {
                    return _this3.onChange();
                  } },
                show ? 'Hide' : 'Show',
                ' Animation'
              )
            ),
            _react2.default.createElement(
              _Grid.Col,
              { span: '10' },
              _react2.default.createElement(children.type, _extends({}, children.props, { show: this.state.show }), children)
            )
          )
        )
      );
    }
  }]);

  return ShowHideControls;
}(_react2.default.Component);

ShowHideControls.propTypes = {
  children: _propTypes2.default.element.isRequired,
  interval: _propTypes2.default.number,
  height: _propTypes2.default.number
};

exports.default = ShowHideControls;
