// wix-animations/dist/stories/utils/Components/TabbedView/TabbedView.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _Tabs = require('../../../../src/Tabs/Tabs');

var _Tabs2 = _interopRequireDefault(_Tabs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var TabbedView = function (_Component) {
  _inherits(TabbedView, _Component);

  function TabbedView(props) {
    _classCallCheck(this, TabbedView);

    var _this = _possibleConstructorReturn(this, (TabbedView.__proto__ || Object.getPrototypeOf(TabbedView)).call(this, props));

    _this.onTabClicked = function (tab) {
      _this.setState({ activeTabId: tab.id });
    };

    _this.getTabsProps = function () {
      return {
        items: _this.props.tabs.map(function (tab) {
          return {
            title: tab,
            id: tab
          };
        }),
        onClick: _this.onTabClicked
      };
    };

    _this.state = { activeTabId: props.tabs[0] };
    _this.tabsProps = _this.getTabsProps();
    return _this;
  }

  _createClass(TabbedView, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var shouldHideForE2E = process.env.STORYBOOK_E2E;

      return _react2.default.createElement(
        'div',
        null,
        !shouldHideForE2E ? _react2.default.createElement(_Tabs2.default, _extends({ activeId: this.state.activeTabId }, this.tabsProps)) : null,
        this.props.children.map(function (child, index) {
          return _this2.state.activeTabId === _this2.props.tabs[index] ? child : null;
        })
      );
    }
  }]);

  return TabbedView;
}(_react.Component);

TabbedView.propTypes = {
  tabs: _propTypes2.default.arrayOf(_propTypes2.default.string),
  children: _propTypes2.default.arrayOf(_propTypes2.default.node)
};
exports.default = TabbedView;
