// wix-ui-test-utils/dist/src/index.js
"use strict";
// DO NOT USE THIS FILE!
// This is file is deprecated, do not import from it.
// It's used for backward compatability in wix-style-react 1.*
Object.defineProperty(exports, "__esModule", { value: true });
var driver_factory_1 = require("./driver-factory");
exports.createDriverFactory = driver_factory_1.createDriverFactory;
var react_helpers_1 = require("./react-helpers");
exports.isClassExists = react_helpers_1.isClassExists;
exports.makeControlled = react_helpers_1.makeControlled;
exports.sleep = react_helpers_1.sleep;
var protractor_1 = require("./protractor");
exports.getStoryUrl = protractor_1.getStoryUrl;
exports.scrollToElement = protractor_1.scrollToElement;
exports.waitForVisibilityOf = protractor_1.waitForVisibilityOf;
exports.protractorTestkitFactoryCreator = protractor_1.protractorTestkitFactoryCreator;
var vanilla_1 = require("./vanilla");
exports.testkitFactoryCreator = vanilla_1.testkitFactoryCreator;
exports.isTestkitExists = vanilla_1.isTestkitExists;
var enzyme_1 = require("./enzyme");
exports.enzymeTestkitFactoryCreator = enzyme_1.enzymeTestkitFactoryCreator;
exports.isEnzymeTestkitExists = enzyme_1.isEnzymeTestkitExists;
var puppeteer_1 = require("./puppeteer");
exports.puppeteerTestkitFactoryCreator = puppeteer_1.puppeteerTestkitFactoryCreator;
var protractor_2 = require("./protractor");
exports.baseProtractorConfig = protractor_2.baseProtractorConfig;
