// wix-ui-core/dist/src/components/TimePicker/Tickers.js
"use strict";
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var prop_types_1 = require("prop-types");
var Tickers_st_css_1 = require("./Tickers.st.css");
exports.Tickers = function (props) { return (React.createElement("div", __assign({}, Tickers_st_css_1.default('root', {}, props)),
    React.createElement("button", { tabIndex: -1, type: "button", onClick: props.onIncrement, className: Tickers_st_css_1.default.ticker, "data-hook": "ticker-button-up" }, props.tickerUpIcon),
    React.createElement("button", { tabIndex: -1, type: "button", onClick: props.onDecrement, className: Tickers_st_css_1.default.ticker, "data-hook": "ticker-button-down" }, props.tickerDownIcon))); };
exports.Tickers.displayName = 'Tickers';
exports.Tickers.propTypes = {
    className: prop_types_1.string,
    /** increment handler */
    onIncrement: prop_types_1.func,
    /** decrement handler */
    onDecrement: prop_types_1.func,
    /** up ticker icon */
    tickerUpIcon: prop_types_1.node,
    /** down ticker icon */
    tickerDownIcon: prop_types_1.node
};
