// wix-ui-core/dist/src/components/TimePicker/TimePicker.driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Input_driver_1 = require("../../components/Input/Input.driver");
exports.timePickerDriverFactory = function (_a) {
    var element = _a.element, eventTrigger = _a.eventTrigger;
    var inputDriver = Input_driver_1.inputDriverFactory({ element: element, eventTrigger: eventTrigger });
    return {
        /** checks if element exists */
        exists: function () { return !!element; },
        /** returns the time picker element */
        element: function () { return element; },
        /** returns the input element */
        getInputElement: function () { return inputDriver.getInput(); },
        /** returns inline style width of TimePicker */
        getInlineStyle: function () { return window.getComputedStyle(inputDriver.getInput().parentNode); },
        /** returns a boolean indicating if the timePicker is disabled */
        isDisabled: function () { return inputDriver.isDisabled(); },
        /** returns the input type attribute */
        getInputType: function () { return inputDriver.getInput().type; },
        /** returns the value (corresponds to state.value) */
        getValue: function () { return inputDriver.getValue(); },
        /** sets the value of the time picker */
        setValue: function (value) { return inputDriver.setValue(value); },
        /** simulates a keyDown event on the input element */
        keyDown: function (key) { return inputDriver.keyDown(key); },
        /** simulates a focus event on the input element */
        focus: function () { return inputDriver.focus(); },
        /** simulates a blur event on the input element */
        blur: function () { return inputDriver.blur(); },
        /** returns the ticker group element */
        getTickers: function () { return inputDriver.getSuffix(); },
        /** clicks on the ticker up button */
        clickTickerUp: function () { return eventTrigger.click(inputDriver.getSuffix().querySelector('[data-hook="ticker-button-up"]')); },
        /** clicks on the ticker down button */
        clickTickerDown: function () { return eventTrigger.click(inputDriver.getSuffix().querySelector('[data-hook="ticker-button-down"]')); }
    };
};
