// wix-style-react/dist/stories/Range/RangeTemplate.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactElementToJsxString = require('react-element-to-jsx-string');

var _reactElementToJsxString2 = _interopRequireDefault(_reactElementToJsxString);

var _Range = require('../../src/Range');

var _Range2 = _interopRequireDefault(_Range);

var _DatePicker = require('../../src/DatePicker');

var _DatePicker2 = _interopRequireDefault(_DatePicker);

var _Input = require('../../src/Input');

var _Input2 = _interopRequireDefault(_Input);

var _Label = require('../../src/Label');

var _Label2 = _interopRequireDefault(_Label);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Form = function (_Component) {
  _inherits(Form, _Component);

  function Form() {
    _classCallCheck(this, Form);

    return _possibleConstructorReturn(this, (Form.__proto__ || Object.getPrototypeOf(Form)).apply(this, arguments));
  }

  _createClass(Form, [{
    key: 'componentDidUpdate',
    value: function componentDidUpdate(props) {
      props.onChange((0, _reactElementToJsxString2.default)(this.getComponent()));
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.props.onChange((0, _reactElementToJsxString2.default)(this.getComponent()));
    }
  }, {
    key: 'getComponent',
    value: function getComponent() {
      return _react2.default.createElement(
        _Range2.default,
        { dataHook: this.props.dataHook, required: this.props.required, info: this.props.info },
        this.props.withLabel ? _react2.default.createElement(_Label2.default, this.props.label) : null,
        this.props.rangeType.value === 'InputRange' ? _react2.default.createElement(_Input2.default, _extends({ dataHook: 'first-item', id: 'first' }, this.props.firstInput)) : _react2.default.createElement(_DatePicker2.default, _extends({ dataHook: 'first-item', placeholderText: 'From', id: 'fromDate' }, this.props.firstDate)),
        this.props.rangeType.value === 'InputRange' ? _react2.default.createElement(_Input2.default, _extends({ dataHook: 'last-item', id: 'last' }, this.props.lastInput)) : _react2.default.createElement(_DatePicker2.default, _extends({ dataHook: 'last-item', placeholderText: 'To', id: 'toDate' }, this.props.lastDate))
      );
    }
  }, {
    key: 'render',
    value: function render() {
      return this.getComponent();
    }
  }]);

  return Form;
}(_react.Component);

Form.propTypes = {
  onChange: _propTypes2.default.func.isRequired,
  withLabel: _propTypes2.default.bool,
  label: _propTypes2.default.object,
  firstInput: _propTypes2.default.object,
  lastInput: _propTypes2.default.object,
  firstDate: _propTypes2.default.object,
  lastDate: _propTypes2.default.object,
  required: _propTypes2.default.bool,
  info: _propTypes2.default.string,
  rangeType: _propTypes2.default.object,
  dataHook: _propTypes2.default.string
};
exports.default = Form;
