// wix-style-react/dist/src/Input/Ticker/testkit/Ticker.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.tickerTestkitFactory = exports.componentFactory = exports.tickerDriverFactory = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _testUtils = require('react-dom/test-utils');

var _Ticker = require('../Ticker');

var _Ticker2 = _interopRequireDefault(_Ticker);

var _Ticker3 = require('../Ticker.scss');

var _Ticker4 = _interopRequireDefault(_Ticker3);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var tickerDriverFactory = exports.tickerDriverFactory = function tickerDriverFactory(component) {
  var handlers = {
    getUp: function getUp() {
      return component.querySelector('.' + _Ticker4.default.up);
    },
    getDown: function getDown() {
      return component.querySelector('.' + _Ticker4.default.down);
    },
    clickUp: function clickUp() {
      return _testUtils.Simulate.click(handlers.getUp());
    },
    clickDown: function clickDown() {
      return _testUtils.Simulate.click(handlers.getDown());
    },
    isUpDisabled: function isUpDisabled() {
      return handlers.getUp().classList.contains(_Ticker4.default.disabled);
    },
    isDownDisabled: function isDownDisabled() {
      return handlers.getDown().classList.contains(_Ticker4.default.disabled);
    },
    exists: function exists() {
      return !!component;
    }
  };
  return handlers;
};

var componentFactory = exports.componentFactory = function componentFactory() {
  var props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _testUtils.renderIntoDocument)(_react2.default.createElement(
    'div',
    null,
    _react2.default.createElement(_Ticker2.default, props)
  )).childNodes[0];
};

var tickerTestkitFactory = exports.tickerTestkitFactory = function tickerTestkitFactory(_ref) {
  var wrapper = _ref.wrapper;
  return tickerDriverFactory(wrapper.querySelector('[data-hook=ticker]'));
};
