// wix-style-react/dist/stories/InputArea/ExampleSizes.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _InputArea = require('wix-style-react/InputArea');

var _InputArea2 = _interopRequireDefault(_InputArea);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var style = {
  display: 'inline-block',
  padding: '0 5px',
  width: '200px',
  lineHeight: '22px',
  verticalAlign: 'top'
};

var defaultProps = {
  size: 'normal',
  magnifyingGlass: true,
  placeholder: 'They did not know it was impossible, so they did it!',
  unit: '$'
};

exports.default = function () {
  return _react2.default.createElement(
    'div',
    { className: 'ltr' },
    _react2.default.createElement(
      'div',
      { style: style },
      'Rows(3) & Max Height & min height',
      _react2.default.createElement('br', null),
      _react2.default.createElement(_InputArea2.default, _extends({}, defaultProps, { rows: 3, maxHeight: '120px', minHeight: '50px', resizable: true })),
      _react2.default.createElement('br', null)
    ),
    _react2.default.createElement(
      'div',
      { style: style },
      'Fixed Size',
      _react2.default.createElement('br', null),
      _react2.default.createElement(_InputArea2.default, defaultProps),
      _react2.default.createElement('br', null)
    )
  );
};
