// wix-style-react/dist/stories/InputArea/ExampleControlled.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _InputArea = require('wix-style-react/InputArea');

var _InputArea2 = _interopRequireDefault(_InputArea);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var style = {
  display: 'inline-block',
  padding: '0 5px',
  width: '200px',
  lineHeight: '22px'
};

var ControlledInputArea = function (_Component) {
  _inherits(ControlledInputArea, _Component);

  function ControlledInputArea(_ref) {
    var _ref$value = _ref.value,
        value = _ref$value === undefined ? '' : _ref$value;

    _classCallCheck(this, ControlledInputArea);

    var _this = _possibleConstructorReturn(this, (ControlledInputArea.__proto__ || Object.getPrototypeOf(ControlledInputArea)).call(this));

    _this.state = { value: value };
    return _this;
  }

  _createClass(ControlledInputArea, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var onChange = function onChange(event) {
        return _this2.setState({ value: event.target.value });
      };
      var onClear = function onClear() {
        _this2.setState({ value: '' });
        _this2.refs.inputtest.focus();
      };

      return _react2.default.createElement(_InputArea2.default, _extends({}, this.props, { error: this.state.value === 'Starwars', value: this.state.value, onChange: onChange, onClear: onClear }));
    }
  }]);

  return ControlledInputArea;
}(_react.Component);

ControlledInputArea.propTypes = {
  value: _propTypes2.default.bool
};

exports.default = function () {
  return _react2.default.createElement(
    'div',
    { className: 'ltr', style: style },
    _react2.default.createElement(
      'span',
      null,
      'This shows an error for "Starwars"'
    ),
    _react2.default.createElement(ControlledInputArea, null)
  );
};
