// wix-style-react/dist/stories/InputArea/index.story.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AUTO_EXAMPLE_DATA_HOOK = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _CodeExample = require('wix-storybook-utils/CodeExample');

var _CodeExample2 = _interopRequireDefault(_CodeExample);

var _InputArea = require('wix-style-react/InputArea');

var _InputArea2 = _interopRequireDefault(_InputArea);

var _ExampleStandard = require('./ExampleStandard');

var _ExampleStandard2 = _interopRequireDefault(_ExampleStandard);

var _ExampleStandard3 = require('!raw-loader!./ExampleStandard');

var _ExampleStandard4 = _interopRequireDefault(_ExampleStandard3);

var _ExampleError = require('./ExampleError');

var _ExampleError2 = _interopRequireDefault(_ExampleError);

var _ExampleError3 = require('!raw-loader!./ExampleError');

var _ExampleError4 = _interopRequireDefault(_ExampleError3);

var _ExamplePaneltitleStyle = require('./ExamplePaneltitleStyle');

var _ExamplePaneltitleStyle2 = _interopRequireDefault(_ExamplePaneltitleStyle);

var _ExamplePaneltitleStyle3 = require('!raw-loader!./ExamplePaneltitleStyle');

var _ExamplePaneltitleStyle4 = _interopRequireDefault(_ExamplePaneltitleStyle3);

var _ExampleControlled = require('./ExampleControlled');

var _ExampleControlled2 = _interopRequireDefault(_ExampleControlled);

var _ExampleControlled3 = require('!raw-loader!./ExampleControlled');

var _ExampleControlled4 = _interopRequireDefault(_ExampleControlled3);

var _ExampleRefs = require('./ExampleRefs');

var _ExampleRefs2 = _interopRequireDefault(_ExampleRefs);

var _ExampleRefs3 = require('!raw-loader!./ExampleRefs');

var _ExampleRefs4 = _interopRequireDefault(_ExampleRefs3);

var _ExampleSizes = require('./ExampleSizes');

var _ExampleSizes2 = _interopRequireDefault(_ExampleSizes);

var _ExampleSizes3 = require('!raw-loader!./ExampleSizes');

var _ExampleSizes4 = _interopRequireDefault(_ExampleSizes3);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var AUTO_EXAMPLE_DATA_HOOK = exports.AUTO_EXAMPLE_DATA_HOOK = 'storybook-inputarea';

exports.default = {
  category: '3. Inputs',
  storyName: '3.2 + InputArea',
  component: _InputArea2.default,
  componentPath: '../../src/InputArea',

  componentProps: {
    dataHook: AUTO_EXAMPLE_DATA_HOOK,
    error: false,
    theme: 'normal',
    hasCounter: false,
    resizable: false
  },

  examples: _react2.default.createElement(
    'div',
    null,
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Standard', code: _ExampleStandard4.default },
      _react2.default.createElement(_ExampleStandard2.default, null)
    ),
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Error', code: _ExampleError4.default },
      _react2.default.createElement(_ExampleError2.default, null)
    ),
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'PaneltitleStyle', code: _ExamplePaneltitleStyle4.default },
      _react2.default.createElement(_ExamplePaneltitleStyle2.default, null)
    ),
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Controlled input', code: _ExampleControlled4.default },
      _react2.default.createElement(_ExampleControlled2.default, null)
    ),
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Sizes', code: _ExampleSizes4.default },
      _react2.default.createElement(_ExampleSizes2.default, null)
    ),
    _react2.default.createElement(
      _CodeExample2.default,
      { title: 'Commands test', code: _ExampleRefs4.default },
      _react2.default.createElement(_ExampleRefs2.default, null)
    )
  )
};
