// wix-ui-core/dist/src/components/RadioButton/RadioButton.driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var RadioButton_st_css_1 = require("./RadioButton.st.css");
var test_utils_1 = require("stylable/test-utils");
exports.radioButtonDriverFactory = function (_a) {
    var element = _a.element, eventTrigger = _a.eventTrigger;
    var domUtils = new test_utils_1.StylableDOMUtil(RadioButton_st_css_1.default, element);
    var getInput = function () { return domUtils.select('.hiddenRadio'); };
    var getIcon = function () { return domUtils.select('.icon'); };
    var getLabel = function () { return domUtils.select('.label'); };
    return {
        exists: function () { return !!element; },
        select: function () { return eventTrigger.change(element, { target: element }); },
        click: function () { return eventTrigger.click(element); },
        value: function () { return getInput().getAttribute('value'); },
        name: function () { return getInput().getAttribute('name'); },
        isInputFocused: function () { return document.activeElement === getInput(); },
        isRequired: function () { return getInput().hasAttribute('required'); },
        iconExists: function () { return !!getIcon(); },
        labelExists: function () { return !!getLabel(); },
        isChecked: function () { return domUtils.hasStyleState(element, 'checked'); },
        isFocused: function () { return domUtils.hasStyleState(element, 'focused'); },
        isDisabled: function () { return domUtils.hasStyleState(element, 'disabled'); },
        isFocusVisible: function () { return domUtils.hasStyleState(element, 'focus-visible'); }
    };
};
