// wix-style-react/dist/src/Tabs/core/TabItem/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _pick = require('lodash/pick');

var _pick2 = _interopRequireDefault(_pick);

var _withTooltip = require('../withTooltip');

var _withTooltip2 = _interopRequireDefault(_withTooltip);

var _tabPropTypes = require('../constants/tab-prop-types');

var TabPropTypes = _interopRequireWildcard(_tabPropTypes);

var _Tabs = require('../../Tabs.scss');

var _Tabs2 = _interopRequireDefault(_Tabs);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var TabItem = function (_React$Component) {
  _inherits(TabItem, _React$Component);

  function TabItem() {
    _classCallCheck(this, TabItem);

    return _possibleConstructorReturn(this, (TabItem.__proto__ || Object.getPrototypeOf(TabItem)).apply(this, arguments));
  }

  _createClass(TabItem, [{
    key: 'getItemStyle',
    value: function getItemStyle() {
      var _props = this.props,
          type = _props.type,
          width = _props.width,
          itemMaxWidth = _props.itemMaxWidth;


      return {
        uniformSide: { width: width },
        compactSide: itemMaxWidth ? { maxWidth: itemMaxWidth + 'pt' } : {}
      }[type] || {};
    }
  }, {
    key: 'render',
    value: function render() {
      var _props2 = this.props,
          item = _props2.item,
          onItemClick = _props2.onItemClick,
          isActive = _props2.isActive,
          initHasTooltip = _props2.initHasTooltip,
          dynamicProperties = _props2.dynamicProperties;


      var containerProps = _extends({
        key: item.id,
        onClick: function onClick() {
          return onItemClick(item);
        },
        className: (0, _classnames2.default)(_Tabs2.default.tab, _defineProperty({}, _Tabs2.default.active, isActive)),
        style: this.getItemStyle()
      }, (0, _pick2.default)(this.props, dynamicProperties));

      return _react2.default.createElement(
        'li',
        _extends({}, containerProps, { 'data-hook': item.dataHook }),
        _react2.default.createElement(
          'span',
          { ref: function ref(el) {
              return initHasTooltip(el);
            } },
          item.title
        )
      );
    }
  }]);

  return TabItem;
}(_react2.default.Component);

TabItem.propTypes = {
  itemMaxWidth: _propTypes2.default.number,
  isActive: _propTypes2.default.bool,
  item: TabPropTypes.item.isRequired,
  onItemClick: TabPropTypes.onClick,
  type: TabPropTypes.type,
  width: TabPropTypes.width,
  initHasTooltip: _propTypes2.default.func.isRequired,
  dynamicProperties: _propTypes2.default.array
};

exports.default = (0, _withTooltip2.default)(TabItem);
