// wix-ui-backoffice/dist/src/components/Badge/Badge.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var prop_types_1 = require("prop-types");
var StylableBadge_1 = require("wix-ui-core/StylableBadge");
var withStylable_1 = require("wix-ui-core/withStylable");
var StylableUIText_1 = require("../StylableUIText");
var constants_1 = require("./constants");
var Badge_st_css_1 = require("./Badge.st.css");
var defaultProps = {
    type: constants_1.TYPE.solid,
    skin: constants_1.SKIN.general
};
var StyledBadge = withStylable_1.withStylable(StylableBadge_1.Badge, Badge_st_css_1.default, function (_a) {
    var skin = _a.skin, type = _a.type;
    return ({ skin: skin, type: type });
}, defaultProps);
var Badge = /** @class */ (function (_super) {
    __extends(Badge, _super);
    function Badge() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Badge.prototype.render = function () {
        var _a = this.props, children = _a.children, prefixIcon = _a.prefixIcon, suffixIcon = _a.suffixIcon, rest = __rest(_a, ["children", "prefixIcon", "suffixIcon"]);
        return (React.createElement(StyledBadge, __assign({}, rest),
            prefixIcon && React.cloneElement(prefixIcon, { className: Badge_st_css_1.default.prefix }),
            React.createElement(StylableUIText_1.UIText, { className: Badge_st_css_1.default.text, appearance: "T5" }, children),
            suffixIcon && React.cloneElement(suffixIcon, { className: Badge_st_css_1.default.suffix })));
    };
    Badge.displayName = 'Badge';
    Badge.propTypes = __assign({}, StylableBadge_1.Badge.propTypes, { 
        /** Type of the badge */
        type: prop_types_1.oneOf(Object.keys(constants_1.TYPE)), 
        /** Skin of the badge */
        skin: prop_types_1.oneOf(Object.keys(constants_1.SKIN)), 
        /** The prefix icon of the badge */
        prefixIcon: prop_types_1.node, 
        /** The suffix icon of the badge */
        suffixIcon: prop_types_1.node });
    Badge.defaultProps = defaultProps;
    return Badge;
}(React.PureComponent));
exports.Badge = Badge;
