// wix-ui-core/dist/src/components/Popover/Popover.driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var queryDocumentOrElement = function (element, query) { return ((element && element.querySelectorAll(query)[0]) || document && document.querySelector(query)); };
var getTargetElement = function (element) { return element && element.querySelectorAll('[data-hook="popover-element"]')[0]; };
var getContentElement = function (element) { return queryDocumentOrElement(element, '[data-hook="popover-content"]'); };
var getArrowElement = function (element) { return element && element.querySelectorAll('[data-hook="popover-arrow"]')[0]; };
exports.popoverDriverFactory = function (_a) {
    var element = _a.element, eventTrigger = _a.eventTrigger;
    return ({
        exists: function () { return !!element; },
        getTargetElement: function () { return getTargetElement(element); },
        getContentElement: function () { return getContentElement(element); },
        isTargetElementExists: function () { return !!getTargetElement(element); },
        isContentElementExists: function () { return !!getContentElement(element); },
        mouseEnter: function () { return eventTrigger.mouseEnter(element); },
        mouseLeave: function () { return eventTrigger.mouseLeave(element); },
        click: function () { return eventTrigger.click(element); },
        getArrowOffset: function () {
            var _a = getArrowElement(element).style, top = _a.top, left = _a.left, right = _a.right, bottom = _a.bottom;
            return { top: top, left: left, right: right, bottom: bottom };
        },
        inlineStyles: function () { return element.style; },
        getElementId: function () { return element.id; }
    });
};
