// wix-style-react/dist/src/EndorseContentLayout/EndorseContentLayout.driver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _enzyme = require('enzyme');

var _ = require('./');

var _2 = _interopRequireDefault(_);

var _styles = require('./styles.scss');

var _styles2 = _interopRequireDefault(_styles);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var EndorseContentLayoutDriver = function EndorseContentLayoutDriver() {
  var _this = this;

  _classCallCheck(this, EndorseContentLayoutDriver);

  this.when = {
    created: function created(props) {
      return _this.component = (0, _enzyme.shallow)(_react2.default.createElement(_2.default, props));
    }
  };
  this.get = {
    root: function root() {
      return _this.component.find('.' + _styles2.default.root);
    },
    head: function head() {
      return _this.component.find('.' + _styles2.default.head);
    },
    content: function content() {
      return _this.component.find('.' + _styles2.default.content);
    },
    primaryCta: function primaryCta() {
      return _this.component.find('.' + _styles2.default.primaryCta);
    },
    secondaryCta: function secondaryCta() {
      return _this.component.find('.' + _styles2.default.secondaryCta);
    }
  };
};

exports.default = EndorseContentLayoutDriver;
