// wix-style-react/dist/stories/Button/close.story.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Button = require('wix-style-react/Button');

var _Button2 = _interopRequireDefault(_Button);

var _newIcons = require('wix-style-react/new-icons');

var Icons = _interopRequireWildcard(_newIcons);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var icons = Object.values(Icons).map(function (icon) {
  return _react2.default.createElement(icon);
});

exports.default = {
  category: '5. Buttons',
  storyName: '5.7 Close',
  component: _Button2.default,
  componentPath: '../../src/Backoffice/Button',

  componentProps: {
    withNewIcons: true,
    theme: 'close-standard',
    children: _react2.default.createElement(Icons.X, null)
  },

  exampleProps: {
    onClick: function onClick() {
      return 'Clicked!';
    },
    onMouseEnter: function onMouseEnter() {
      return 'Mouse Enter!';
    },
    onMouseLeave: function onMouseLeave() {
      return 'Mouse Leave!';
    },
    children: icons
  }
};
