// wix-style-react/dist/stories/AddItem/Example.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _AddItem = require('../../src/AddItem/AddItem');

var _AddItem2 = _interopRequireDefault(_AddItem);

var _Grid = require('../../src/Grid');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Example = function Example() {
  return _react2.default.createElement(
    _Grid.Container,
    null,
    _react2.default.createElement(
      _Grid.Row,
      null,
      _react2.default.createElement(
        _Grid.Col,
        { span: 2 },
        _react2.default.createElement(
          'h3',
          null,
          'Add Item - default'
        ),
        _react2.default.createElement(_AddItem2.default, { dataHook: 'add-item', onAddImage: function onAddImage() {} })
      ),
      _react2.default.createElement(
        _Grid.Col,
        { span: 4 },
        _react2.default.createElement(
          'h3',
          null,
          'Add Item - with tooltip & ratio 16/9 '
        ),
        _react2.default.createElement(_AddItem2.default, {
          aspectRatio: '16/9',
          dataHook: 'add-item',
          tooltipContent: 'Add Tooltip Content',
          onAddImage: function onAddImage() {}
        })
      )
    )
  );
};

Example.propTypes = {
  theme: _propTypes2.default.string
};

exports.default = Example;
