// wix-style-react/dist/src/Search/Search.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _InputWithOptions = require('../InputWithOptions');

var _InputWithOptions2 = _interopRequireDefault(_InputWithOptions);

var _Search = require('wix-ui-icons-common/Search');

var _Search2 = _interopRequireDefault(_Search);

var _WixComponent2 = require('../BaseComponents/WixComponent');

var _WixComponent3 = _interopRequireDefault(_WixComponent2);

var _Search3 = require('./Search.scss');

var _Search4 = _interopRequireDefault(_Search3);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/**
 * Search component with suggestions based on input value listed in dropdown
 */
var Search = function (_WixComponent) {
  _inherits(Search, _WixComponent);

  function Search(props) {
    _classCallCheck(this, Search);

    var _this = _possibleConstructorReturn(this, (Search.__proto__ || Object.getPrototypeOf(Search)).call(this, props));

    _this._onChange = function (e) {
      if (_this._isControlled) {
        _this.props.onChange(e);
      } else {
        _this.setState({
          inputValue: e.target.value
        });
      }
    };

    _this._onClear = function () {
      var onClear = _this.props.onClear;


      if (!_this.state.collapsed && _this.props.expandable) {
        _this.setState({
          collapsed: true
        });
      }

      onClear && onClear();
    };

    _this._currentValue = function () {
      var value = void 0;

      if (_this._isControlled) {
        value = _this.props.value;
      } else {
        value = _this.state.inputValue;
      }

      return value;
    };

    _this._onBlur = function () {
      var onBlur = _this.props.onBlur;


      if (!_this.state.collapsed && _this.props.expandable) {
        var value = _this._currentValue();

        if (value === '') {
          _this.setState({
            collapsed: true
          });
        }
      }

      onBlur && onBlur();
    };

    _this._onWrapperClick = function () {
      if (_this.props.expandable && _this.state.collapsed) {
        _this.refs.searchInput.input.focus();
        _this.setState({ collapsed: false });
      }
    };

    _this._onWrapperMouseDown = function (e) {
      // We need to capture mouse down and prevent it's event if the input
      // is already open
      if (_this.props.expandable && !_this.state.collapsed) {
        var value = _this._currentValue();

        if (value === '') {
          e.preventDefault();
        }
      }
    };

    var initialValue = !_this._isControlled && props.defaultValue || '';

    _this.state = {
      inputValue: initialValue,
      collapsed: props.expandable && initialValue === '' && !props.autoFocus
    };
    return _this;
  }

  _createClass(Search, [{
    key: 'render',
    value: function render() {
      var _classNames;

      var wrapperClasses = (0, _classnames2.default)((_classNames = {}, _defineProperty(_classNames, _Search4.default.expandableStyles, this.props.expandable), _defineProperty(_classNames, _Search4.default.collapsed, this.state.collapsed && this.props.expandable), _defineProperty(_classNames, _Search4.default.expanded, !this.state.collapsed && this.props.expandable), _classNames));

      return _react2.default.createElement(
        'div',
        { className: wrapperClasses, onClick: this._onWrapperClick, onMouseDown: this._onWrapperMouseDown },
        _react2.default.createElement(_InputWithOptions2.default, _extends({}, this.props, {
          ref: 'searchInput',
          roundInput: true,
          prefix: _react2.default.createElement(
            'div',
            { className: _Search4.default.leftIcon },
            _react2.default.createElement(_Search2.default, null)
          ),
          menuArrow: false,
          clearButton: true,
          closeOnSelect: true,
          showOptionsIfEmptyInput: false,
          options: this._filteredOptions,
          onClear: this._onClear,
          onChange: this._onChange,
          onBlur: this._onBlur,
          highlight: true
        }))
      );
    }
  }, {
    key: '_isControlled',
    get: function get() {
      return 'value' in this.props && 'onChange' in this.props;
    }
  }, {
    key: '_filteredOptions',
    get: function get() {
      var _props = this.props,
          options = _props.options,
          value = _props.value;


      var searchText = this._isControlled ? value : this.state.inputValue;

      return options.filter(function (option) {
        return searchText && searchText.length ? new RegExp(searchText.trim(), 'i').test(option.value) : true;
      });
    }
  }]);

  return Search;
}(_WixComponent3.default);

Search.displayName = 'Search';
Search.propTypes = _extends({}, _InputWithOptions2.default.propTypes, {
  placeholder: _propTypes2.default.string,
  expandable: _propTypes2.default.bool
});
Search.defaultProps = _extends({}, _InputWithOptions2.default.defaultProps, {
  placeholder: 'Search',
  expandable: false
});
exports.default = Search;
