// wix-style-react/dist/src/Search/Search.protractor.driver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _InputProtractor = require('../Input/Input.protractor.driver');

var _InputProtractor2 = _interopRequireDefault(_InputProtractor);

var _DropdownProtractor = require('../Dropdown/Dropdown.protractor.driver');

var _DropdownProtractor2 = _interopRequireDefault(_DropdownProtractor);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function (component) {
  var inputDriver = (0, _InputProtractor2.default)(component);
  var dropdownDriver = (0, _DropdownProtractor2.default)(component);

  return _extends({}, inputDriver, {
    element: function element() {
      return component;
    },
    exists: function exists() {
      return component.isPresent();
    },
    clickOnInput: function clickOnInput() {
      return dropdownDriver.getInput().click();
    },
    getInput: dropdownDriver.getInput,
    getSearchDropdown: dropdownDriver.getDropdown,
    getSearchOptionAt: dropdownDriver.getDropdownItem,
    clickSearchOptionAt: function clickSearchOptionAt(index) {
      return dropdownDriver.getDropdownItem(index).click();
    },
    getSearchOptionsCount: dropdownDriver.getDropdownItemsCount
  });
};
