// wix-ui-core/dist/src/components/Button/Button.driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buttonDriverFactory = function (_a) {
    var element = _a.element, eventTrigger = _a.eventTrigger;
    var getButtonStyle = function () { return window.getComputedStyle(element); };
    return {
        exists: function () { return !!element; },
        click: function () { return eventTrigger.click(element); },
        getType: function () { return element.getAttribute('type'); },
        getTextContent: function () { return element.textContent; },
        isDisabled: function () { return element.getAttribute('disabled') === ''; },
        styles: {
            getMinWidth: function () { return getButtonStyle().minWidth; },
            getWidth: function () { return getButtonStyle().width; },
            getHeight: function () { return getButtonStyle().height; },
            getPadding: function () { return getButtonStyle().padding; },
            getBorderRadius: function () { return getButtonStyle().borderRadius; },
        }
    };
};
