// wix-ui-core/dist/src/components/LinearProgressBar/LinearProgressBar.driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var test_utils_1 = require("stylable/test-utils");
var LinearProgressBar_st_css_1 = require("./LinearProgressBar.st.css");
exports.linearProgressBarDriverFactory = function (_a) {
    var element = _a.element;
    var stylableDOMUtil = new test_utils_1.StylableDOMUtil(LinearProgressBar_st_css_1.default);
    var getElement = function (dataHook) { return element.querySelector("[data-hook=\"" + dataHook + "\"]"); };
    var getValue = function () { return !element ? null : getElement('progress-percentages').querySelector('span').innerHTML; };
    var driver = {
        exists: function () { return !!element; },
        getWidth: function () {
            var el = getElement('progressbar-foreground');
            return el ? el.style.width : '0';
        },
        isSuccessIconDisplayed: function () { return !!getElement('success-icon'); },
        isErrorIconDisplayed: function () { return !!getElement('error-icon'); },
        isPercentagesProgressDisplayed: function () { return !!getElement('progress-percentages'); },
        getValue: function () { return getValue(); },
        isCompleted: function () { return getValue() === '100'; },
        hasError: function () { return stylableDOMUtil.hasStyleState(element, 'error'); },
    };
    return driver;
};
