// wix-ui-core/dist/src/components/LinearProgressBar/LinearProgressBar.protractor.driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.linearProgressBarDriverFactory = function (element) {
    var findByDataHook = function (dataHook) { return element.$("[data-hook=\"" + dataHook + "\"]"); };
    var foregroundBar = function () { return findByDataHook('progressbar-foreground'); };
    var backgroundBar = function () { return findByDataHook('progressbar-background'); };
    var getElementWidth = function (e) { return e.getSize().then((function (size) { return size.width; })); };
    var progressIndication = function () { return findByDataHook('progress-indicator'); };
    return {
        element: function () { return element; },
        exists: function () { return element.isPresent(); },
        getForegroundBarWidth: function () { return getElementWidth(foregroundBar()); },
        getBackgroundBarWidth: function () { return getElementWidth(backgroundBar()); },
        isProgressIndicationDisplayed: function () { return progressIndication().isPresent(); },
        progressIndicationValue: function () { return progressIndication().getText(); }
    };
};
