// wix-style-react/dist/stories/RichTextAreaComposite/RichTextAreaCompositeTemplate.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _RichTextAreaComposite = require('../../src/RichTextAreaComposite');

var _RichTextAreaComposite2 = _interopRequireDefault(_RichTextAreaComposite);

var _RichTextArea = require('../../src/RichTextArea');

var _RichTextArea2 = _interopRequireDefault(_RichTextArea);

var _Label = require('../../src/Label');

var _Label2 = _interopRequireDefault(_Label);

var _reactElementToJsxString = require('react-element-to-jsx-string');

var _reactElementToJsxString2 = _interopRequireDefault(_reactElementToJsxString);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Form = function (_Component) {
  _inherits(Form, _Component);

  function Form() {
    _classCallCheck(this, Form);

    return _possibleConstructorReturn(this, (Form.__proto__ || Object.getPrototypeOf(Form)).apply(this, arguments));
  }

  _createClass(Form, [{
    key: 'componentDidUpdate',
    value: function componentDidUpdate(props) {
      props.onChange((0, _reactElementToJsxString2.default)(this.getComponent()));
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.props.onChange((0, _reactElementToJsxString2.default)(this.getComponent()));
    }
  }, {
    key: 'getComponent',
    value: function getComponent() {
      return _react2.default.createElement(
        _RichTextAreaComposite2.default,
        { required: this.props.required, info: this.props.info },
        this.props.withLabel ? _react2.default.createElement(_Label2.default, this.props.label) : null,
        _react2.default.createElement(_RichTextArea2.default, this.props.richTextArea)
      );
    }
  }, {
    key: 'render',
    value: function render() {
      return this.getComponent();
    }
  }]);

  return Form;
}(_react.Component);

Form.propTypes = {
  withLabel: _propTypes2.default.bool,
  label: _propTypes2.default.object,
  richTextArea: _propTypes2.default.object,
  onChange: _propTypes2.default.func,
  required: _propTypes2.default.bool,
  info: _propTypes2.default.string
};
exports.default = Form;
