// wix-style-react/dist/src/Composite/InputAreaWithLabelComposite/InputAreaWithLabelComposite.driver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _FieldLabelAttributes = require('../../FieldLabelAttributes/FieldLabelAttributes.driver');

var _FieldLabelAttributes2 = _interopRequireDefault(_FieldLabelAttributes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var inputAreaWithLabelCompositeDriverFactory = function inputAreaWithLabelCompositeDriverFactory(_ref) {
  var element = _ref.element,
      wrapper = _ref.wrapper,
      component = _ref.component;

  var label = element.childNodes[0].childNodes[0];
  return {
    exists: function exists() {
      return !!element;
    },
    getLabel: function getLabel() {
      return label.textContent;
    },
    hasLabel: function hasLabel() {
      return label.tagName.toLowerCase() === 'label';
    },
    getAttr: function getAttr(attrName) {
      return element.getAttribute(attrName);
    },
    getNumberOfChildren: function getNumberOfChildren() {
      return element.childElementCount;
    },
    getInfoTooltipTestKit: function getInfoTooltipTestKit() {
      return (0, _FieldLabelAttributes2.default)({ wrapper: wrapper, element: element.querySelector('[data-hook="field-label-attributes"]') }).getTooltipTestKit();
    },
    hasFieldLabelAttributes: function hasFieldLabelAttributes() {
      return !!element.querySelectorAll('[data-hook="field-label-attributes"]').length;
    },
    setProps: function setProps(props) {
      var ClonedWithProps = _react2.default.cloneElement.apply(_react2.default, [component, Object.assign({}, component.props, props)].concat(_toConsumableArray(component.props.children || [])));
      _reactDom2.default.render(_react2.default.createElement(
        'div',
        { ref: function ref(r) {
            return element = r;
          } },
        ClonedWithProps
      ), wrapper);
    }
  };
};

exports.default = inputAreaWithLabelCompositeDriverFactory;
