// wix-style-react/dist/src/Composite/InputAreaWithLabelComposite/InputAreaWithLabelComposite.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _last = require('lodash/last');

var _last2 = _interopRequireDefault(_last);

var _WixComponent2 = require('../../BaseComponents/WixComponent');

var _WixComponent3 = _interopRequireDefault(_WixComponent2);

var _InputAreaWithLabelComposite = require('./InputAreaWithLabelComposite.scss');

var _InputAreaWithLabelComposite2 = _interopRequireDefault(_InputAreaWithLabelComposite);

var _FieldLabelAttributes = require('../../FieldLabelAttributes/FieldLabelAttributes');

var _FieldLabelAttributes2 = _interopRequireDefault(_FieldLabelAttributes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var InputAreaWithLabelComposite = function (_WixComponent) {
  _inherits(InputAreaWithLabelComposite, _WixComponent);

  function InputAreaWithLabelComposite() {
    _classCallCheck(this, InputAreaWithLabelComposite);

    return _possibleConstructorReturn(this, (InputAreaWithLabelComposite.__proto__ || Object.getPrototypeOf(InputAreaWithLabelComposite)).apply(this, arguments));
  }

  _createClass(InputAreaWithLabelComposite, [{
    key: 'getFieldLabelAttributesComponent',
    value: function getFieldLabelAttributesComponent() {
      return _react2.default.createElement(_FieldLabelAttributes2.default, {
        appendToParent: this.props.appendToParent,
        required: this.props.required,
        info: this.props.info,
        tooltip: this.props.tooltip
      });
    }
  }, {
    key: 'shouldShowFieldLabelAttributesComponent',
    value: function shouldShowFieldLabelAttributesComponent() {
      return this.props.required || this.props.info || this.props.tooltip;
    }
  }, {
    key: 'render',
    value: function render() {
      var children = _react.Children.toArray(this.props.children);
      return _react2.default.createElement(
        'div',
        null,
        children.length === 2 ? _react2.default.createElement(
          'div',
          { className: _InputAreaWithLabelComposite2.default.label },
          children[0],
          this.shouldShowFieldLabelAttributesComponent() ? this.getFieldLabelAttributesComponent() : null
        ) : null,
        children.length === 1 && this.shouldShowFieldLabelAttributesComponent() ? _react2.default.createElement(
          'div',
          { className: _InputAreaWithLabelComposite2.default.withLabelAttributes },
          _react2.default.createElement(
            'div',
            { className: _InputAreaWithLabelComposite2.default.inputWrapper },
            (0, _last2.default)(children)
          ),
          this.getFieldLabelAttributesComponent()
        ) : (0, _last2.default)(children)
      );
    }
  }]);

  return InputAreaWithLabelComposite;
}(_WixComponent3.default);

InputAreaWithLabelComposite.propTypes = {
  children: _propTypes.any,
  required: _propTypes.bool,
  appendToParent: _propTypes.bool,
  info: _propTypes.node,
  tooltip: _propTypes.node
};

InputAreaWithLabelComposite.defaultProps = {
  appendToParent: false
};

InputAreaWithLabelComposite.displayName = 'InputAreaWithLabelComposite';

exports.default = InputAreaWithLabelComposite;
