// wix-ui-core/dist/src/components/Tooltip/Tooltip.driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Popover_driver_1 = require("../Popover/Popover.driver");
exports.tooltipDriverFactory = function (args) {
    var popoverDriver = Popover_driver_1.popoverDriverFactory(args);
    var getTooltipStyle = function () { return window.getComputedStyle(args.element.querySelector('[data-hook="popover-content"]')); };
    return Object.assign({
        styles: {
            /** Gets background color */
            getBackgroundColor: function () { return getTooltipStyle().backgroundColor; },
            /** Gets border width */
            getBorderWidth: function () { return getTooltipStyle().borderWidth; },
            /** Gets border style */
            getBorderStyle: function () { return getTooltipStyle().borderStyle; },
            /** Gets border color */
            getBorderColor: function () { return getTooltipStyle().borderColor; },
            /** Gets border radius */
            getBorderRadius: function () { return getTooltipStyle().borderRadius; },
            /** Gets content padding */
            getContentPadding: function () { return getTooltipStyle().padding; }
        }
    }, popoverDriver);
};
