// wix-style-react/dist/src/Backoffice/Button/Button.driver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _testUtils = require('react-dom/test-utils');

var _testUtils2 = _interopRequireDefault(_testUtils);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

exports.default = function (_ref) {
  var element = _ref.element,
      wrapper = _ref.wrapper,
      component = _ref.component;
  return {
    exists: function exists() {
      return !!element;
    },
    click: function click() {
      return _testUtils2.default.Simulate.click(element);
    },
    focus: function focus() {
      return _testUtils2.default.Simulate.focus(element);
    },
    blur: function blur() {
      return _testUtils2.default.Simulate.blur(element);
    },
    mouseEnter: function mouseEnter() {
      return _testUtils2.default.Simulate.mouseEnter(element);
    },
    mouseLeave: function mouseLeave() {
      return _testUtils2.default.Simulate.mouseLeave(element);
    },
    getButtonTextContent: function getButtonTextContent() {
      return element.textContent;
    },
    isButtonDisabled: function isButtonDisabled() {
      return element.getAttribute('disabled') === '';
    },
    isPrefixIconExists: function isPrefixIconExists() {
      return element.innerHTML.indexOf('prefix') !== -1;
    },
    isSuffixIconExists: function isSuffixIconExists() {
      return element.innerHTML.indexOf('suffix') !== -1;
    },
    setProps: function setProps(props) {
      var ClonedWithProps = _react2.default.cloneElement.apply(_react2.default, [component, Object.assign({}, component.props, props)].concat(_toConsumableArray(component.props.children || [])));
      _reactDom2.default.render(_react2.default.createElement(
        'div',
        { ref: function ref(r) {
            return element = r;
          } },
        ClonedWithProps
      ), wrapper);
    }
  };
};
