// playable/dist/src/modules/ui/bottom-block/bottom-block.view.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var view_1 = tslib_1.__importDefault(require("../core/view"));
var htmlToElement_1 = tslib_1.__importDefault(require("../core/htmlToElement"));
var getElementByHook_1 = tslib_1.__importDefault(require("../core/getElementByHook"));
var toggleNodeClass_1 = tslib_1.__importDefault(require("../core/toggleNodeClass"));
var templates_1 = require("./templates");
var bottom_block_scss_1 = tslib_1.__importDefault(require("./bottom-block.scss"));
var BottomBlockView = /** @class */ (function (_super) {
    tslib_1.__extends(BottomBlockView, _super);
    function BottomBlockView(config) {
        var _this = _super.call(this) || this;
        var callbacks = config.callbacks, elements = config.elements;
        _this._callbacks = callbacks;
        _this._initDOM(elements);
        _this._bindEvents();
        return _this;
    }
    BottomBlockView.prototype._initDOM = function (elements) {
        this._$node = htmlToElement_1.default(templates_1.bottomBlockTemplate({
            styles: this.styleNames,
        }));
        var $playContainer = getElementByHook_1.default(this._$node, 'play-container');
        var $volumeContainer = getElementByHook_1.default(this._$node, 'volume-container');
        var $timeContainer = getElementByHook_1.default(this._$node, 'time-container');
        var $fullScreenContainer = getElementByHook_1.default(this._$node, 'full-screen-container');
        var $logoContainer = getElementByHook_1.default(this._$node, 'logo-container');
        var $progressBarContainer = getElementByHook_1.default(this._$node, 'progress-bar-container');
        $playContainer.appendChild(elements.play);
        $volumeContainer.appendChild(elements.volume);
        $timeContainer.appendChild(elements.time);
        $fullScreenContainer.appendChild(elements.fullScreen);
        $logoContainer.appendChild(elements.logo);
        $progressBarContainer.appendChild(elements.progress);
    };
    BottomBlockView.prototype._preventClickPropagation = function (e) {
        e.stopPropagation();
    };
    BottomBlockView.prototype._bindEvents = function () {
        this._$node.addEventListener('click', this._preventClickPropagation);
        this._$node.addEventListener('mousemove', this._callbacks.onBlockMouseMove);
        this._$node.addEventListener('mouseleave', this._callbacks.onBlockMouseOut);
    };
    BottomBlockView.prototype._unbindEvents = function () {
        this._$node.removeEventListener('click', this._preventClickPropagation);
        this._$node.removeEventListener('mousemove', this._callbacks.onBlockMouseMove);
        this._$node.removeEventListener('mouseleave', this._callbacks.onBlockMouseOut);
    };
    BottomBlockView.prototype.setShouldLogoShowAlwaysFlag = function (isShowAlways) {
        toggleNodeClass_1.default(this._$node, this.styleNames.showLogoAlways, isShowAlways);
        this.showLogo();
    };
    BottomBlockView.prototype.showPlayControl = function () {
        this._$node.classList.remove(this.styleNames.playControlHidden);
    };
    BottomBlockView.prototype.hidePlayControl = function () {
        this._$node.classList.add(this.styleNames.playControlHidden);
    };
    BottomBlockView.prototype.showTimeControl = function () {
        this._$node.classList.remove(this.styleNames.timeControlHidden);
    };
    BottomBlockView.prototype.hideTimeControl = function () {
        this._$node.classList.add(this.styleNames.timeControlHidden);
    };
    BottomBlockView.prototype.showVolumeControl = function () {
        this._$node.classList.remove(this.styleNames.volumeControlHidden);
    };
    BottomBlockView.prototype.hideVolumeControl = function () {
        this._$node.classList.add(this.styleNames.volumeControlHidden);
    };
    BottomBlockView.prototype.showFullScreenControl = function () {
        this._$node.classList.remove(this.styleNames.fullScreenControlHidden);
    };
    BottomBlockView.prototype.hideFullScreenControl = function () {
        this._$node.classList.add(this.styleNames.fullScreenControlHidden);
    };
    BottomBlockView.prototype.showLogo = function () {
        this._$node.classList.remove(this.styleNames.logoHidden);
    };
    BottomBlockView.prototype.hideLogo = function () {
        this._$node.classList.add(this.styleNames.logoHidden);
    };
    BottomBlockView.prototype.showProgressControl = function () {
        this._$node.classList.remove(this.styleNames.progressControlHidden);
    };
    BottomBlockView.prototype.hideProgressControl = function () {
        this._$node.classList.add(this.styleNames.progressControlHidden);
    };
    BottomBlockView.prototype.show = function () {
        this._$node.classList.remove(this.styleNames.hidden);
    };
    BottomBlockView.prototype.hide = function () {
        this._$node.classList.add(this.styleNames.hidden);
    };
    BottomBlockView.prototype.getNode = function () {
        return this._$node;
    };
    BottomBlockView.prototype.showContent = function () {
        this._$node.classList.add(this.styleNames.activated);
    };
    BottomBlockView.prototype.hideContent = function () {
        this._$node.classList.remove(this.styleNames.activated);
    };
    BottomBlockView.prototype.destroy = function () {
        this._unbindEvents();
        if (this._$node.parentNode) {
            this._$node.parentNode.removeChild(this._$node);
        }
        this._$node = null;
    };
    return BottomBlockView;
}(view_1.default));
BottomBlockView.extendStyleNames(bottom_block_scss_1.default);
exports.default = BottomBlockView;
