// wix-style-react/dist/stories/SideMenu/ExampleStandard.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _SideMenu = require('wix-style-react/SideMenu');

var _SideMenu2 = _interopRequireDefault(_SideMenu);

var _Button = require('wix-style-react/Button');

var _Button2 = _interopRequireDefault(_Button);

var _Tooltip = require('wix-style-react/Tooltip');

var _Tooltip2 = _interopRequireDefault(_Tooltip);

var _Edit = require('wix-style-react/new-icons/Edit');

var _Edit2 = _interopRequireDefault(_Edit);

var _Chat = require('wix-style-react/new-icons/Chat');

var _Chat2 = _interopRequireDefault(_Chat);

var _ExternalLink = require('wix-style-react/new-icons/ExternalLink');

var _ExternalLink2 = _interopRequireDefault(_ExternalLink);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function () {
  return _react2.default.createElement(
    'div',
    { style: { width: 220, height: 700 } },
    _react2.default.createElement(
      _SideMenu2.default,
      null,
      _react2.default.createElement(
        _SideMenu2.default.Header,
        null,
        _react2.default.createElement(
          'div',
          { style: { padding: '26px 30px', fontSize: '20px', color: 'white' } },
          _react2.default.createElement(
            _Tooltip2.default,
            {
              content: 'wix-style-react',
              placement: 'bottom',
              alignment: 'left',
              dataHook: 'site-name-tooltip',
              maxWidth: '250'
            },
            _react2.default.createElement(
              'div',
              { style: { display: 'flex' } },
              _react2.default.createElement(
                'span',
                { style: { fontSize: '20px' } },
                'wix-style-react'
              ),
              _react2.default.createElement(_ExternalLink2.default, null)
            )
          ),
          _react2.default.createElement(
            'div',
            { style: { marginTop: '5px', fontSize: '13px' } },
            'Role: Owner'
          )
        )
      ),
      _react2.default.createElement(
        _SideMenu2.default.Navigation,
        null,
        _react2.default.createElement(
          _SideMenu2.default.NavigationLink,
          { onClick: function onClick() {
              return console.log('Dashboard clicked');
            } },
          'Dashboard'
        ),
        _react2.default.createElement(
          _SideMenu2.default.NavigationLink,
          { onClick: function onClick() {
              return console.log('Rest. Menus clicked');
            } },
          'Rest. Menus'
        ),
        _react2.default.createElement(
          _SideMenu2.default.NavigationLink,
          { withArrow: true, onClick: function onClick() {
              return console.log('Rest. Products clicked');
            } },
          'Rest. Products'
        ),
        _react2.default.createElement(
          _SideMenu2.default.NavigationLink,
          { isActive: true, onClick: function onClick() {
              return console.log('Store Orders clicked');
            } },
          'Store Orders'
        ),
        _react2.default.createElement(
          _SideMenu2.default.NavigationLink,
          { withArrow: true, onClick: function onClick() {
              return console.log('Contacts & CRM clicked');
            } },
          'Contacts & CRM'
        ),
        _react2.default.createElement(
          _SideMenu2.default.NavigationLink,
          { withArrow: true, onClick: function onClick() {
              return console.log('Marketing Tools clicked');
            } },
          'Marketing Tools'
        ),
        _react2.default.createElement(_SideMenu2.default.NavigationSeparator, null),
        _react2.default.createElement(
          _SideMenu2.default.NavigationLink,
          { onClick: function onClick() {
              return console.log('Manage Website clicked');
            } },
          'Manage Website'
        ),
        _react2.default.createElement(
          _SideMenu2.default.NavigationLink,
          { badge: _react2.default.createElement(_SideMenu2.default.NavigationBadge, null), onClick: function onClick() {
              return console.log('Settings clicked');
            } },
          'Settings'
        ),
        _react2.default.createElement(
          _SideMenu2.default.NavigationLink,
          { onClick: function onClick() {
              return console.log('Apps clicked');
            } },
          'Apps'
        ),
        _react2.default.createElement(
          _SideMenu2.default.NavigationLink,
          {
            href: 'https://academy.wix.com/en/seo',
            target: '_blank',
            isDiminishedHover: true
          },
          'Dim Hover link #1'
        )
      ),
      _react2.default.createElement(
        _SideMenu2.default.Promotion,
        null,
        _react2.default.createElement(
          _Button2.default,
          {
            theme: 'fullpurple',
            onClick: function onClick() {
              return console.log('Promotion button clicked!');
            }
          },
          'Upgrade'
        )
      ),
      _react2.default.createElement(_SideMenu2.default.NavigationSeparator, null),
      _react2.default.createElement(
        _SideMenu2.default.Footer,
        null,
        _react2.default.createElement(
          _SideMenu2.default.FooterLink,
          {
            href: 'https://support.wix.com/',
            target: '_blank',
            icon: _react2.default.createElement(_Edit2.default, null)
          },
          'Edit'
        ),
        _react2.default.createElement(_SideMenu2.default.FooterTinyLink, {
          href: 'https://support.wix.com/en/article/wix-seo-wiz-suggestions-and-feedback',
          target: '_blank',
          icon: _react2.default.createElement(
            'div',
            { style: { marginTop: 2 } },
            _react2.default.createElement(_Chat2.default, null)
          ),
          tooltip: 'Hey, come talk to me!',
          onClick: function onClick() {
            return console.log('clicked on tiny link yay!');
          }
        })
      )
    )
  );
};
