// wix-ui-jss/dist/src/domStyleRenderer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var jss_1 = require("jss");
var jss_preset_default_1 = require("jss-preset-default");
var jss = jss_1.create(jss_preset_default_1.default());
if (process.env.NODE_ENV === 'test') {
    jss.setup({
        createGenerateClassName: function (args) {
            return function (rule, sheet) { return "" + rule.key; };
        }
    });
}
exports.sheetManager = new jss_1.SheetsManager();
exports.sheetMapper = {};
var atachStyleSheetToDom = function (styles, componentId) {
    var newSheet = jss.createStyleSheet(styles);
    if (exports.sheetMapper[componentId]) {
        detachStyleSheetFromDom(componentId);
    }
    exports.sheetMapper[componentId] = {
        styleElement: newSheet.renderer.element,
        styles: styles
    };
    exports.sheetManager.add(styles, newSheet);
    exports.sheetManager.manage(styles);
    return newSheet;
};
exports.generateClasses = function (styles, componentId) {
    var classes = atachStyleSheetToDom(styles, componentId).classes;
    return classes;
};
function detachStyleSheetFromDom(componentId) {
    exports.sheetManager.unmanage(exports.sheetMapper[componentId].styles);
    delete exports.sheetMapper[componentId];
}
exports.detachStyleSheetFromDom = detachStyleSheetFromDom;
