// wix-ui-core/dist/src/hocs/Focusable/FocusableHOC.driver.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var driver_factory_1 = require("wix-ui-test-utils/driver-factory");
var test_utils_1 = require("stylable/test-utils");
var FocusableHOC_1 = require("./FocusableHOC");
var Focusable_st_css_1 = require("./Focusable.st.css");
var stylableUtil = new test_utils_1.StylableDOMUtil(Focusable_st_css_1.default);
var hasFocusState = function (element) { return stylableUtil.hasStyleState(element, 'focus'); };
var hasFocusVisibleState = function (element) { return stylableUtil.hasStyleState(element, 'focus-visible'); };
var PureChildComponent = /** @class */ (function (_super) {
    __extends(PureChildComponent, _super);
    function PureChildComponent(props) {
        var _this = _super.call(this, props) || this;
        _this.unboundMethod = function () { return 'unboundMethod'; };
        _this.boundMethod = function () { return _this.id; };
        _this.id = props.id;
        _this.boundMethod = _this.boundMethod.bind(_this);
        return _this;
    }
    PureChildComponent.prototype.render = function () {
        return (React.createElement("div", __assign({ onFocus: this.props.focusableOnFocus, onBlur: this.props.focusableOnBlur }, Focusable_st_css_1.default('root', {}, this.props)), "Hello"));
    };
    PureChildComponent.staticVariable = 'staticVariable';
    PureChildComponent.staticMethod = function () { return 'staticMethod'; };
    return PureChildComponent;
}(React.PureComponent));
exports.PureChildComponent = PureChildComponent;
var focusableDriverFactory = function (_a) {
    var element = _a.element, eventTrigger = _a.eventTrigger;
    return {
        exists: function () { return true; },
        focus: function () { return eventTrigger.focus(element); },
        blur: function () { return eventTrigger.blur(element); },
        hasFocusState: function () { return hasFocusState(element); },
        hasFocusVisibleState: function () { return hasFocusVisibleState(element); }
    };
};
var driverFactory = driver_factory_1.createDriverFactory(focusableDriverFactory);
exports.createDriver = function (Component) {
    var driver = driverFactory(Component);
    var fireMouseDown = function () { return window.dispatchEvent(new window.Event('mousedown')); };
    var fireMouseUp = function () { return window.dispatchEvent(new window.Event('mouseup')); };
    var fireKeyDown = function () { return window.dispatchEvent(new window.Event('keydown')); };
    var fireKeyUp = function () { return window.dispatchEvent(new window.Event('keyup')); };
    var tabOut = function () {
        fireKeyDown();
        driver.blur();
        fireKeyUp();
    };
    var tabIn = function () {
        fireKeyDown();
        driver.focus();
        fireKeyUp();
    };
    var click = function () {
        fireMouseDown();
        driver.focus();
        fireMouseUp();
    };
    return __assign({}, driver, { fireMouseDown: fireMouseDown,
        fireKeyDown: fireKeyDown,
        fireKeyUp: fireKeyUp,
        tabOut: tabOut,
        tabIn: tabIn,
        click: click });
};
exports.WithFocusableComp = FocusableHOC_1.withFocusable(PureChildComponent);
