// wix-style-react/dist/stories/MultiSelectComposite/MultiSelectCompositeTemplate.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactElementToJsxString = require('react-element-to-jsx-string');

var _reactElementToJsxString2 = _interopRequireDefault(_reactElementToJsxString);

var _MultiSelectComposite = require('../../src/MultiSelectComposite');

var _MultiSelectComposite2 = _interopRequireDefault(_MultiSelectComposite);

var _MultiSelect = require('../../src/MultiSelect');

var _MultiSelect2 = _interopRequireDefault(_MultiSelect);

var _Label = require('../../src/Label');

var _Label2 = _interopRequireDefault(_Label);

var _ExampleStandard = require('../MultiSelect/ExampleStandard');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Form = function (_Component) {
  _inherits(Form, _Component);

  _createClass(Form, [{
    key: 'componentDidUpdate',
    value: function componentDidUpdate(props) {
      props.onChange((0, _reactElementToJsxString2.default)(this.getComponent()));
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.props.onChange((0, _reactElementToJsxString2.default)(this.getComponent()));
    }
  }]);

  function Form(props) {
    _classCallCheck(this, Form);

    var _this = _possibleConstructorReturn(this, (Form.__proto__ || Object.getPrototypeOf(Form)).call(this, props));

    _this.getValue = function (option) {
      return typeof option.value === 'string' ? option.value : option.value.props.children[0].props.children;
    };

    _this.handleOnSelect = function (tags) {
      return Array.isArray(tags) ? _this.setState({ tags: [].concat(_toConsumableArray(_this.state.tags), _toConsumableArray(tags)) }) : _this.setState({ tags: [].concat(_toConsumableArray(_this.state.tags), [tags]) });
    };

    _this.handleOnRemoveTag = function (tagId) {
      return _this.setState({ tags: _this.state.tags.filter(function (currTag) {
          return currTag.id !== tagId;
        }) });
    };

    _this.handleOnChange = function (event) {
      return _this.setState({ inputValue: event.target.value });
    };

    _this.predicate = function (option) {
      return _this.getValue(option).toLowerCase().includes(_this.state.inputValue.toLowerCase());
    };

    _this.state = {
      tags: [],
      options: _ExampleStandard.options,
      inputValue: ''
    };
    return _this;
  }

  _createClass(Form, [{
    key: 'getComponent',
    value: function getComponent() {
      return _react2.default.createElement(
        _MultiSelectComposite2.default,
        { required: this.props.required, info: this.props.info },
        this.props.withLabel ? _react2.default.createElement(_Label2.default, _extends({ 'for': 'firstName' }, this.props.label)) : null,
        _react2.default.createElement(_MultiSelect2.default, {
          tags: this.state.tags,
          onSelect: this.handleOnSelect,
          onRemoveTag: this.handleOnRemoveTag,
          onChange: this.handleOnChange,
          onManuallyInput: function onManuallyInput() {
            return console.log('NOW');
          },
          options: this.state.options,
          value: this.state.inputValue,
          predicate: this.predicate,
          valueParser: _ExampleStandard.valueParser
        })
      );
    }
  }, {
    key: 'render',
    value: function render() {
      return this.getComponent();
    }
  }]);

  return Form;
}(_react.Component);

Form.propTypes = {
  onChange: _propTypes2.default.func.isRequired,
  withLabel: _propTypes2.default.bool,
  label: _propTypes2.default.object,
  input: _propTypes2.default.object,
  required: _propTypes2.default.bool,
  info: _propTypes2.default.string
};
exports.default = Form;
