// wix-style-react/dist/src/InputWithOptions/InputWithOptions.protractor.driver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _DropdownLayoutProtractor = require('../DropdownLayout/DropdownLayout.protractor.driver');

var _DropdownLayoutProtractor2 = _interopRequireDefault(_DropdownLayoutProtractor);

var _protractor = require('wix-ui-test-utils/protractor');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var driverFactory = function driverFactory(component) {
  var dropdownLayoutDriver = (0, _DropdownLayoutProtractor2.default)(component.$('[data-hook="dropdown-layout-wrapper"]'));
  var input = component.$('input');

  return _extends({}, dropdownLayoutDriver, {
    click: function click() {
      return component.click();
    },
    getInput: function getInput() {
      return input;
    },
    isFocused: function isFocused() {
      return (0, _protractor.isFocused)(input);
    },
    element: function element() {
      return component;
    },
    /** Check wether the options dropdown is open */
    isOptionsShown: function isOptionsShown() {
      return dropdownLayoutDriver.getDropdown().isDisplayed();
    }
  });
};

exports.default = driverFactory;
