// playable/dist/src/core/config.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
require("jsdom-global/register");
var chai_1 = require("chai");
var config_1 = tslib_1.__importDefault(require("./config"));
var testkit_1 = require("../testkit");
describe('getUIConfig function', function () {
    afterEach(function () {
        testkit_1.resetProperty(navigator, 'userAgent');
    });
    it('should convert config if iPod', function () {
        var params = {};
        var expectedConfig = {
            loader: false,
            controls: false,
            title: false,
            showInteractionIndicator: false,
            screen: {
                disableClickProcessing: true,
                nativeControls: true,
            },
        };
        testkit_1.setProperty(navigator, 'userAgent', 'iPod');
        chai_1.expect(config_1.default(params)).to.be.deep.equal(expectedConfig);
    });
    it('should convert config if iPhone', function () {
        var params = {};
        var expectedConfig = {
            loader: false,
            controls: false,
            title: false,
            showInteractionIndicator: false,
            screen: {
                disableClickProcessing: true,
                nativeControls: true,
            },
        };
        testkit_1.setProperty(navigator, 'userAgent', 'iPhone');
        chai_1.expect(config_1.default(params)).to.be.deep.equal(expectedConfig);
    });
    it('should convert config if iPad', function () {
        var params = {};
        var expectedConfig = {
            loader: false,
            controls: false,
            title: false,
            showInteractionIndicator: false,
            screen: {
                disableClickProcessing: true,
                nativeControls: true,
            },
        };
        testkit_1.setProperty(navigator, 'userAgent', 'iPad');
        chai_1.expect(config_1.default(params)).to.be.deep.equal(expectedConfig);
    });
    it('should convert config if Android', function () {
        testkit_1.setProperty(navigator, 'userAgent', 'Android');
        var params = {
            title: { text: 'test' },
            controls: {},
            loader: true,
        };
        chai_1.expect(config_1.default(params)).to.be.deep.equal({
            title: { text: 'test' },
            controls: {},
            loader: true,
            screen: {
                disableClickProcessing: true,
            },
        });
    });
});
