// wix-style-react/dist/src/TPA/Button/Button.driver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _testUtils = require('react-dom/test-utils');

var _testUtils2 = _interopRequireDefault(_testUtils);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var buttonDriverFactory = function buttonDriverFactory(_ref) {
  var element = _ref.element;

  return {
    exists: function exists() {
      return !!element;
    },
    click: function click() {
      try {
        _testUtils2.default.Simulate.click(element);
      } catch (e) {
        element.simulate('click');
      }
    },
    hasTheme: function hasTheme(theme) {
      return element.getAttribute('data-theme') === theme;
    },
    hasClass: function hasClass(className) {
      return element.getAttribute('class').indexOf(className) > -1;
    },
    getButtonTextContent: function getButtonTextContent() {
      return element.textContent;
    },
    isButtonDisabled: function isButtonDisabled() {
      return element.getAttribute('disabled') === '';
    }
  };
};

exports.default = buttonDriverFactory;
