// wix-ui-core/dist/src/baseComponents/DropdownOption/DropdownOption.driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var test_utils_1 = require("stylable/test-utils");
var DropdownOption_st_css_1 = require("./DropdownOption.st.css");
exports.dropdownOptionDriverFactory = function (_a) {
    var element = _a.element, eventTrigger = _a.eventTrigger;
    var domUtils = new test_utils_1.StylableDOMUtil(DropdownOption_st_css_1.default, element);
    return {
        exists: function () { return !!element; },
        click: function () { return element && eventTrigger.click(element); },
        mouseEnter: function () { return element && eventTrigger.mouseEnter(element); },
        className: function () { return element && element.className; },
        isHovered: function () { return element && domUtils.hasStyleState(element, 'hovered'); },
        isSelected: function () { return element && domUtils.hasStyleState(element, 'selected'); },
        isDisabled: function () { return element && domUtils.hasStyleState(element, 'disabled'); },
        getText: function () { return element && element.textContent; },
        getElement: function () { return element; },
    };
};
