// wix-animations/dist/stories/Animations/CssStyles/ExampleTranslate.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Animator = require('../../../src/components/Animator');

var _Animator2 = _interopRequireDefault(_Animator);

var _ShowHideControls = require('../ShowHideControls');

var _ShowHideControls2 = _interopRequireDefault(_ShowHideControls);

var _StyledDiv = require('../StyledDiv');

var _StyledDiv2 = _interopRequireDefault(_StyledDiv);

var _ExampleTranslateHelpers = require('./ExampleTranslateHelpers');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var getValue = function getValue(value) {
  return '"' + value + '"';
};

var ExampleTranslate = function (_React$Component) {
  _inherits(ExampleTranslate, _React$Component);

  function ExampleTranslate(props) {
    _classCallCheck(this, ExampleTranslate);

    var _this = _possibleConstructorReturn(this, (ExampleTranslate.__proto__ || Object.getPrototypeOf(ExampleTranslate)).call(this, props));

    _this.state = {
      enterDirection: 'top',
      enterSize: '100%',
      exitDirection: 'left',
      exitSize: '150px'
    };
    return _this;
  }

  _createClass(ExampleTranslate, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var show = this.props.show;
      var _state = this.state,
          enterDirection = _state.enterDirection,
          exitDirection = _state.exitDirection,
          enterSize = _state.enterSize,
          exitSize = _state.exitSize;

      return _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
          'div',
          { style: { display: 'flex', marginBottom: '20px' } },
          _react2.default.createElement(_ExampleTranslateHelpers.DirectionDropDown, { title: 'Enter Direction', selectedId: enterDirection, onSelect: function onSelect(enterDirection) {
              return _this2.setState({ enterDirection: enterDirection });
            } }),
          _react2.default.createElement(_ExampleTranslateHelpers.SizeInput, { title: 'Enter Size', value: enterSize, onChange: function onChange(enterSize) {
              return _this2.setState({ enterSize: enterSize });
            } }),
          _react2.default.createElement(_ExampleTranslateHelpers.DirectionDropDown, { title: 'Exit Direction', selectedId: exitDirection, onSelect: function onSelect(exitDirection) {
              return _this2.setState({ exitDirection: exitDirection });
            } }),
          _react2.default.createElement(_ExampleTranslateHelpers.SizeInput, { title: 'Exit Size', value: exitSize, onChange: function onChange(exitSize) {
              return _this2.setState({ exitSize: exitSize });
            } })
        ),
        _react2.default.createElement(
          _Animator2.default,
          {
            show: show,
            opacity: true,
            translate: {
              enter: {
                direction: enterDirection,
                size: enterSize
              },
              exit: {
                direction: exitDirection,
                size: exitSize
              }
            }
          },
          _react2.default.createElement(
            _StyledDiv2.default,
            null,
            _react2.default.createElement(
              'pre',
              { style: { margin: 0 } },
              'translate=',
              '{',
              '{',
              'enter:',
              '{',
              'direction: ',
              getValue(enterDirection),
              ', size: ',
              getValue(enterSize),
              '}',
              ', exit: ',
              '{',
              'direction: ',
              getValue(exitDirection),
              ', size: ',
              getValue(exitSize),
              '}',
              '}',
              '}'
            )
          )
        )
      );
    }
  }]);

  return ExampleTranslate;
}(_react2.default.Component);

ExampleTranslate.propTypes = {
  show: _propTypes2.default.bool
};

exports.default = function () {
  return _react2.default.createElement(
    _ShowHideControls2.default,
    { height: '190px' },
    _react2.default.createElement(ExampleTranslate, null)
  );
};
