// wix-animations/dist/stories/Animations/CssStyles/ExampleTranslateHelpers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SizeInput = exports.DirectionDropDown = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Dropdown = require('wix-style-react/dist/src/Dropdown');

var _Dropdown2 = _interopRequireDefault(_Dropdown);

var _Input = require('wix-style-react/dist/src/Input');

var _Input2 = _interopRequireDefault(_Input);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var directions = [{ id: 'top', value: 'top' }, { id: 'bottom', value: 'bottom' }, { id: 'left', value: 'left' }, { id: 'right', value: 'right' }];

var DirectionDropDown = function DirectionDropDown(_ref) {
  var title = _ref.title,
      selectedId = _ref.selectedId,
      _onSelect = _ref.onSelect;
  return _react2.default.createElement(
    'div',
    { style: { marginRight: '20px' } },
    _react2.default.createElement(
      'div',
      { style: { marginBottom: '3px' } },
      title
    ),
    _react2.default.createElement(_Dropdown2.default, {
      width: '80px',
      selectedId: selectedId,
      onSelect: function onSelect(option) {
        return _onSelect(option.id);
      },
      options: directions
    })
  );
};

DirectionDropDown.propTypes = {
  title: _propTypes2.default.string,
  selectedId: _propTypes2.default.string,
  onSelect: _propTypes2.default.func
};

var SizeInput = function SizeInput(_ref2) {
  var title = _ref2.title,
      value = _ref2.value,
      _onChange = _ref2.onChange;
  return _react2.default.createElement(
    'div',
    { style: { marginRight: '20px' } },
    _react2.default.createElement(
      'div',
      { style: { marginBottom: '3px' } },
      title
    ),
    _react2.default.createElement(_Input2.default, {
      width: '80px',
      value: value,
      onChange: function onChange(e) {
        return _onChange(e.target.value);
      }
    })
  );
};

SizeInput.propTypes = {
  title: _propTypes2.default.string,
  value: _propTypes2.default.string,
  onChange: _propTypes2.default.func
};

exports.DirectionDropDown = DirectionDropDown;
exports.SizeInput = SizeInput;
