// wix-style-react/dist/stories/FilePicker/Example.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _FilePicker = require('wix-style-react/FilePicker');

var _FilePicker2 = _interopRequireDefault(_FilePicker);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function () {
  return _react2.default.createElement(
    'div',
    { style: { marginBottom: '100px' } },
    _react2.default.createElement(
      'div',
      { style: { marginBottom: '30px' } },
      _react2.default.createElement(_FilePicker2.default, {
        dataHook: 'story-filepicker',
        supportedFormats: '.png, .pdf',
        onChange: function onChange(file) {
          return console.log(file.name);
        }
      })
    ),
    _react2.default.createElement(_FilePicker2.default, {
      dataHook: 'story-filepicker2',
      supportedFormats: '.png, .pdf',
      onChange: function onChange(file) {
        return console.log(file.name);
      },
      mainLabel: 'FilePicker with error',
      error: true,
      errorMessage: 'File is too large'
    })
  );
};
