// wix-style-react/dist/src/ModalSelectorLayout/ModalSelectorLayout.protractor.driver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _protractor = require('wix-ui-test-utils/protractor');

var _LoaderProtractor = require('../Loader/Loader.protractor.driver');

var _LoaderProtractor2 = _interopRequireDefault(_LoaderProtractor);

var _ButtonProtractor = require('../Backoffice/Button/Button.protractor.driver');

var _ButtonProtractor2 = _interopRequireDefault(_ButtonProtractor);

var _TextProtractor = require('../Text/Text.protractor.driver');

var _TextProtractor2 = _interopRequireDefault(_TextProtractor);

var _SearchProtractor = require('../Search/Search.protractor.driver');

var _SearchProtractor2 = _interopRequireDefault(_SearchProtractor);

var _ModalSelectorLayout = require('./ModalSelectorLayout.helpers');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var loaderTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_LoaderProtractor2.default);
var buttonTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_ButtonProtractor2.default);
var textTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_TextProtractor2.default);
var searchTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_SearchProtractor2.default);

var modalSelectorLayoutDriverFactory = function modalSelectorLayoutDriverFactory(component) {
  var findByDataHook = function findByDataHook(dataHook) {
    return component.$('[data-hook="' + dataHook + '"]');
  };
  var findAllByDataHook = function findAllByDataHook(dataHook) {
    return component.$$('[data-hook="' + dataHook + '"]');
  };
  var mainLoaderDriver = function mainLoaderDriver() {
    return loaderTestkitFactory({ dataHook: _ModalSelectorLayout.dataHooks.mainLoader });
  };
  var nextPageLoaderDriver = function nextPageLoaderDriver() {
    return loaderTestkitFactory({ dataHook: _ModalSelectorLayout.dataHooks.nextPageLoader });
  };
  var cancelButtonDriver = function cancelButtonDriver() {
    return buttonTestkitFactory({ dataHook: 'cancellation-button' });
  };
  var okButtonDriver = function okButtonDriver() {
    return buttonTestkitFactory({ dataHook: 'confirmation-button' });
  };
  var subtitleTextDriver = function subtitleTextDriver() {
    return textTestkitFactory({ dataHook: _ModalSelectorLayout.dataHooks.subtitle });
  };
  var searchDriver = function searchDriver() {
    return searchTestkitFactory({ dataHook: _ModalSelectorLayout.dataHooks.search });
  };

  return {
    element: function element() {
      return component;
    },
    mainLoaderDriver: mainLoaderDriver,
    nextPageLoaderDriver: nextPageLoaderDriver,
    cancelButtonDriver: cancelButtonDriver,
    okButtonDriver: okButtonDriver,
    searchDriver: searchDriver,
    subtitleTextDriver: subtitleTextDriver,
    getTitle: function getTitle() {
      return findByDataHook('header-layout-title').getText();
    },
    clickOnClose: function clickOnClose() {
      return findByDataHook('header-close-button').click();
    },
    getEmptyState: function getEmptyState() {
      return findByDataHook(_ModalSelectorLayout.dataHooks.emptyState);
    },
    getNoResultsFoundState: function getNoResultsFoundState() {
      return findByDataHook(_ModalSelectorLayout.dataHooks.noResultsFoundState);
    },
    listExists: function listExists() {
      return findByDataHook(_ModalSelectorLayout.dataHooks.list).isPresent();
    },
    numberOfItemsInList: function numberOfItemsInList() {
      return findAllByDataHook(_ModalSelectorLayout.dataHooks.selector).count();
    }
  };
};

exports.default = modalSelectorLayoutDriverFactory;
