// wix-style-react/dist/src/Card/LinkHeader/LinkHeader.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _LinkHeader = require('./LinkHeader.scss');

var _LinkHeader2 = _interopRequireDefault(_LinkHeader);

var _TextLink = require('../../../src/Backoffice/TextLink');

var _TextLink2 = _interopRequireDefault(_TextLink);

var _WixComponent2 = require('../../../src/BaseComponents/WixComponent');

var _WixComponent3 = _interopRequireDefault(_WixComponent2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var LinkHeader = function (_WixComponent) {
  _inherits(LinkHeader, _WixComponent);

  function LinkHeader() {
    _classCallCheck(this, LinkHeader);

    return _possibleConstructorReturn(this, (LinkHeader.__proto__ || Object.getPrototypeOf(LinkHeader)).apply(this, arguments));
  }

  _createClass(LinkHeader, [{
    key: 'render',
    value: function render() {
      var _classNames;

      var _props = this.props,
          title = _props.title,
          subtitle = _props.subtitle,
          linkTitle = _props.linkTitle,
          linkTo = _props.linkTo,
          withoutDivider = _props.withoutDivider,
          tooltip = _props.tooltip;


      var headerClasses = (0, _classnames2.default)((_classNames = {}, _defineProperty(_classNames, _LinkHeader2.default.headerOnlyTitle, !subtitle), _defineProperty(_classNames, _LinkHeader2.default.headerTitleSubtitle, subtitle), _defineProperty(_classNames, _LinkHeader2.default.withDivider, !withoutDivider), _classNames));

      var linkElement = _react2.default.createElement(
        'div',
        { className: _LinkHeader2.default.link },
        _react2.default.createElement(
          _TextLink2.default,
          { dataHook: 'link', link: linkTo },
          linkTitle
        )
      );

      var titleElement = _react2.default.createElement(
        'div',
        { 'data-hook': 'title', className: _LinkHeader2.default.title },
        title
      );

      var tooltipElement = tooltip ? _react2.default.cloneElement(tooltip, {}, linkElement) : null;

      var actionElement = tooltipElement ? tooltipElement : linkElement;

      var subtitleElement = subtitle ? _react2.default.createElement(
        'div',
        { 'data-hook': 'subtitle', className: _LinkHeader2.default.subtitle },
        this.props.subtitle
      ) : null;

      return _react2.default.createElement(
        'div',
        { className: headerClasses },
        _react2.default.createElement(
          'div',
          null,
          titleElement,
          subtitleElement
        ),
        actionElement
      );
    }
  }]);

  return LinkHeader;
}(_WixComponent3.default);

LinkHeader.propTypes = {
  title: _propTypes.node.isRequired,
  linkTitle: _propTypes.string.isRequired,
  linkTo: _propTypes.string.isRequired,
  subtitle: _propTypes.node,
  tooltip: _propTypes.node,
  withoutDivider: _propTypes.bool
};
LinkHeader.defaultProps = {
  subtitle: null,
  tooltip: null,
  withoutDivider: false
};
exports.default = LinkHeader;
