// wix-style-react/dist/src/Card/LinkHeader/LinkHeader.driver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var linkHeaderDriverFactory = function linkHeaderDriverFactory(_ref) {
  var _element = _ref.element,
      wrapper = _ref.wrapper,
      component = _ref.component;


  var _title = _element.querySelector('[data-hook="title"]');
  var _subtitle = _element.querySelector('[data-hook="subtitle"]');

  return {
    exists: function exists() {
      return !!_element;
    },
    title: function title() {
      return _title && _title.innerHTML;
    },
    subtitle: function subtitle() {
      return _subtitle && _subtitle.innerHTML;
    },
    element: function element() {
      return _element;
    },
    linkDataHook: function linkDataHook() {
      return 'link';
    },
    setProps: function setProps(props) {
      var ClonedWithProps = _react2.default.cloneElement.apply(_react2.default, [component, Object.assign({}, component.props, props)].concat(_toConsumableArray(component.props.children || [])));
      _reactDom2.default.render(_react2.default.createElement(
        'div',
        { ref: function ref(r) {
            return _element = r;
          } },
        ClonedWithProps
      ), wrapper);
    }
  };
};

exports.default = linkHeaderDriverFactory;
