// wix-style-react/dist/src/AddItem/AddItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _AddItem = require('./AddItem.scss');

var _AddItem2 = _interopRequireDefault(_AddItem);

var _Tooltip = require('../Tooltip');

var _Tooltip2 = _interopRequireDefault(_Tooltip);

var _AddMedia = require('wix-ui-icons-common/system/AddMedia');

var _AddMedia2 = _interopRequireDefault(_AddMedia);

var _WixComponent2 = require('../BaseComponents/WixComponent');

var _WixComponent3 = _interopRequireDefault(_WixComponent2);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var tooltipCommonProps = {
  showDelay: 0,
  theme: 'dark',
  hideDelay: 0,
  align: 'center',
  placement: 'top'
};

var ratioClasses = {
  '16/9': _AddItem2.default.ratio16x9,
  '3/4': _AddItem2.default.ratio3x4,
  '4/3': _AddItem2.default.ratio4x3,
  '1/1': _AddItem2.default.ratio1x1
};

var renderInnerAddItem = function renderInnerAddItem() {
  return _react2.default.createElement(
    'div',
    { className: _AddItem2.default.dashedBorder },
    _react2.default.createElement(_AddMedia2.default, { className: _AddItem2.default.plusIcon, size: '31px' })
  );
};

var AddItem = function (_WixComponent) {
  _inherits(AddItem, _WixComponent);

  function AddItem() {
    _classCallCheck(this, AddItem);

    return _possibleConstructorReturn(this, (AddItem.__proto__ || Object.getPrototypeOf(AddItem)).apply(this, arguments));
  }

  _createClass(AddItem, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          onClick = _props.onClick,
          height = _props.height,
          tooltipContent = _props.tooltipContent,
          aspectRatio = _props.aspectRatio;

      var ratio = !height && ratioClasses[aspectRatio];
      return _react2.default.createElement(
        'div',
        { className: (0, _classnames2.default)(ratio, _AddItem2.default.box), style: { height: height } },
        _react2.default.createElement(
          'div',
          { className: _AddItem2.default.container, onClick: onClick, 'data-hook': 'add-container' },
          tooltipContent ? _react2.default.createElement(
            _Tooltip2.default,
            _extends({ content: tooltipContent, dataHook: 'add-tooltip' }, tooltipCommonProps),
            renderInnerAddItem()
          ) : renderInnerAddItem()
        )
      );
    }
  }]);

  return AddItem;
}(_WixComponent3.default);

AddItem.propTypes = {
  /** Funciton called upon click */
  onClick: _propTypes2.default.func,
  /** The elemnt's asspect ratio   */
  aspectRatio: _propTypes2.default.oneOf(Object.keys(ratioClasses)),
  /** Element's height - overrides the asspect ratio */
  height: _propTypes2.default.number,
  /** Content of the tooltip */
  tooltipContent: _propTypes2.default.string
};

AddItem.defaultProps = {
  aspectRatio: '1/1'
};

exports.default = AddItem;
