// wix-style-react/dist/stories/Card/ExampleCollapsableHeaders.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Grid = require('wix-style-react/Grid');

var _Card = require('wix-style-react/Card');

var _Card2 = _interopRequireDefault(_Card);

var _FormField = require('wix-style-react/FormField');

var _FormField2 = _interopRequireDefault(_FormField);

var _Input = require('wix-style-react/Input');

var _Input2 = _interopRequireDefault(_Input);

var _Button = require('wix-style-react/Button');

var _Button2 = _interopRequireDefault(_Button);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

exports.default = function () {
  return _react2.default.createElement(
    'div',
    { 'data-hook': 'card-example', style: { background: '#F0F4F7', padding: 30 } },
    _react2.default.createElement(
      _Grid.Container,
      null,
      _react2.default.createElement(
        _Grid.Row,
        null,
        _react2.default.createElement(
          _Grid.Col,
          { span: 6 },
          _react2.default.createElement(
            _Card2.default,
            null,
            _react2.default.createElement(
              _Card2.default.CollapsedHeader,
              { title: 'Card with collapsable content' },
              _react2.default.createElement(
                _Card2.default.Content,
                null,
                field()
              )
            )
          )
        ),
        _react2.default.createElement(
          _Grid.Col,
          { span: 6 },
          _react2.default.createElement(
            _Card2.default,
            null,
            _react2.default.createElement(
              _Card2.default.CollapsedHeader,
              {
                title: 'Card with collapsable content',
                subtitle: 'with button toggle',
                toggleStyle: 'button'
              },
              _react2.default.createElement(
                _Card2.default.Content,
                null,
                field()
              )
            )
          )
        )
      ),
      _react2.default.createElement(ControlledExample, null)
    )
  );
};

var ControlledExample = function (_React$Component) {
  _inherits(ControlledExample, _React$Component);

  function ControlledExample() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ControlledExample);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ControlledExample.__proto__ || Object.getPrototypeOf(ControlledExample)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      collapsed: false,
      toggled: false
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ControlledExample, [{
    key: 'onCollapsedChange',
    value: function onCollapsedChange() {
      var _this2 = this;

      this.setState({ toggled: true });
      setTimeout(function () {
        return _this2.setState({ toggled: false });
      }, 500);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      var collapsed = this.state.collapsed;

      return _react2.default.createElement(
        _Grid.Row,
        null,
        _react2.default.createElement(
          _Grid.Col,
          { span: 10 },
          _react2.default.createElement(
            _Card2.default,
            null,
            _react2.default.createElement(
              _Card2.default.CollapsedHeader,
              {
                collapsed: collapsed,
                controlled: true,
                title: 'Controlled <Card collapsed={' + collapsed.toString() + '}>',
                onCollapsedChange: this.onCollapsedChange
              },
              _react2.default.createElement(
                _Card2.default.Content,
                null,
                field()
              )
            )
          )
        ),
        _react2.default.createElement(
          _Grid.Col,
          { span: 2 },
          _react2.default.createElement(
            _Button2.default,
            {
              onClick: function onClick() {
                return _this3.setState(function (_ref2) {
                  var collapsed = _ref2.collapsed;
                  return { collapsed: !collapsed };
                });
              },
              height: 'large',
              theme: collapsed ? 'fullred' : 'fullblue'
            },
            collapsed ? 'Expand' : 'Collapse'
          )
        )
      );
    }
  }]);

  return ControlledExample;
}(_react2.default.Component);

function field() {
  return _react2.default.createElement(
    _FormField2.default,
    { label: 'Text Field' },
    _react2.default.createElement(_Input2.default, { placeholder: 'You can type here' })
  );
}
