// wix-style-react/dist/stories/Card/ExampleActionHeaders.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Card = require('wix-style-react/Card');

var _Card2 = _interopRequireDefault(_Card);

var _Grid = require('wix-style-react/Grid');

var _FormField = require('wix-style-react/FormField');

var _FormField2 = _interopRequireDefault(_FormField);

var _Input = require('wix-style-react/Input');

var _Input2 = _interopRequireDefault(_Input);

var _Tooltip = require('wix-style-react/Tooltip');

var _Tooltip2 = _interopRequireDefault(_Tooltip);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function () {
  return _react2.default.createElement(
    'div',
    { style: { background: '#F0F4F7', padding: 30 } },
    _react2.default.createElement(
      _Grid.Container,
      null,
      _react2.default.createElement(
        _Grid.Row,
        null,
        _react2.default.createElement(
          _Grid.Col,
          { span: 6 },
          _react2.default.createElement(
            _Card2.default,
            null,
            _react2.default.createElement(_Card2.default.ButtonHeader, {
              withNewIcons: true,
              title: 'Card header',
              buttonOnClick: function buttonOnClick() {
                return alert('Clicked!');
              },
              theme: 'fullblue',
              buttonTitle: 'Click Me!'
            }),
            _react2.default.createElement(
              _Card2.default.Content,
              null,
              field()
            )
          )
        ),
        _react2.default.createElement(
          _Grid.Col,
          { span: 6 },
          _react2.default.createElement(
            _Card2.default,
            null,
            _react2.default.createElement(_Card2.default.LinkHeader, {
              tooltip: _react2.default.createElement(_Tooltip2.default, { placement: 'top', alignment: 'center', content: 'And a tooltip!' }),
              title: 'Card header',
              linkTo: 'http://www.wix.com/',
              linkTitle: 'Link to Wix',
              subtitle: 'Subtitle'
            }),
            _react2.default.createElement(
              _Card2.default.Content,
              null,
              field()
            )
          )
        )
      ),
      _react2.default.createElement(
        _Grid.Row,
        null,
        _react2.default.createElement(
          _Grid.Col,
          null,
          _react2.default.createElement(
            _Card2.default,
            null,
            _react2.default.createElement(_Card2.default.ButtonHeader, {
              withoutDivider: true,
              title: 'Card header without no content',
              buttonOnClick: function buttonOnClick() {
                return alert('Clicked!');
              },
              buttonTitle: 'Click Me!',
              theme: 'fullblue'
            })
          )
        )
      )
    )
  );
};

function field() {
  return _react2.default.createElement(
    _FormField2.default,
    { label: 'Text Field' },
    _react2.default.createElement(_Input2.default, { placeholder: 'You can type here' })
  );
}
