// wix-style-react/dist/stories/TextField/ExampleStandard.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _TextFieldTemplate = require('./TextFieldTemplate');

var _TextFieldTemplate2 = _interopRequireDefault(_TextFieldTemplate);

var _Input = require('../../src/Input');

var _Input2 = _interopRequireDefault(_Input);

var _Label = require('../../src/Label');

var _Label2 = _interopRequireDefault(_Label);

var _ToggleSwitch = require('../../src/ToggleSwitch');

var _ToggleSwitch2 = _interopRequireDefault(_ToggleSwitch);

var _RadioGroup = require('../../src/RadioGroup');

var _RadioGroup2 = _interopRequireDefault(_RadioGroup);

var _ExampleStandard = require('./ExampleStandard.scss');

var _ExampleStandard2 = _interopRequireDefault(_ExampleStandard);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ExampleStandard = function (_Component) {
  _inherits(ExampleStandard, _Component);

  function ExampleStandard() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, ExampleStandard);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ExampleStandard.__proto__ || Object.getPrototypeOf(ExampleStandard)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      withLabel: true,
      prefixSuffixValue: '',
      prefixUnit: '',
      prefixTicker: false,
      suffixUnit: '',
      suffixTicker: false,
      label: {
        appearance: 'T1.1',
        children: 'First name'
      },
      input: {
        size: 'normal',
        placeholder: 'Please type in your first name...',
        disabled: false,
        clearButton: false
      },
      required: false,
      info: ''
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(ExampleStandard, [{
    key: 'setComponentState',
    value: function setComponentState(componentName, obj) {
      var _this2 = this;

      this.setState(function (prevState) {
        prevState[componentName] = _extends({}, _this2.state[componentName], obj);
        Object.keys(prevState[componentName]).forEach(function (k) {
          return !prevState[componentName][k] && delete prevState[componentName][k];
        });
        return prevState;
      });
    }
  }, {
    key: 'updateAddons',
    value: function updateAddons(name, state) {
      var _this3 = this;

      this.setState(state, function () {
        var unit = _this3.state[name + 'Unit'];
        var ticker = _this3.state[name + 'Ticker'];
        if (unit && ticker) {
          _this3.setComponentState('input', _defineProperty({}, name, _react2.default.createElement(
            _Input2.default.Group,
            null,
            _react2.default.createElement(_Input2.default.Unit, { value: unit }),
            _react2.default.createElement(_Input2.default.Ticker, { onUp: function onUp() {}, onDown: function onDown() {} })
          )));
        } else if (unit) {
          _this3.setComponentState('input', _defineProperty({}, name, _react2.default.createElement(_Input2.default.Unit, { value: unit })));
        } else if (ticker) {
          _this3.setComponentState('input', _defineProperty({}, name, _react2.default.createElement(_Input2.default.Ticker, { onUp: function onUp() {}, onDown: function onDown() {} })));
        } else {
          _this3.setComponentState('input', _defineProperty({}, name, null));
        }
      });
    }
  }, {
    key: 'render',
    value: function render() {
      var _this4 = this;

      return _react2.default.createElement(
        'form',
        { className: _ExampleStandard2.default.form },
        _react2.default.createElement(
          'div',
          { className: _ExampleStandard2.default.input },
          _react2.default.createElement(
            'div',
            { className: _ExampleStandard2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Show label'
            ),
            _react2.default.createElement(
              'div',
              { className: _ExampleStandard2.default.flex },
              _react2.default.createElement(_Input2.default, {
                size: 'small',
                value: this.state.label.children,
                onChange: function onChange(e) {
                  return _this4.setComponentState('label', { children: e.target.value });
                }
              }),
              '\xA0',
              _react2.default.createElement(_ToggleSwitch2.default, {
                size: 'small',
                checked: this.state.withLabel,
                onChange: function onChange() {
                  return _this4.setState({ withLabel: !_this4.state.withLabel });
                }
              })
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _ExampleStandard2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Placeholder'
            ),
            _react2.default.createElement(
              'div',
              { className: _ExampleStandard2.default.flex },
              _react2.default.createElement(_Input2.default, {
                size: 'small',
                value: this.state.input.placeholder,
                onChange: function onChange(e) {
                  return _this4.setComponentState('input', { placeholder: e.target.value });
                }
              })
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _ExampleStandard2.default.option },
            _react2.default.createElement(
              'div',
              { className: _ExampleStandard2.default.flex },
              _react2.default.createElement(
                'div',
                { className: _ExampleStandard2.default.paddRight },
                _react2.default.createElement(
                  _Label2.default,
                  null,
                  'Required Field: '
                )
              ),
              _react2.default.createElement(_ToggleSwitch2.default, {
                size: 'small',
                checked: this.state.required,
                onChange: function onChange() {
                  return _this4.setState({ required: !_this4.state.required });
                }
              })
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _ExampleStandard2.default.option },
            _react2.default.createElement(
              'div',
              { className: _ExampleStandard2.default.flex },
              _react2.default.createElement(
                'div',
                { className: _ExampleStandard2.default.paddRight },
                _react2.default.createElement(
                  _Label2.default,
                  null,
                  'Clear Button: '
                )
              ),
              _react2.default.createElement(_ToggleSwitch2.default, {
                size: 'small',
                checked: this.state.input.clearButton,
                onChange: function onChange() {
                  return _this4.setComponentState('input', { clearButton: !_this4.state.input.clearButton });
                }
              })
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _ExampleStandard2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Info Tooltip'
            ),
            _react2.default.createElement(
              'div',
              { className: _ExampleStandard2.default.flex },
              _react2.default.createElement(_Input2.default, {
                size: 'small',
                value: this.state.info,
                onChange: function onChange(e) {
                  return _this4.setState({ info: e.target.value });
                }
              })
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _ExampleStandard2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Size'
            ),
            _react2.default.createElement(
              'div',
              { className: _ExampleStandard2.default.flex },
              _react2.default.createElement(
                _RadioGroup2.default,
                {
                  display: 'horizontal',
                  value: this.state.input.size,
                  onChange: function onChange(size) {
                    return _this4.setComponentState('input', { size: size });
                  }
                },
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'small' },
                  'Small'
                ),
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'normal' },
                  'Normal'
                ),
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'large' },
                  'Large'
                )
              )
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _ExampleStandard2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Type'
            ),
            _react2.default.createElement(
              'div',
              { className: _ExampleStandard2.default.flex },
              _react2.default.createElement(
                _RadioGroup2.default,
                {
                  display: 'horizontal',
                  value: this.state.input.type,
                  onChange: function onChange(type) {
                    return _this4.setComponentState('input', { type: type });
                  }
                },
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: '' },
                  'Text'
                ),
                _react2.default.createElement(
                  _RadioGroup2.default.Radio,
                  { value: 'number' },
                  'Number'
                )
              )
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _ExampleStandard2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Prefix'
            ),
            _react2.default.createElement(
              'div',
              { className: _ExampleStandard2.default.flex },
              _react2.default.createElement(
                'div',
                { className: _ExampleStandard2.default.smallInput },
                _react2.default.createElement(_Input2.default, {
                  size: 'small',
                  value: this.state.prefixUnit,
                  onChange: function onChange(e) {
                    return _this4.updateAddons('prefix', { prefixUnit: e.target.value });
                  }
                })
              ),
              '\xA0Unit\xA0',
              _react2.default.createElement(_ToggleSwitch2.default, {
                size: 'small',
                checked: this.state.prefixTicker,
                onChange: function onChange() {
                  return _this4.updateAddons('prefix', { prefixTicker: !_this4.state.prefixTicker });
                }
              }),
              '\xA0with ticker'
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _ExampleStandard2.default.option },
            _react2.default.createElement(
              _Label2.default,
              null,
              'Suffix'
            ),
            _react2.default.createElement(
              'div',
              { className: _ExampleStandard2.default.flex },
              _react2.default.createElement(
                'div',
                { className: _ExampleStandard2.default.smallInput },
                _react2.default.createElement(_Input2.default, {
                  size: 'small',
                  value: this.state.suffixUnit,
                  onChange: function onChange(e) {
                    return _this4.updateAddons('suffix', { suffixUnit: e.target.value });
                  }
                })
              ),
              '\xA0Unit\xA0',
              _react2.default.createElement(_ToggleSwitch2.default, {
                size: 'small',
                checked: this.state.suffixTicker,
                onChange: function onChange() {
                  return _this4.updateAddons('suffix', { suffixTicker: !_this4.state.suffixTicker });
                }
              }),
              '\xA0with ticker'
            )
          ),
          _react2.default.createElement(
            'div',
            { className: _ExampleStandard2.default.option },
            _react2.default.createElement(
              'div',
              { className: _ExampleStandard2.default.flex },
              _react2.default.createElement(
                'div',
                { className: _ExampleStandard2.default.paddRight },
                _react2.default.createElement(
                  _Label2.default,
                  null,
                  'Enabled state: '
                )
              ),
              _react2.default.createElement(_ToggleSwitch2.default, {
                size: 'small',
                checked: !this.state.input.disabled,
                onChange: function onChange() {
                  return _this4.setComponentState('input', { disabled: !_this4.state.input.disabled });
                }
              }),
              '\xA0',
              this.state.input.disabled ? 'Disabled' : 'Enabled'
            )
          )
        ),
        _react2.default.createElement(
          'div',
          { className: _ExampleStandard2.default.output },
          _react2.default.createElement(_TextFieldTemplate2.default, _extends({}, this.state, { onChange: this.props.onChange }))
        )
      );
    }
  }]);

  return ExampleStandard;
}(_react.Component);

ExampleStandard.propTypes = {
  onChange: _propTypes2.default.func
};
exports.default = ExampleStandard;
