// wix-style-react/dist/stories/TextField/TextFieldTemplate.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactElementToJsxString = require('react-element-to-jsx-string');

var _reactElementToJsxString2 = _interopRequireDefault(_reactElementToJsxString);

var _TextField = require('../../src/TextField');

var _TextField2 = _interopRequireDefault(_TextField);

var _Input = require('../../src/Input');

var _Input2 = _interopRequireDefault(_Input);

var _Label = require('../../src/Label');

var _Label2 = _interopRequireDefault(_Label);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Form = function (_Component) {
  _inherits(Form, _Component);

  function Form() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, Form);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = Form.__proto__ || Object.getPrototypeOf(Form)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      value: ''
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(Form, [{
    key: 'componentDidUpdate',
    value: function componentDidUpdate(props) {
      props.onChange((0, _reactElementToJsxString2.default)(this.getComponent()));
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.props.onChange((0, _reactElementToJsxString2.default)(this.getComponent()));
    }
  }, {
    key: 'getComponent',
    value: function getComponent() {
      var _this2 = this;

      return _react2.default.createElement(
        _TextField2.default,
        { required: this.props.required, info: this.props.info },
        this.props.withLabel ? _react2.default.createElement(_Label2.default, _extends({ 'for': 'firstName' }, this.props.label)) : null,
        _react2.default.createElement(_Input2.default, _extends({
          id: 'firstName'
        }, this.props.input, {
          value: this.state.value,
          onChange: function onChange(e) {
            return _this2.setState({ value: e.target.value });
          }
        }))
      );
    }
  }, {
    key: 'render',
    value: function render() {
      return this.getComponent();
    }
  }]);

  return Form;
}(_react.Component);

Form.propTypes = {
  onChange: _propTypes2.default.func.isRequired,
  withLabel: _propTypes2.default.bool,
  label: _propTypes2.default.object,
  input: _propTypes2.default.object,
  required: _propTypes2.default.bool,
  info: _propTypes2.default.string
};
exports.default = Form;
