// wix-style-react/dist/src/TPA/TextLink/TextLink.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _WixComponent2 = require('../../BaseComponents/WixComponent');

var _WixComponent3 = _interopRequireDefault(_WixComponent2);

var _TpaStyleInjector = require('../TpaStyleInjector');

var _TpaStyleInjector2 = _interopRequireDefault(_TpaStyleInjector);

var _omit = require('omit');

var _omit2 = _interopRequireDefault(_omit);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var styles = { locals: {} };
try {
  styles = require('!css-loader?modules&camelCase&localIdentName="[path][name]__[local]__[hash:base64:5]"!sass-loader!./TextLink.scss');
} catch (e) {}

var TextLink = function (_WixComponent) {
  _inherits(TextLink, _WixComponent);

  function TextLink() {
    _classCallCheck(this, TextLink);

    return _possibleConstructorReturn(this, (TextLink.__proto__ || Object.getPrototypeOf(TextLink)).apply(this, arguments));
  }

  _createClass(TextLink, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          className = _props.className,
          link = _props.link;
      var _styles = styles,
          locals = _styles.locals;

      var classes = (0, _classnames2.default)([locals['wix-style-react-text-link']], className).trim();
      var propsToOmit = ['children', 'className', 'link', 'href', 'dataHook', 'injectedStyles'];

      return _react2.default.createElement(
        'a',
        _extends({ className: classes, href: link }, (0, _omit2.default)(propsToOmit, this.props)),
        children
      );
    }
  }]);

  return TextLink;
}(_WixComponent3.default);

TextLink.propTypes = {
  link: _propTypes.string.isRequired,
  id: _propTypes.string,
  children: _propTypes.string
};
TextLink.defaultProps = {
  disabled: false,
  rel: null,
  target: null
};


TextLink.displayName = 'TextLink';

exports.default = (0, _TpaStyleInjector2.default)(TextLink, styles);
