// wix-style-react/dist/testkit/enzyme.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.circularProgressBarTestkitFactory = exports.linearProgressBarTestkitFactory = exports.floatingHelperTestkitFactory = exports.labelTestkitFactory = exports.toggleSwitchTestkitFactory = exports.counterBadgeTestkitFactory = exports.badgeTestkitFactory = exports.textTestkitFactory = exports.headingTestkitFactory = exports.formFieldTestkitFactory = exports.sectionHelperTestkitFactory = exports.modalSelectorLayoutTestkitFactory = exports.statsWidgetTestkitFactory = exports.highlighterTestkitFactory = exports.searchTestkitFactory = exports.iconWithOptionsTestkitFactory = exports.editableRowTestkitFactory = exports.editableSelectorTestkitFactory = exports.selectorTestkitFactory = exports.datePickerTestkitFactory = exports.popoverMenuTestkitFactory = exports.fieldLabelAttributesTestkitFactory = exports.tpaFloatingTabsTestkitFactory = exports.tpaTextLinkTestkitFactory = exports.tpaInputTestkitFactory = exports.tpaButtonTestkitFactory = exports.sideMenuDrillTestkitFactory = exports.sideMenuTestkitFactory = exports.imageViewerTestkitFactory = exports.addItemTestkitFactory = exports.sliderTestkitFactory = exports.timeInputTestkitFactory = exports.filePickerTestkitFactory = exports.modalTestkitFactory = exports.richTextAreaTestkitFactory = exports.richTextAreaCompositeTestkitFactory = exports.headerTestkitFactory = exports.collapsedHeaderTestkitFactory = exports.linkHeaderTestkitFactory = exports.buttonHeaderTestkitFactory = exports.messageBoxMarketerialLayoutTestkitFactory = exports.messageBoxFunctionalLayoutTestkitFactory = exports.googleAddressWithInputTestkitFactory = exports.fieldWithSelectionCompositeTestkitFactory = exports.rangeTestkitFactory = exports.notificationTestkitFactory = exports.radioButtonTestkitFactory = exports.radioGroupTestkitFactory = exports.loaderTestkitFactory = exports.multiSelectCompositeTestkitFactory = exports.autoCompleteCompositeTestkitFactory = exports.dropdownCompositeTestkitFactory = exports.textFieldTestkitFactory = exports.textAreaTestkitFactory = exports.tableTestkitFactory = exports.dataTableTestkitFactory = exports.textLinkLayoutTestkitFactory = exports.textLinkTestkitFactory = exports.buttonTestkitFactory = exports.checkboxTestkitFactory = exports.tagTestkitFactory = exports.multiSelectCheckboxTestkitFactory = exports.multiSelectTestkitFactory = exports.dropdownTestkitFactory = exports.autoCompleteTestkitFactory = exports.languagePickerTestkitFactory = exports.buttonWithOptionsTestkitFactory = exports.inputWithOptionsTestkitFactory = exports.dropdownLayoutTestkitFactory = exports.pageHeaderTestkitFactory = exports.pageTestkitFactory = exports.tabsTestkitFactory = exports.breadcrumbsTestkitFactory = exports.backofficeTooltipTestkitFactory = exports.tooltipTestkitFactory = exports.inputAreaTestkitFactory = exports.inputTestkitFactory = undefined;

var _enzyme = require('wix-ui-backoffice/dist/src/testkit/enzyme');

Object.defineProperty(exports, 'badgeTestkitFactory', {
  enumerable: true,
  get: function get() {
    return _enzyme.badgeTestkitFactory;
  }
});
Object.defineProperty(exports, 'counterBadgeTestkitFactory', {
  enumerable: true,
  get: function get() {
    return _enzyme.stylableCounterBadgeTestkitFactory;
  }
});
Object.defineProperty(exports, 'toggleSwitchTestkitFactory', {
  enumerable: true,
  get: function get() {
    return _enzyme.toggleSwitchTestkitFactory;
  }
});
Object.defineProperty(exports, 'labelTestkitFactory', {
  enumerable: true,
  get: function get() {
    return _enzyme.labelTestkitFactory;
  }
});
Object.defineProperty(exports, 'floatingHelperTestkitFactory', {
  enumerable: true,
  get: function get() {
    return _enzyme.floatingHelperTestkitFactory;
  }
});
Object.defineProperty(exports, 'linearProgressBarTestkitFactory', {
  enumerable: true,
  get: function get() {
    return _enzyme.linearProgressBarTestkitFactory;
  }
});
Object.defineProperty(exports, 'circularProgressBarTestkitFactory', {
  enumerable: true,
  get: function get() {
    return _enzyme.circularProgressBarTestkitFactory;
  }
});

var _enzyme2 = require('wix-ui-test-utils/enzyme');

var _Input = require('../src/Input/Input.driver');

var _Input2 = _interopRequireDefault(_Input);

var _InputArea = require('../src/InputArea/InputArea.driver');

var _InputArea2 = _interopRequireDefault(_InputArea);

var _Tooltip = require('../src/Tooltip/Tooltip.driver');

var _Tooltip2 = _interopRequireDefault(_Tooltip);

var _Tooltip3 = require('../src/Backoffice/Tooltip/Tooltip.driver');

var _Tooltip4 = _interopRequireDefault(_Tooltip3);

var _Breadcrumbs = require('../src/Breadcrumbs/Breadcrumbs.driver');

var _Breadcrumbs2 = _interopRequireDefault(_Breadcrumbs);

var _Tabs = require('../src/Tabs/Tabs.driver');

var _Tabs2 = _interopRequireDefault(_Tabs);

var _Page = require('../src/Page/Page.driver');

var _Page2 = _interopRequireDefault(_Page);

var _PageHeader = require('../src/PageHeader/PageHeader.driver');

var _PageHeader2 = _interopRequireDefault(_PageHeader);

var _DropdownLayout = require('../src/DropdownLayout/DropdownLayout.driver');

var _DropdownLayout2 = _interopRequireDefault(_DropdownLayout);

var _InputWithOptions = require('../src/InputWithOptions/InputWithOptions.driver');

var _InputWithOptions2 = _interopRequireDefault(_InputWithOptions);

var _ButtonWithOptions = require('../src/ButtonWithOptions/ButtonWithOptions.driver');

var _ButtonWithOptions2 = _interopRequireDefault(_ButtonWithOptions);

var _LanguagePicker = require('../src/LanguagePicker/LanguagePicker.driver');

var _LanguagePicker2 = _interopRequireDefault(_LanguagePicker);

var _AutoComplete = require('../src/AutoComplete/AutoComplete.driver');

var _AutoComplete2 = _interopRequireDefault(_AutoComplete);

var _Dropdown = require('../src/Dropdown/Dropdown.driver');

var _Dropdown2 = _interopRequireDefault(_Dropdown);

var _MultiSelect = require('../src/MultiSelect/MultiSelect.driver');

var _MultiSelect2 = _interopRequireDefault(_MultiSelect);

var _MultiSelectCheckbox = require('../src/MultiSelectCheckbox/MultiSelectCheckbox.driver');

var _MultiSelectCheckbox2 = _interopRequireDefault(_MultiSelectCheckbox);

var _Tag = require('../src/Tag/Tag.driver');

var _Tag2 = _interopRequireDefault(_Tag);

var _Checkbox = require('../src/Checkbox/Checkbox.driver');

var _Checkbox2 = _interopRequireDefault(_Checkbox);

var _Button = require('../src/Backoffice/Button/Button.driver');

var _Button2 = _interopRequireDefault(_Button);

var _TextLink = require('../src/BaseComponents/TextLink/TextLink.driver');

var _TextLink2 = _interopRequireDefault(_TextLink);

var _TextLinkLayout = require('../src/BaseComponents/TextLinkLayout/TextLinkLayout.driver');

var _TextLinkLayout2 = _interopRequireDefault(_TextLinkLayout);

var _DataTable = require('../src/DataTable/DataTable.driver');

var _DataTable2 = _interopRequireDefault(_DataTable);

var _Table = require('../src/Table/Table.driver');

var _Table2 = _interopRequireDefault(_Table);

var _TextArea = require('../src/TextArea/TextArea.driver');

var _TextArea2 = _interopRequireDefault(_TextArea);

var _TextField = require('../src/TextField/TextField.driver');

var _TextField2 = _interopRequireDefault(_TextField);

var _DropdownComposite = require('../src/DropdownComposite/DropdownComposite.driver');

var _DropdownComposite2 = _interopRequireDefault(_DropdownComposite);

var _AutoCompleteComposite = require('../src/AutoCompleteComposite/AutoCompleteComposite.driver');

var _AutoCompleteComposite2 = _interopRequireDefault(_AutoCompleteComposite);

var _MultiSelectComposite = require('../src/MultiSelectComposite/MultiSelectComposite.driver');

var _MultiSelectComposite2 = _interopRequireDefault(_MultiSelectComposite);

var _Loader = require('../src/Loader/Loader.driver');

var _Loader2 = _interopRequireDefault(_Loader);

var _RadioGroup = require('../src/RadioGroup/RadioGroup.driver');

var _RadioGroup2 = _interopRequireDefault(_RadioGroup);

var _RadioButton = require('../src/RadioGroup/RadioButton/RadioButton.driver');

var _RadioButton2 = _interopRequireDefault(_RadioButton);

var _Notification = require('../src/Notification/Notification.driver');

var _Notification2 = _interopRequireDefault(_Notification);

var _Range = require('../src/Range/Range.driver');

var _Range2 = _interopRequireDefault(_Range);

var _FieldWithSelectionComposite = require('../src/Composite/FieldWithSelectionComposite/FieldWithSelectionComposite.driver');

var _FieldWithSelectionComposite2 = _interopRequireDefault(_FieldWithSelectionComposite);

var _GoogleAddressInputWithLabel = require('../src/GoogleAddressInputWithLabel/GoogleAddressInputWithLabel.driver');

var _GoogleAddressInputWithLabel2 = _interopRequireDefault(_GoogleAddressInputWithLabel);

var _MessageBoxFunctionalLayout = require('../src/MessageBox/MessageBoxFunctionalLayout.driver');

var _MessageBoxFunctionalLayout2 = _interopRequireDefault(_MessageBoxFunctionalLayout);

var _MessageBoxMarketerialLayout = require('../src/MessageBox/MessageBoxMarketerialLayout.driver');

var _MessageBoxMarketerialLayout2 = _interopRequireDefault(_MessageBoxMarketerialLayout);

var _ButtonHeader = require('../src/Card/ButtonHeader/ButtonHeader.driver');

var _ButtonHeader2 = _interopRequireDefault(_ButtonHeader);

var _LinkHeader = require('../src/Card/LinkHeader/LinkHeader.driver');

var _LinkHeader2 = _interopRequireDefault(_LinkHeader);

var _CollapsedHeader = require('../src/Card/CollapsedHeader/CollapsedHeader.driver');

var _CollapsedHeader2 = _interopRequireDefault(_CollapsedHeader);

var _Header = require('../src/Card/Header/Header.driver');

var _Header2 = _interopRequireDefault(_Header);

var _RichTextAreaComposite = require('../src/RichTextAreaComposite/RichTextAreaComposite.driver');

var _RichTextAreaComposite2 = _interopRequireDefault(_RichTextAreaComposite);

var _RichTextArea = require('../src/RichTextArea/RichTextArea.driver');

var _RichTextArea2 = _interopRequireDefault(_RichTextArea);

var _Modal = require('../src/Modal/Modal.driver');

var _Modal2 = _interopRequireDefault(_Modal);

var _FilePicker = require('../src/FilePicker/FilePicker.driver');

var _FilePicker2 = _interopRequireDefault(_FilePicker);

var _TimeInput = require('../src/TimeInput/TimeInput.driver');

var _TimeInput2 = _interopRequireDefault(_TimeInput);

var _Slider = require('../src/Slider/Slider.driver');

var _Slider2 = _interopRequireDefault(_Slider);

var _AddItem = require('../src/AddItem/AddItem.driver');

var _AddItem2 = _interopRequireDefault(_AddItem);

var _ImageViewer = require('../src/ImageViewer/ImageViewer.driver');

var _ImageViewer2 = _interopRequireDefault(_ImageViewer);

var _SideMenu = require('../src/SideMenu/core/SideMenu.driver');

var _SideMenu2 = _interopRequireDefault(_SideMenu);

var _DrillView = require('../src/SideMenu/DrillView/DrillView.driver');

var _DrillView2 = _interopRequireDefault(_DrillView);

var _Button3 = require('../src/TPA/Button/Button.driver');

var _Button4 = _interopRequireDefault(_Button3);

var _Input3 = require('../src/TPA/Input/Input.driver');

var _Input4 = _interopRequireDefault(_Input3);

var _TextLink3 = require('../src/TPA/TextLink/TextLink.driver');

var _TextLink4 = _interopRequireDefault(_TextLink3);

var _FloatingTabs = require('../src/TPA/FloatingTabs/FloatingTabs.driver');

var _FloatingTabs2 = _interopRequireDefault(_FloatingTabs);

var _FieldLabelAttributes = require('../src/FieldLabelAttributes/FieldLabelAttributes.driver');

var _FieldLabelAttributes2 = _interopRequireDefault(_FieldLabelAttributes);

var _PopoverMenu = require('../src/PopoverMenu/PopoverMenu.driver');

var _PopoverMenu2 = _interopRequireDefault(_PopoverMenu);

var _DatePicker = require('../src/DatePicker/DatePicker.driver');

var _DatePicker2 = _interopRequireDefault(_DatePicker);

var _Selector = require('../src/Selector/Selector.driver');

var _Selector2 = _interopRequireDefault(_Selector);

var _EditableSelector = require('../src/EditableSelector/EditableSelector.driver');

var _EditableSelector2 = _interopRequireDefault(_EditableSelector);

var _EditableRow = require('../src/EditableSelector/EditableRow/EditableRow.driver');

var _EditableRow2 = _interopRequireDefault(_EditableRow);

var _IconWithOptions = require('../src/IconWithOptions/IconWithOptions.driver');

var _IconWithOptions2 = _interopRequireDefault(_IconWithOptions);

var _Search = require('../src/Search/Search.driver');

var _Search2 = _interopRequireDefault(_Search);

var _Highlighter = require('../src/Highlighter/Highlighter.driver');

var _Highlighter2 = _interopRequireDefault(_Highlighter);

var _StatsWidget = require('../src/StatsWidget/StatsWidget.driver');

var _StatsWidget2 = _interopRequireDefault(_StatsWidget);

var _ModalSelectorLayout = require('../src/ModalSelectorLayout/ModalSelectorLayout.driver');

var _ModalSelectorLayout2 = _interopRequireDefault(_ModalSelectorLayout);

var _SectionHelper = require('../src/SectionHelper/SectionHelper.driver');

var _SectionHelper2 = _interopRequireDefault(_SectionHelper);

var _FormField = require('../src/FormField/FormField.driver');

var _FormField2 = _interopRequireDefault(_FormField);

var _Heading = require('../src/Heading/Heading.driver');

var _Heading2 = _interopRequireDefault(_Heading);

var _Text = require('../src/Text/Text.driver');

var _Text2 = _interopRequireDefault(_Text);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var inputTestkitFactory = exports.inputTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Input2.default);

var inputAreaTestkitFactory = exports.inputAreaTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_InputArea2.default);

var tooltipTestkitFactory = exports.tooltipTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Tooltip2.default);

var backofficeTooltipTestkitFactory = exports.backofficeTooltipTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Tooltip4.default);

var breadcrumbsTestkitFactory = exports.breadcrumbsTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Breadcrumbs2.default);

var tabsTestkitFactory = exports.tabsTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Tabs2.default);

var pageTestkitFactory = exports.pageTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Page2.default);

var pageHeaderTestkitFactory = exports.pageHeaderTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_PageHeader2.default);

var dropdownLayoutTestkitFactory = exports.dropdownLayoutTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_DropdownLayout2.default);

var inputWithOptionsTestkitFactory = exports.inputWithOptionsTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_InputWithOptions2.default);

var buttonWithOptionsTestkitFactory = exports.buttonWithOptionsTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_ButtonWithOptions2.default);

var languagePickerTestkitFactory = exports.languagePickerTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_LanguagePicker2.default);

var autoCompleteTestkitFactory = exports.autoCompleteTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_AutoComplete2.default);

var dropdownTestkitFactory = exports.dropdownTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Dropdown2.default);

var multiSelectTestkitFactory = exports.multiSelectTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_MultiSelect2.default);

var multiSelectCheckboxTestkitFactory = exports.multiSelectCheckboxTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_MultiSelectCheckbox2.default);

var tagTestkitFactory = exports.tagTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Tag2.default);

var checkboxTestkitFactory = exports.checkboxTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Checkbox2.default);

var buttonTestkitFactory = exports.buttonTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Button2.default);

var textLinkTestkitFactory = exports.textLinkTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_TextLink2.default);

var textLinkLayoutTestkitFactory = exports.textLinkLayoutTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_TextLinkLayout2.default);

var dataTableTestkitFactory = exports.dataTableTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_DataTable2.default);

var tableTestkitFactory = exports.tableTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Table2.default);

var textAreaTestkitFactory = exports.textAreaTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_TextArea2.default);

var textFieldTestkitFactory = exports.textFieldTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_TextField2.default);

var dropdownCompositeTestkitFactory = exports.dropdownCompositeTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_DropdownComposite2.default);

var autoCompleteCompositeTestkitFactory = exports.autoCompleteCompositeTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_AutoCompleteComposite2.default);

var multiSelectCompositeTestkitFactory = exports.multiSelectCompositeTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_MultiSelectComposite2.default);

var loaderTestkitFactory = exports.loaderTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Loader2.default);

var radioGroupTestkitFactory = exports.radioGroupTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_RadioGroup2.default);

var radioButtonTestkitFactory = exports.radioButtonTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_RadioButton2.default);

var notificationTestkitFactory = exports.notificationTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Notification2.default);

var rangeTestkitFactory = exports.rangeTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Range2.default);

var fieldWithSelectionCompositeTestkitFactory = exports.fieldWithSelectionCompositeTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_FieldWithSelectionComposite2.default);

var googleAddressWithInputTestkitFactory = exports.googleAddressWithInputTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_GoogleAddressInputWithLabel2.default);

var messageBoxFunctionalLayoutTestkitFactory = exports.messageBoxFunctionalLayoutTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_MessageBoxFunctionalLayout2.default);

var messageBoxMarketerialLayoutTestkitFactory = exports.messageBoxMarketerialLayoutTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_MessageBoxMarketerialLayout2.default);

var buttonHeaderTestkitFactory = exports.buttonHeaderTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_ButtonHeader2.default);

var linkHeaderTestkitFactory = exports.linkHeaderTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_LinkHeader2.default);

var collapsedHeaderTestkitFactory = exports.collapsedHeaderTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_CollapsedHeader2.default);

var headerTestkitFactory = exports.headerTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Header2.default);

var richTextAreaCompositeTestkitFactory = exports.richTextAreaCompositeTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_RichTextAreaComposite2.default);

var richTextAreaTestkitFactory = exports.richTextAreaTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_RichTextArea2.default);

var modalTestkitFactory = exports.modalTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Modal2.default);

var filePickerTestkitFactory = exports.filePickerTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_FilePicker2.default);

var timeInputTestkitFactory = exports.timeInputTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_TimeInput2.default);

var sliderTestkitFactory = exports.sliderTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Slider2.default);

var addItemTestkitFactory = exports.addItemTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_AddItem2.default);

var imageViewerTestkitFactory = exports.imageViewerTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_ImageViewer2.default);

var sideMenuTestkitFactory = exports.sideMenuTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_SideMenu2.default);

var sideMenuDrillTestkitFactory = exports.sideMenuDrillTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_DrillView2.default);

var tpaButtonTestkitFactory = exports.tpaButtonTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Button4.default);

var tpaInputTestkitFactory = exports.tpaInputTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Input4.default);

var tpaTextLinkTestkitFactory = exports.tpaTextLinkTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_TextLink4.default);

var tpaFloatingTabsTestkitFactory = exports.tpaFloatingTabsTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_FloatingTabs2.default);

var fieldLabelAttributesTestkitFactory = exports.fieldLabelAttributesTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_FieldLabelAttributes2.default);

var popoverMenuTestkitFactory = exports.popoverMenuTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_PopoverMenu2.default);

var datePickerTestkitFactory = exports.datePickerTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_DatePicker2.default);

var selectorTestkitFactory = exports.selectorTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Selector2.default);

var editableSelectorTestkitFactory = exports.editableSelectorTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_EditableSelector2.default);

var editableRowTestkitFactory = exports.editableRowTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_EditableRow2.default);

var iconWithOptionsTestkitFactory = exports.iconWithOptionsTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_IconWithOptions2.default);

var searchTestkitFactory = exports.searchTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Search2.default);

var highlighterTestkitFactory = exports.highlighterTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Highlighter2.default);

var statsWidgetTestkitFactory = exports.statsWidgetTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_StatsWidget2.default);

var modalSelectorLayoutTestkitFactory = exports.modalSelectorLayoutTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_ModalSelectorLayout2.default);

var sectionHelperTestkitFactory = exports.sectionHelperTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_SectionHelper2.default);

var formFieldTestkitFactory = exports.formFieldTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_FormField2.default);

var headingTestkitFactory = exports.headingTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Heading2.default);

var textTestkitFactory = exports.textTestkitFactory = (0, _enzyme2.enzymeTestkitFactoryCreator)(_Text2.default);

// wix-ui-backoffice proxy
