// wix-style-react/dist/testkit/protractor.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.circularProgressBarTestkitFactory = exports.linearProgressBarTestkitFactory = exports.floatingHelperTestkitFactory = exports.labelTestkitFactory = exports.toggleSwitchTestkitFactory = exports.counterBadgeTestkitFactory = exports.badgeTestkitFactory = exports.headingTestkitFactory = exports.messageBoxFunctionalLayoutTestkitFactory = exports.textTestkitFactory = exports.formFieldTestkitFactory = exports.tooltipTestkitFactory = exports.drillViewTestkitFactory = exports.tabsTestkitFactory = exports.sliderTestkitFactory = exports.modalSelectorLayoutTestkitFactory = exports.loaderTestkitFactory = exports.statsWidgetTestkitFactory = exports.highlighterTestkitFactory = exports.searchTestkitFactory = exports.iconWithOptionsTestkitFactory = exports.editableSelectorTestkitFactory = exports.datePickerTestkitFactory = exports.multiSelectCheckboxTestkitFactory = exports.multiSelectTestkitFactory = exports.popoverMenuTestkitFactory = exports.tpaTextLinkTestkitFactory = exports.imageViewerTestkitFactory = exports.googleAddressWithInputTestkitFactory = exports.fieldWithSelectionCompositeTestkitFactory = exports.rangeTestkitFactory = exports.autoCompleteCompositeTestkitFactory = exports.multiSelectCompositeTestkitFactory = exports.dropdownCompositeTestkitFactory = exports.textFieldTestkitFactory = exports.textAreaTestkitFactory = exports.radioGroupTestkitFactory = exports.filePickerTestkitFactory = exports.languagePickerTestkitFactory = exports.buttonWithOptionsTestkitFactory = exports.dropdownTestkitFactory = exports.checkboxTestkitFactory = exports.autoCompleteTestkitFactory = exports.breadcrumbsTestkitFactory = exports.tableTestkitFactory = exports.dataTableTestkitFactory = exports.pageTestkitFactory = exports.tpaInputTestkitFactory = exports.tpaButtonTestkitFactory = exports.buttonTestkitFactory = exports.inputAreaTestkitFactory = exports.richTextAreaTestkitFactory = exports.addItemTestkitFactory = exports.inputTestkitFactory = exports.scrollToElement = exports.waitForVisibilityOf = undefined;

var _protractor = require('wix-ui-test-utils/protractor');

Object.defineProperty(exports, 'waitForVisibilityOf', {
  enumerable: true,
  get: function get() {
    return _protractor.waitForVisibilityOf;
  }
});
Object.defineProperty(exports, 'scrollToElement', {
  enumerable: true,
  get: function get() {
    return _protractor.scrollToElement;
  }
});

var _protractor2 = require('wix-ui-backoffice/dist/src/testkit/protractor');

Object.defineProperty(exports, 'badgeTestkitFactory', {
  enumerable: true,
  get: function get() {
    return _protractor2.badgeTestkitFactory;
  }
});
Object.defineProperty(exports, 'counterBadgeTestkitFactory', {
  enumerable: true,
  get: function get() {
    return _protractor2.stylableCounterBadgeTestkitFactory;
  }
});
Object.defineProperty(exports, 'toggleSwitchTestkitFactory', {
  enumerable: true,
  get: function get() {
    return _protractor2.toggleSwitchTestkitFactory;
  }
});
Object.defineProperty(exports, 'labelTestkitFactory', {
  enumerable: true,
  get: function get() {
    return _protractor2.labelTestkitFactory;
  }
});
Object.defineProperty(exports, 'floatingHelperTestkitFactory', {
  enumerable: true,
  get: function get() {
    return _protractor2.floatingHelperTestkitFactory;
  }
});
Object.defineProperty(exports, 'linearProgressBarTestkitFactory', {
  enumerable: true,
  get: function get() {
    return _protractor2.linearProgressBarTestkitFactory;
  }
});
Object.defineProperty(exports, 'circularProgressBarTestkitFactory', {
  enumerable: true,
  get: function get() {
    return _protractor2.circularProgressBarTestkitFactory;
  }
});

require('regenerator-runtime/runtime');

var _InputProtractor = require('../src/Input/Input.protractor.driver');

var _InputProtractor2 = _interopRequireDefault(_InputProtractor);

var _AddItemProtractor = require('../src/AddItem/AddItem.protractor.driver');

var _AddItemProtractor2 = _interopRequireDefault(_AddItemProtractor);

var _RichTextAreaProtractor = require('../src/RichTextArea/RichTextArea.protractor.driver');

var _RichTextAreaProtractor2 = _interopRequireDefault(_RichTextAreaProtractor);

var _InputAreaProtractor = require('../src/InputArea/InputArea.protractor.driver');

var _InputAreaProtractor2 = _interopRequireDefault(_InputAreaProtractor);

var _ButtonProtractor = require('../src/Backoffice/Button/Button.protractor.driver');

var _ButtonProtractor2 = _interopRequireDefault(_ButtonProtractor);

var _ButtonProtractor3 = require('../src/TPA/Button/Button.protractor.driver');

var _ButtonProtractor4 = _interopRequireDefault(_ButtonProtractor3);

var _InputProtractor3 = require('../src/TPA/Input/Input.protractor.driver');

var _InputProtractor4 = _interopRequireDefault(_InputProtractor3);

var _PageProtractor = require('../src/Page/Page.protractor.driver');

var _PageProtractor2 = _interopRequireDefault(_PageProtractor);

var _DataTableProtractor = require('../src/DataTable/DataTable.protractor.driver');

var _DataTableProtractor2 = _interopRequireDefault(_DataTableProtractor);

var _TableProtractor = require('../src/Table/Table.protractor.driver');

var _TableProtractor2 = _interopRequireDefault(_TableProtractor);

var _BreadcrumbsProtractor = require('../src/Breadcrumbs/Breadcrumbs.protractor.driver');

var _BreadcrumbsProtractor2 = _interopRequireDefault(_BreadcrumbsProtractor);

var _AutoCompleteProtractor = require('../src/AutoComplete/AutoComplete.protractor.driver');

var _AutoCompleteProtractor2 = _interopRequireDefault(_AutoCompleteProtractor);

var _CheckboxProtractor = require('../src/Checkbox/Checkbox.protractor.driver');

var _CheckboxProtractor2 = _interopRequireDefault(_CheckboxProtractor);

var _DropdownProtractor = require('../src/Dropdown/Dropdown.protractor.driver');

var _DropdownProtractor2 = _interopRequireDefault(_DropdownProtractor);

var _ButtonWithOptionsProtractor = require('../src/ButtonWithOptions/ButtonWithOptions.protractor.driver');

var _ButtonWithOptionsProtractor2 = _interopRequireDefault(_ButtonWithOptionsProtractor);

var _LanguagePickerProtractor = require('../src/LanguagePicker/LanguagePicker.protractor.driver');

var _LanguagePickerProtractor2 = _interopRequireDefault(_LanguagePickerProtractor);

var _FilePickerProtractor = require('../src/FilePicker/FilePicker.protractor.driver');

var _FilePickerProtractor2 = _interopRequireDefault(_FilePickerProtractor);

var _RadioGroupProtractor = require('../src/RadioGroup/RadioGroup.protractor.driver');

var _RadioGroupProtractor2 = _interopRequireDefault(_RadioGroupProtractor);

var _TextAreaProtractor = require('../src/TextArea/TextArea.protractor.driver');

var _TextAreaProtractor2 = _interopRequireDefault(_TextAreaProtractor);

var _TextFieldProtractor = require('../src/TextField/TextField.protractor.driver');

var _TextFieldProtractor2 = _interopRequireDefault(_TextFieldProtractor);

var _DropdownCompositeProtractor = require('../src/DropdownComposite/DropdownComposite.protractor.driver');

var _DropdownCompositeProtractor2 = _interopRequireDefault(_DropdownCompositeProtractor);

var _MultiSelectCompositeProtractor = require('../src/MultiSelectComposite/MultiSelectComposite.protractor.driver');

var _MultiSelectCompositeProtractor2 = _interopRequireDefault(_MultiSelectCompositeProtractor);

var _AutoCompleteCompositeProtractor = require('../src/AutoCompleteComposite/AutoCompleteComposite.protractor.driver');

var _AutoCompleteCompositeProtractor2 = _interopRequireDefault(_AutoCompleteCompositeProtractor);

var _RangeProtractor = require('../src/Range/Range.protractor.driver');

var _RangeProtractor2 = _interopRequireDefault(_RangeProtractor);

var _FieldWithSelectionCompositeProtractor = require('../src/Composite/FieldWithSelectionComposite/FieldWithSelectionComposite.protractor.driver');

var _FieldWithSelectionCompositeProtractor2 = _interopRequireDefault(_FieldWithSelectionCompositeProtractor);

var _GoogleAddressInputWithLabelProtractor = require('../src/GoogleAddressInputWithLabel/GoogleAddressInputWithLabel.protractor.driver');

var _GoogleAddressInputWithLabelProtractor2 = _interopRequireDefault(_GoogleAddressInputWithLabelProtractor);

var _ImageViewerProtractor = require('../src/ImageViewer/ImageViewer.protractor.driver');

var _ImageViewerProtractor2 = _interopRequireDefault(_ImageViewerProtractor);

var _TextLinkProtractor = require('../src/TPA/TextLink/TextLink.protractor.driver');

var _TextLinkProtractor2 = _interopRequireDefault(_TextLinkProtractor);

var _PopoverMenuProtractor = require('../src/PopoverMenu/PopoverMenu.protractor.driver');

var _PopoverMenuProtractor2 = _interopRequireDefault(_PopoverMenuProtractor);

var _MultiSelectProtractor = require('../src/MultiSelect/MultiSelect.protractor.driver');

var _MultiSelectProtractor2 = _interopRequireDefault(_MultiSelectProtractor);

var _MultiSelectCheckboxProtractor = require('../src/MultiSelectCheckbox/MultiSelectCheckbox.protractor.driver');

var _MultiSelectCheckboxProtractor2 = _interopRequireDefault(_MultiSelectCheckboxProtractor);

var _DatePickerProtractor = require('../src/DatePicker/DatePicker.protractor.driver');

var _DatePickerProtractor2 = _interopRequireDefault(_DatePickerProtractor);

var _EditableSelectorProtractor = require('../src/EditableSelector/EditableSelector.protractor.driver');

var _EditableSelectorProtractor2 = _interopRequireDefault(_EditableSelectorProtractor);

var _IconWithOptionsProtractor = require('../src/IconWithOptions/IconWithOptions.protractor.driver');

var _IconWithOptionsProtractor2 = _interopRequireDefault(_IconWithOptionsProtractor);

var _SearchProtractor = require('../src/Search/Search.protractor.driver');

var _SearchProtractor2 = _interopRequireDefault(_SearchProtractor);

var _HighlighterProtractor = require('../src/Highlighter/Highlighter.protractor.driver');

var _HighlighterProtractor2 = _interopRequireDefault(_HighlighterProtractor);

var _StatsWidgetProtractor = require('../src/StatsWidget/StatsWidget.protractor.driver');

var _StatsWidgetProtractor2 = _interopRequireDefault(_StatsWidgetProtractor);

var _LoaderProtractor = require('../src/Loader/Loader.protractor.driver');

var _LoaderProtractor2 = _interopRequireDefault(_LoaderProtractor);

var _ModalSelectorLayoutProtractor = require('../src/ModalSelectorLayout/ModalSelectorLayout.protractor.driver');

var _ModalSelectorLayoutProtractor2 = _interopRequireDefault(_ModalSelectorLayoutProtractor);

var _SliderProtractor = require('../src/Slider/Slider.protractor.driver');

var _SliderProtractor2 = _interopRequireDefault(_SliderProtractor);

var _TabsProtractor = require('../src/Tabs/Tabs.protractor.driver');

var _TabsProtractor2 = _interopRequireDefault(_TabsProtractor);

var _DrillViewProtractor = require('../src/SideMenu/DrillView/DrillView.protractor.driver');

var _DrillViewProtractor2 = _interopRequireDefault(_DrillViewProtractor);

var _TooltipProtractor = require('../src/Tooltip/Tooltip.protractor.driver');

var _TooltipProtractor2 = _interopRequireDefault(_TooltipProtractor);

var _FormFieldProtractor = require('../src/FormField/FormField.protractor.driver');

var _FormFieldProtractor2 = _interopRequireDefault(_FormFieldProtractor);

var _TextProtractor = require('../src/Text/Text.protractor.driver');

var _TextProtractor2 = _interopRequireDefault(_TextProtractor);

var _MessageBoxFunctionalLayoutProtractor = require('../src/MessageBox/MessageBoxFunctionalLayout.protractor.driver');

var _MessageBoxFunctionalLayoutProtractor2 = _interopRequireDefault(_MessageBoxFunctionalLayoutProtractor);

var _HeadingProtractor = require('../src/Heading/Heading.protractor.driver');

var _HeadingProtractor2 = _interopRequireDefault(_HeadingProtractor);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var inputTestkitFactory = exports.inputTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_InputProtractor2.default);

var addItemTestkitFactory = exports.addItemTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_AddItemProtractor2.default);

var richTextAreaTestkitFactory = exports.richTextAreaTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_RichTextAreaProtractor2.default);

var inputAreaTestkitFactory = exports.inputAreaTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_InputAreaProtractor2.default);

var buttonTestkitFactory = exports.buttonTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_ButtonProtractor2.default);

var tpaButtonTestkitFactory = exports.tpaButtonTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_ButtonProtractor4.default);

var tpaInputTestkitFactory = exports.tpaInputTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_InputProtractor4.default);

var pageTestkitFactory = exports.pageTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_PageProtractor2.default);

var dataTableTestkitFactory = exports.dataTableTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_DataTableProtractor2.default);

var tableTestkitFactory = exports.tableTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_TableProtractor2.default);

var breadcrumbsTestkitFactory = exports.breadcrumbsTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_BreadcrumbsProtractor2.default);

var autoCompleteTestkitFactory = exports.autoCompleteTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_AutoCompleteProtractor2.default);

var checkboxTestkitFactory = exports.checkboxTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_CheckboxProtractor2.default);

var dropdownTestkitFactory = exports.dropdownTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_DropdownProtractor2.default);

var buttonWithOptionsTestkitFactory = exports.buttonWithOptionsTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_ButtonWithOptionsProtractor2.default);

var languagePickerTestkitFactory = exports.languagePickerTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_LanguagePickerProtractor2.default);

var filePickerTestkitFactory = exports.filePickerTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_FilePickerProtractor2.default);

var radioGroupTestkitFactory = exports.radioGroupTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_RadioGroupProtractor2.default);

var textAreaTestkitFactory = exports.textAreaTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_TextAreaProtractor2.default);

var textFieldTestkitFactory = exports.textFieldTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_TextFieldProtractor2.default);

var dropdownCompositeTestkitFactory = exports.dropdownCompositeTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_DropdownCompositeProtractor2.default);

var multiSelectCompositeTestkitFactory = exports.multiSelectCompositeTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_MultiSelectCompositeProtractor2.default);

var autoCompleteCompositeTestkitFactory = exports.autoCompleteCompositeTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_AutoCompleteCompositeProtractor2.default);

var rangeTestkitFactory = exports.rangeTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_RangeProtractor2.default);

var fieldWithSelectionCompositeTestkitFactory = exports.fieldWithSelectionCompositeTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_FieldWithSelectionCompositeProtractor2.default);

var googleAddressWithInputTestkitFactory = exports.googleAddressWithInputTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_GoogleAddressInputWithLabelProtractor2.default);

var imageViewerTestkitFactory = exports.imageViewerTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_ImageViewerProtractor2.default);

var tpaTextLinkTestkitFactory = exports.tpaTextLinkTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_TextLinkProtractor2.default);

var popoverMenuTestkitFactory = exports.popoverMenuTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_PopoverMenuProtractor2.default);

var multiSelectTestkitFactory = exports.multiSelectTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_MultiSelectProtractor2.default);

var multiSelectCheckboxTestkitFactory = exports.multiSelectCheckboxTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_MultiSelectCheckboxProtractor2.default);

var datePickerTestkitFactory = exports.datePickerTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_DatePickerProtractor2.default);

var editableSelectorTestkitFactory = exports.editableSelectorTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_EditableSelectorProtractor2.default);

var iconWithOptionsTestkitFactory = exports.iconWithOptionsTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_IconWithOptionsProtractor2.default);

var searchTestkitFactory = exports.searchTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_SearchProtractor2.default);

var highlighterTestkitFactory = exports.highlighterTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_HighlighterProtractor2.default);

var statsWidgetTestkitFactory = exports.statsWidgetTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_StatsWidgetProtractor2.default);

var loaderTestkitFactory = exports.loaderTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_LoaderProtractor2.default);

var modalSelectorLayoutTestkitFactory = exports.modalSelectorLayoutTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_ModalSelectorLayoutProtractor2.default);

var sliderTestkitFactory = exports.sliderTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_SliderProtractor2.default);

var tabsTestkitFactory = exports.tabsTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_TabsProtractor2.default);

var drillViewTestkitFactory = exports.drillViewTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_DrillViewProtractor2.default);

var tooltipTestkitFactory = exports.tooltipTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_TooltipProtractor2.default);

var formFieldTestkitFactory = exports.formFieldTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_FormFieldProtractor2.default);

var textTestkitFactory = exports.textTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_TextProtractor2.default);

var messageBoxFunctionalLayoutTestkitFactory = exports.messageBoxFunctionalLayoutTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_MessageBoxFunctionalLayoutProtractor2.default);

var headingTestkitFactory = exports.headingTestkitFactory = (0, _protractor.protractorTestkitFactoryCreator)(_HeadingProtractor2.default);

// wix-ui-backoffice proxy
