// wix-style-react/dist/testkit/puppeteer.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.tooltipTestkitFactory = exports.textTestkitFactory = exports.headingTestkitFactory = exports.tableTestkitFactory = exports.formFieldTestkitFactory = exports.buttonTestkitFactory = exports.inputTestkitFactory = undefined;

require('regenerator-runtime/runtime');

var _puppeteer = require('wix-ui-test-utils/puppeteer');

var _InputPuppeteer = require('../src/Input/Input.puppeteer.driver');

var _InputPuppeteer2 = _interopRequireDefault(_InputPuppeteer);

var _ButtonPuppeteer = require('../src/Backoffice/Button/Button.puppeteer.driver');

var _ButtonPuppeteer2 = _interopRequireDefault(_ButtonPuppeteer);

var _FormFieldPuppeteer = require('../src/FormField/FormField.puppeteer.driver');

var _FormFieldPuppeteer2 = _interopRequireDefault(_FormFieldPuppeteer);

var _TablePuppeteer = require('../src/Table/Table.puppeteer.driver');

var _TablePuppeteer2 = _interopRequireDefault(_TablePuppeteer);

var _HeadingPuppeteer = require('../src/Heading/Heading.puppeteer.driver');

var _HeadingPuppeteer2 = _interopRequireDefault(_HeadingPuppeteer);

var _TextPuppeteer = require('../src/Text/Text.puppeteer.driver');

var _TextPuppeteer2 = _interopRequireDefault(_TextPuppeteer);

var _TooltipPuppeteer = require('../src/Tooltip/Tooltip.puppeteer.driver');

var _TooltipPuppeteer2 = _interopRequireDefault(_TooltipPuppeteer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var inputTestkitFactory = exports.inputTestkitFactory = (0, _puppeteer.puppeteerTestkitFactoryCreator)(_InputPuppeteer2.default);

var buttonTestkitFactory = exports.buttonTestkitFactory = (0, _puppeteer.puppeteerTestkitFactoryCreator)(_ButtonPuppeteer2.default);

var formFieldTestkitFactory = exports.formFieldTestkitFactory = (0, _puppeteer.puppeteerTestkitFactoryCreator)(_FormFieldPuppeteer2.default);

var tableTestkitFactory = exports.tableTestkitFactory = (0, _puppeteer.puppeteerTestkitFactoryCreator)(_TablePuppeteer2.default);

var headingTestkitFactory = exports.headingTestkitFactory = (0, _puppeteer.puppeteerTestkitFactoryCreator)(_HeadingPuppeteer2.default);

var textTestkitFactory = exports.textTestkitFactory = (0, _puppeteer.puppeteerTestkitFactoryCreator)(_TextPuppeteer2.default);

var tooltipTestkitFactory = exports.tooltipTestkitFactory = (0, _puppeteer.puppeteerTestkitFactoryCreator)(_TooltipPuppeteer2.default);
