// wix-style-react/dist/stories/Tabs/TabsExample.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _RadioGroup = require('../../src/RadioGroup');

var _RadioGroup2 = _interopRequireDefault(_RadioGroup);

var _Label = require('../../src/Label');

var _Label2 = _interopRequireDefault(_Label);

var _Input = require('../../src/Input');

var _Input2 = _interopRequireDefault(_Input);

var _TabsTemplate = require('./TabsTemplate');

var _TabsTemplate2 = _interopRequireDefault(_TabsTemplate);

var _Button = require('../../src/Button');

var _Button2 = _interopRequireDefault(_Button);

var _TabsExample = require('./TabsExample.scss');

var _TabsExample2 = _interopRequireDefault(_TabsExample);

var _TextField = require('../../src/TextField');

var _TextField2 = _interopRequireDefault(_TextField);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var SideContentExample = function SideContentExample() {
  return _react2.default.createElement(
    'div',
    { className: _TabsExample2.default.sideContentExample },
    _react2.default.createElement(
      _Button2.default,
      { theme: 'whiteblue' },
      'Cancel'
    ),
    _react2.default.createElement(
      _Button2.default,
      null,
      'Save'
    )
  );
};

var TabsExample = function (_Component) {
  _inherits(TabsExample, _Component);

  function TabsExample() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, TabsExample);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = TabsExample.__proto__ || Object.getPrototypeOf(TabsExample)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      type: '',
      hasDivider: true,
      width: 30,
      showSideContent: false
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(TabsExample, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      return _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
          'div',
          { className: _TabsExample2.default.controlGroup },
          _react2.default.createElement(
            _Label2.default,
            null,
            'Type'
          ),
          _react2.default.createElement(
            'div',
            { className: _TabsExample2.default.radioGroup },
            _react2.default.createElement(
              _RadioGroup2.default,
              {
                display: 'horizontal',
                value: this.state.type,
                onChange: function onChange(type) {
                  return _this2.setState({ type: type });
                }
              },
              _react2.default.createElement(
                _RadioGroup2.default.Radio,
                { value: '' },
                'Default'
              ),
              _react2.default.createElement(
                _RadioGroup2.default.Radio,
                { value: 'compact' },
                'Compact'
              ),
              _react2.default.createElement(
                _RadioGroup2.default.Radio,
                { value: 'compactSide' },
                'Compact (Side)'
              ),
              _react2.default.createElement(
                _RadioGroup2.default.Radio,
                { value: 'uniformSide' },
                'Uniform (Side)'
              ),
              _react2.default.createElement(
                _RadioGroup2.default.Radio,
                { value: 'uniformFull' },
                'Uniform (Full)'
              )
            )
          )
        ),
        this.state.type === 'uniformSide' ? _react2.default.createElement(
          'div',
          { className: _TabsExample2.default.option },
          _react2.default.createElement(
            'div',
            { className: _TabsExample2.default.column, style: { width: '100px' } },
            _react2.default.createElement('br', null),
            _react2.default.createElement(
              _TextField2.default,
              null,
              _react2.default.createElement(
                _Label2.default,
                {
                  appearance: 'T1.1',
                  'for': 'firstName'
                },
                'Tab Width'
              ),
              _react2.default.createElement(_Input2.default, {
                errorMessage: '',
                id: 'firstName',
                placeholder: 'e.g. 100',
                size: 'normal',
                onChange: function onChange(e) {
                  return _this2.setState({ width: e.target.value });
                },
                value: this.state.width,
                suffix: _react2.default.createElement(
                  _Input2.default.Group,
                  null,
                  _react2.default.createElement(_Input2.default.Unit, { value: 'px' }),
                  _react2.default.createElement(_Input2.default.Ticker, {
                    onDown: function onDown() {
                      _this2.setState({ width: _this2.state.width - 1 });
                    },
                    onUp: function onUp() {
                      _this2.setState({ width: _this2.state.width + 1 });
                    }
                  })
                ),
                theme: 'normal',
                type: 'number'
              })
            ),
            _react2.default.createElement('br', null)
          )
        ) : null,
        _react2.default.createElement(
          'div',
          { className: _TabsExample2.default.controlGroup },
          _react2.default.createElement(
            _Label2.default,
            null,
            'Divider'
          ),
          _react2.default.createElement(
            'div',
            { className: _TabsExample2.default.radioGroup },
            _react2.default.createElement(
              _RadioGroup2.default,
              {
                display: 'horizontal',
                value: this.state.hasDivider,
                onChange: function onChange(hasDivider) {
                  return _this2.setState({ hasDivider: hasDivider });
                }
              },
              _react2.default.createElement(
                _RadioGroup2.default.Radio,
                { value: true },
                'Visible'
              ),
              _react2.default.createElement(
                _RadioGroup2.default.Radio,
                { value: false },
                'Hidden'
              )
            )
          )
        ),
        _react2.default.createElement(
          'div',
          { className: _TabsExample2.default.controlGroup },
          _react2.default.createElement(
            _Label2.default,
            null,
            'Side Content'
          ),
          _react2.default.createElement(
            'div',
            { className: _TabsExample2.default.radioGroup },
            _react2.default.createElement(
              _RadioGroup2.default,
              {
                display: 'horizontal',
                value: this.state.showSideContent,
                onChange: function onChange(showSideContent) {
                  return _this2.setState({ showSideContent: showSideContent });
                }
              },
              _react2.default.createElement(
                _RadioGroup2.default.Radio,
                { value: false },
                'None'
              ),
              _react2.default.createElement(
                _RadioGroup2.default.Radio,
                { value: true },
                'Buttons'
              )
            )
          )
        ),
        _react2.default.createElement(
          'div',
          { className: _TabsExample2.default.preview },
          _react2.default.createElement(_TabsTemplate2.default, {
            onChange: this.props.onChange,
            type: this.state.type,
            hasDivider: this.state.hasDivider,
            width: this.state.width,
            sideContent: this.state.showSideContent ? _react2.default.createElement(SideContentExample, null) : undefined
          })
        )
      );
    }
  }]);

  return TabsExample;
}(_react.Component);

TabsExample.propTypes = {
  onChange: _propTypes2.default.func
};

exports.default = TabsExample;
