// wix-style-react/dist/stories/Tabs/TabsTemplate.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactElementToJsxString = require('react-element-to-jsx-string');

var _reactElementToJsxString2 = _interopRequireDefault(_reactElementToJsxString);

var _Tabs = require('../../src/Tabs');

var _Tabs2 = _interopRequireDefault(_Tabs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var items = [{ id: 1, title: 'First Item' }, { id: 2, title: 'Second Item' }, { id: 3, title: 'Third Item' }, { id: 4, title: 'Fourth item' }, { id: 6, title: 'Fifth item' }, { id: 7, title: 'Sixth item' }, { id: 8, title: 'A very long tab item that may not fit' }];

var TabsTemplate = function (_Component) {
  _inherits(TabsTemplate, _Component);

  function TabsTemplate() {
    var _ref;

    var _temp, _this, _ret;

    _classCallCheck(this, TabsTemplate);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = TabsTemplate.__proto__ || Object.getPrototypeOf(TabsTemplate)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      activeId: items[0].id
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  _createClass(TabsTemplate, [{
    key: 'componentDidUpdate',
    value: function componentDidUpdate(props) {
      props.onChange(getExampleCode(this.getComponent()));
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.props.onChange(getExampleCode(this.getComponent()));
    }
  }, {
    key: 'getComponent',
    value: function getComponent() {
      var _this2 = this;

      var activeId = this.state.activeId;
      var _props = this.props,
          type = _props.type,
          hasDivider = _props.hasDivider,
          width = _props.width,
          sideContent = _props.sideContent;

      var props = { items: items, activeId: activeId, hasDivider: hasDivider };

      if (type) {
        props.type = type;
      }

      if (type === 'uniformSide' && width) {
        props.width = width;
      }

      return _react2.default.createElement(_Tabs2.default, _extends({
        sideContent: sideContent,
        onClick: function onClick(item) {
          return _this2.setState({ activeId: item.id });
        },
        dataHook: 'story-tabs'
      }, props));
    }
  }, {
    key: 'render',
    value: function render() {
      return this.getComponent();
    }
  }]);

  return TabsTemplate;
}(_react.Component);

TabsTemplate.propTypes = {
  onChange: _propTypes2.default.func.isRequired,
  type: _propTypes2.default.oneOf(_Tabs2.default.tabTypes),
  hasDivider: _propTypes2.default.bool,
  width: _propTypes2.default.string,
  sideContent: _propTypes2.default.node
};

function getExampleCode(element) {
  return (0, _reactElementToJsxString2.default)(element, {
    filterProps: ['onClick'],
    showDefaultProps: false
  });
}

exports.default = TabsTemplate;
