// wix-ui-backoffice/dist/src/components/core/CoreText/Text.driver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var test_utils_1 = require("stylable/test-utils");
var Text_st_css_1 = require("./Text.st.css");
exports.textDriverFactory = function (_a) {
    var element = _a.element;
    var stylableDOMUtil = new test_utils_1.StylableDOMUtil(Text_st_css_1.default);
    return {
        /** check if element exists */
        exists: function () { return !!element; },
        /** check if component has ellipsis */
        hasEllipsis: function () { return stylableDOMUtil.hasStyleState(element, 'ellipsis'); },
        /** check if element has title attribute */
        hasTitleAttribute: function () { return element.getAttribute('title') !== null; },
        /** check if element has title attribute */
        getTitle: function () { return element.title; },
        /** get the rendered tag name */
        getTagName: function () { return element.tagName.toLowerCase(); },
        /** get the rendered content */
        getText: function () { return element.innerHTML; }
    };
};
