// wix-ui-core/dist/src/testkit/enzyme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var enzyme_1 = require("wix-ui-test-utils/enzyme");
var Input_driver_1 = require("../components/Input/Input.driver");
//Ugly workaround for now
exports.inputTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Input_driver_1.inputDriverFactory);
var Badge_driver_1 = require("../components/Badge/Badge.driver");
exports.badgeTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Badge_driver_1.badgeDriverFactory);
var Tooltip_driver_1 = require("../components/Tooltip/Tooltip.driver");
exports.tooltipTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Tooltip_driver_1.tooltipDriverFactory);
var LinearProgressBar_driver_1 = require("../components/LinearProgressBar/LinearProgressBar.driver");
exports.linearProgressBarTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(LinearProgressBar_driver_1.linearProgressBarDriverFactory);
var CircularProgressBar_driver_1 = require("../components/CircularProgressBar/CircularProgressBar.driver");
exports.circularProgressBarTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(CircularProgressBar_driver_1.circularProgressBarDriverFactory);
var Pagination_driver_1 = require("../components/Pagination/Pagination.driver");
exports.paginationTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Pagination_driver_1.paginationDriverFactory);
var Divider_driver_1 = require("../components/Divider/Divider.driver");
exports.dividerTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Divider_driver_1.dividerDriverFactory);
var Checkbox_driver_1 = require("../components/Checkbox/Checkbox.driver");
exports.checkboxTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Checkbox_driver_1.checkboxDriverFactory);
// Stylable
var ToggleSwitch_driver_1 = require("../components/ToggleSwitch/ToggleSwitch.driver");
exports.toggleSwitchTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(ToggleSwitch_driver_1.toggleSwitchDriverFactory);
var Button_driver_1 = require("../components/Button/Button.driver");
exports.buttonTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Button_driver_1.buttonDriverFactory);
var Badge_driver_2 = require("../components/StylableBadge/Badge.driver");
exports.stylableBadgeTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Badge_driver_2.badgeDriverFactory);
var RadioButton_driver_1 = require("../components/RadioButton/RadioButton.driver");
exports.radioButtonTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(RadioButton_driver_1.radioButtonDriverFactory);
var Autocomplete_driver_1 = require("../components/Autocomplete/Autocomplete.driver");
exports.autocompleteTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Autocomplete_driver_1.autocompleteDriverFactory);
var Slider_driver_1 = require("../components/Slider/Slider.driver");
exports.sliderTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Slider_driver_1.sliderDriverFactory);
var AddressInput_driver_1 = require("../components/AddressInput/AddressInput.driver");
exports.addressInputTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(AddressInput_driver_1.addressInputDriverFactory);
var Label_driver_1 = require("../components/Label/Label.driver");
exports.labelTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Label_driver_1.labelDriverFactory);
var TimePicker_driver_1 = require("../components/TimePicker/TimePicker.driver");
exports.timePickerTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(TimePicker_driver_1.timePickerDriverFactory);
var LabelWithOptions_driver_1 = require("../components/LabelWithOptions/LabelWithOptions.driver");
exports.labelWithOptionsTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(LabelWithOptions_driver_1.labelWithOptionsDriverFactory);
var Thumbnail_driver_1 = require("../components/Thumbnail/Thumbnail.driver");
exports.thumbnailTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Thumbnail_driver_1.thumbnailDriverFactory);
var Popover_driver_1 = require("../components/Popover/Popover.driver");
exports.popoverTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Popover_driver_1.popoverDriverFactory);
var Link_driver_1 = require("../components/Link/Link.driver");
exports.linkTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(Link_driver_1.linkDriverFactory);
var NavStepper_driver_1 = require("../components/NavStepper/NavStepper.driver");
exports.NavStepperDriver = NavStepper_driver_1.NavStepperDriver;
exports.navStepperTestkitFactory = enzyme_1.enzymeTestkitFactoryCreator(NavStepper_driver_1.navStepperDriverFactory);
