// wix-ui-core/dist/src/testkit/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var vanilla_1 = require("wix-ui-test-utils/vanilla");
var Input_driver_1 = require("../components/Input/Input.driver");
exports.inputTestkitFactory = vanilla_1.testkitFactoryCreator(Input_driver_1.inputDriverFactory);
var Pagination_driver_1 = require("../components/Pagination/Pagination.driver");
exports.paginationTestkitFactory = vanilla_1.testkitFactoryCreator(Pagination_driver_1.paginationDriverFactory);
var Badge_driver_1 = require("../components/Badge/Badge.driver");
exports.badgeTestkitFactory = vanilla_1.testkitFactoryCreator(Badge_driver_1.badgeDriverFactory);
var Tooltip_driver_1 = require("../components/Tooltip/Tooltip.driver");
exports.tooltipTestkitFactory = vanilla_1.testkitFactoryCreator(Tooltip_driver_1.tooltipDriverFactory);
var Divider_driver_1 = require("../components/Divider/Divider.driver");
exports.dividerTestkitFactory = vanilla_1.testkitFactoryCreator(Divider_driver_1.dividerDriverFactory);
var Checkbox_driver_1 = require("../components/Checkbox/Checkbox.driver");
exports.checkboxTestkitFactory = vanilla_1.testkitFactoryCreator(Checkbox_driver_1.checkboxDriverFactory);
//Stylable
var ToggleSwitch_driver_1 = require("../components/ToggleSwitch/ToggleSwitch.driver");
exports.toggleSwitchTestkitFactory = vanilla_1.testkitFactoryCreator(ToggleSwitch_driver_1.toggleSwitchDriverFactory);
var Button_driver_1 = require("../components/Button/Button.driver");
exports.buttonTestkitFactory = vanilla_1.testkitFactoryCreator(Button_driver_1.buttonDriverFactory);
var LinearProgressBar_driver_1 = require("../components/LinearProgressBar/LinearProgressBar.driver");
exports.linearProgressBarTestkitFactory = vanilla_1.testkitFactoryCreator(LinearProgressBar_driver_1.linearProgressBarDriverFactory);
var CircularProgressBar_driver_1 = require("../components/CircularProgressBar/CircularProgressBar.driver");
exports.circularProgressBarTestkitFactory = vanilla_1.testkitFactoryCreator(CircularProgressBar_driver_1.circularProgressBarDriverFactory);
var Badge_driver_2 = require("../components/StylableBadge/Badge.driver");
exports.stylableBadgeTestkitFactory = vanilla_1.testkitFactoryCreator(Badge_driver_2.badgeDriverFactory);
var RadioButton_driver_1 = require("../components/RadioButton/RadioButton.driver");
exports.radioButtonTestkitFactory = vanilla_1.testkitFactoryCreator(RadioButton_driver_1.radioButtonDriverFactory);
var Autocomplete_driver_1 = require("../components/Autocomplete/Autocomplete.driver");
exports.autocompleteTestkitFactory = vanilla_1.testkitFactoryCreator(Autocomplete_driver_1.autocompleteDriverFactory);
var Slider_driver_1 = require("../components/Slider/Slider.driver");
exports.sliderTestkitFactory = vanilla_1.testkitFactoryCreator(Slider_driver_1.sliderDriverFactory);
var AddressInput_driver_1 = require("../components/AddressInput/AddressInput.driver");
exports.addressInputTestkitFactory = vanilla_1.testkitFactoryCreator(AddressInput_driver_1.addressInputDriverFactory);
var Label_driver_1 = require("../components/Label/Label.driver");
exports.labelTestkitFactory = vanilla_1.testkitFactoryCreator(Label_driver_1.labelDriverFactory);
var TimePicker_driver_1 = require("../components/TimePicker/TimePicker.driver");
exports.timePickerTestkitFactory = vanilla_1.testkitFactoryCreator(TimePicker_driver_1.timePickerDriverFactory);
var LabelWithOptions_driver_1 = require("../components/LabelWithOptions/LabelWithOptions.driver");
exports.labelWithOptionsTestkitFactory = vanilla_1.testkitFactoryCreator(LabelWithOptions_driver_1.labelWithOptionsDriverFactory);
var Thumbnail_driver_1 = require("../components/Thumbnail/Thumbnail.driver");
exports.thumbnailTestkitFactory = vanilla_1.testkitFactoryCreator(Thumbnail_driver_1.thumbnailDriverFactory);
var Popover_driver_1 = require("../components/Popover/Popover.driver");
exports.popoverTestkitFactory = vanilla_1.testkitFactoryCreator(Popover_driver_1.popoverDriverFactory);
var Link_driver_1 = require("../components/Link/Link.driver");
exports.linkTestkitFactory = vanilla_1.testkitFactoryCreator(Link_driver_1.linkDriverFactory);
