// wix-ui-backoffice/dist/src/components/CloseButton/CloseButton.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var defaults = require("lodash/defaults");
var driver_factory_1 = require("wix-ui-test-utils/driver-factory");
var CloseButton_driver_1 = require("./CloseButton.driver");
var testkit_1 = require("../../testkit");
var enzyme_1 = require("../../testkit/enzyme");
var testkitTests_1 = require("../../common/testkitTests");
var CloseButton_1 = require("./CloseButton");
var constants_1 = require("./constants");
var utils_1 = require("../../utils");
describe('CloseButton', function () {
    var createDriver = driver_factory_1.createDriverFactory(CloseButton_driver_1.closeButtonDriverFactory);
    var CloseButtonWithDefaults = withDefaultsHOC(CloseButton_1.CloseButton, {
        children: 'Click me!'
    });
    describe('skin prop', function () {
        it('should be standard by default', function () {
            var driver = createDriver(React.createElement(CloseButtonWithDefaults, null));
            expect(driver.getSkin()).toBe(constants_1.Skin.standard);
        });
        utils_1.enumValues(constants_1.Skin).forEach(function (skin) {
            it("should be '" + skin + "'", function () {
                var driver = createDriver(React.createElement(CloseButtonWithDefaults, { skin: skin }));
                expect(driver.getSkin()).toBe(skin);
            });
        });
    });
    describe('size prop', function () {
        it('should be small by default', function () {
            var driver = createDriver(React.createElement(CloseButtonWithDefaults, null));
            expect(driver.getSize()).toBe(constants_1.Size.small);
        });
        utils_1.enumValues(constants_1.Size).forEach(function (size) {
            it("should be '" + size + "'", function () {
                var driver = createDriver(React.createElement(CloseButtonWithDefaults, { size: size }));
                expect(driver.getSize()).toBe(size);
            });
        });
    });
    testkitTests_1.runTestkitExistsSuite({
        Element: React.createElement(CloseButton_1.CloseButton, null),
        testkitFactory: testkit_1.closeButtonTestkitFactory,
        enzymeTestkitFactory: enzyme_1.closeButtonTestkitFactory
    });
});
// TODO: consider putting this in 'test/utils'
/**
 * Create a Component with applied default props.
 * The new component can receive Partial<P> instead of P.
 */
function withDefaultsHOC(Component, defaultProps) {
    return function (partialProps) {
        return React.createElement(Component, defaults({}, partialProps, defaultProps));
    };
}
